--- mozilla/toolkit/components/Makefile.in.orig	2007-11-21 23:42:23.000000000 +0200
+++ mozilla/toolkit/components/Makefile.in	2007-11-22 03:08:21.000000000 +0200
@@ -113,19 +113,21 @@ endif # MOZ_THUNDERBIRD
 ifdef MOZ_XUL
 DIRS += help
 endif
 
 ifndef MOZ_SUITE
 # XXX Suite doesn't want these just yet
 ifdef MOZ_XPINSTALL
 ifdef MOZ_RDF
+ifdef MOZ_STORAGE
 DIRS +=	downloads
 endif
 endif
+endif
 endif # MOZ_SUITE
 
 ifdef MOZ_URL_CLASSIFIER
 DIRS += url-classifier
 endif
 
 DIRS += \
 	commandlines \
--- mozilla/toolkit/components/build/Makefile.in.orig	2007-05-22 03:03:35.000000000 +0300
+++ mozilla/toolkit/components/build/Makefile.in	2007-11-22 03:10:01.000000000 +0200
@@ -111,20 +111,22 @@ endif
 SHARED_LIBRARY_LIBS = \
 	../startup/src/$(LIB_PREFIX)appstartup_s.$(LIB_SUFFIX) \
 	$(NULL)
 
 ifndef MOZ_SUITE
 # XXX Suite isn't ready to build this just yet
 ifdef MOZ_XPINSTALL
 ifdef MOZ_RDF
+ifdef MOZ_STORAGE
 SHARED_LIBRARY_LIBS += ../downloads/src/$(LIB_PREFIX)download_s.$(LIB_SUFFIX)
 endif
 endif
 endif
+endif
 
 ifdef ALERTS_SERVICE
 SHARED_LIBRARY_LIBS += ../alerts/src/$(LIB_PREFIX)alerts_s.$(LIB_SUFFIX)
 endif
 ifdef ALERTS_SERVICE_MAC
 SHARED_LIBRARY_LIBS += ../alerts/src/mac/$(LIB_PREFIX)alerts_s.$(LIB_SUFFIX)
 SHARED_LIBRARY_LIBS += ../alerts/src/mac/growl/$(LIB_PREFIX)growl_s.$(LIB_SUFFIX)
 endif
--- mozilla/toolkit/components/build/nsToolkitCompsModule.cpp.orig	2007-05-02 23:16:32.000000000 +0300
+++ mozilla/toolkit/components/build/nsToolkitCompsModule.cpp	2007-11-22 03:09:20.000000000 +0200
@@ -43,20 +43,22 @@
 #ifdef ALERTS_SERVICE
 #include "nsAlertsService.h"
 #endif
 
 #ifndef MOZ_SUITE
 // XXX Suite isn't ready to include this just yet
 #ifdef MOZ_XPINSTALL
 #ifdef MOZ_RDF
+#ifdef MOZ_STORAGE
 #include "nsDownloadManager.h"
 #include "nsDownloadProxy.h"
 #endif
 #endif
+#endif
 #endif // MOZ_SUITE
 
 #include "nsTypeAheadFind.h"
 
 #ifdef MOZ_URL_CLASSIFIER
 #include "nsUrlClassifierDBService.h"
 #include "nsUrlClassifierStreamUpdater.h"
 #include "nsUrlClassifierUtils.h"
@@ -74,20 +76,22 @@ NS_GENERIC_FACTORY_CONSTRUCTOR(nsUserInf
 #ifdef ALERTS_SERVICE
 NS_GENERIC_FACTORY_CONSTRUCTOR(nsAlertsService)
 #endif
 
 #ifndef MOZ_SUITE
 // XXX Suite isn't ready to include this just yet
 #ifdef MOZ_XPINSTALL
 #ifdef MOZ_RDF
+#ifdef MOZ_STORAGE
 NS_GENERIC_FACTORY_CONSTRUCTOR_INIT(nsDownloadManager, Init) 
 NS_GENERIC_FACTORY_CONSTRUCTOR(nsDownloadProxy)
 #endif
 #endif
+#endif
 #endif // MOZ_SUITE
 
 NS_GENERIC_FACTORY_CONSTRUCTOR(nsTypeAheadFind)
 
 #ifdef MOZ_URL_CLASSIFIER
 NS_GENERIC_FACTORY_SINGLETON_CONSTRUCTOR(nsUrlClassifierDBService,
                                          nsUrlClassifierDBService::GetInstance)
 NS_GENERIC_FACTORY_CONSTRUCTOR(nsUrlClassifierStreamUpdater)
@@ -116,26 +120,28 @@ static const nsModuleComponentInfo compo
     NS_ALERTSSERVICE_CID, 
     NS_ALERTSERVICE_CONTRACTID,
     nsAlertsServiceConstructor },
 #endif
 #ifndef MOZ_SUITE
 // XXX Suite isn't ready to include this just yet
 #ifdef MOZ_XPINSTALL
 #ifdef MOZ_RDF
+#ifdef MOZ_STORAGE
   { "Download Manager",
     NS_DOWNLOADMANAGER_CID,
     NS_DOWNLOADMANAGER_CONTRACTID,
     nsDownloadManagerConstructor },
   { "Download",
     NS_DOWNLOAD_CID,
     NS_TRANSFER_CONTRACTID,
     nsDownloadProxyConstructor },
 #endif
 #endif
+#endif
 #endif // MOZ_SUITE
   { "TypeAheadFind Component",
     NS_TYPEAHEADFIND_CID,
     NS_TYPEAHEADFIND_CONTRACTID,
     nsTypeAheadFindConstructor
   },
 #ifdef MOZ_URL_CLASSIFIER
   { "Url Classifier DB Service",
