# https://bugzilla.mozilla.org/show_bug.cgi?id=437018
--- mozilla/widget/src/gtk2/nsWindow.cpp.orig	2008-06-03 11:38:27.000000000 +0300
+++ mozilla/widget/src/gtk2/nsWindow.cpp	2008-06-03 11:43:37.000000000 +0300
@@ -531,10 +531,35 @@ nsWindow::SetParent(nsIWidget *aNewParen
 
     if (!mShell && mDrawingarea) {
         moz_drawingarea_reparent(mDrawingarea, newParentWindow);
     } else {
         NS_NOTREACHED("nsWindow::SetParent - reparenting a non-child window");
+
+        // Get the container
+        gpointer user_data = NULL;
+        gdk_window_get_user_data(newParentWindow, &user_data);
+        if (!user_data)
+           return NS_OK;
+        // Make sure its really a container
+        MozContainer *parent_container = MOZ_CONTAINER(user_data);
+        if (!parent_container)
+           return NS_OK;
+        GtkWidget *temp_shell = GTK_WIDGET(gtk_widget_get_toplevel(GTK_WIDGET(parent_container)));
+        if (!mShell)
+           return NS_OK;
+
+        g_object_steal_data(G_OBJECT(mShell), "nsWindow");
+        mShell = temp_shell;
+        moz_drawingarea_reparent(mDrawingarea, newParentWindow);
+
+        g_object_set_data(G_OBJECT(mShell), "nsWindow", this);
+        g_signal_connect(G_OBJECT(mShell), "configure_event",
+                         G_CALLBACK(configure_event_cb), NULL);
+        g_signal_connect(G_OBJECT(mShell), "delete_event",
+                         G_CALLBACK(delete_event_cb), NULL);
+        g_signal_connect(G_OBJECT(mShell), "window_state_event",
+                         G_CALLBACK(window_state_event_cb), NULL);
     }
     return NS_OK;
 }
 
 NS_IMETHODIMP
