--- mozilla/toolkit/mozapps/extensions/src/nsExtensionManager.js.in.orig	2008-02-29 14:22:41.000000000 +0200
+++ mozilla/toolkit/mozapps/extensions/src/nsExtensionManager.js.in	2008-03-07 14:49:16.000000000 +0200
@@ -5618,10 +5618,11 @@ ExtensionManager.prototype = {
    */
   removeDownload: function(url) {
     for (var i = 0; i < this._transactions.length; ++i) {
       if (this._transactions[i].containsURL(url)) {
         this._transactions[i].removeDownload(url);
+        gOS.notifyObservers(this._transactions[i], "xpinstall-progress", "cancel");
         return;
       }
     } 
   },
   
