--- mozilla/toolkit/mozapps/extensions/src/nsExtensionManager.js.in.orig	2008-03-20 18:40:36.000000000 +0200
+++ mozilla/toolkit/mozapps/extensions/src/nsExtensionManager.js.in	2008-03-20 18:40:45.000000000 +0200
@@ -4063,10 +4063,11 @@ ExtensionManager.prototype = {
         
     var extensionSectionHeader = "[ExtensionDirs]\r\n";
     fos.write(extensionSectionHeader, extensionSectionHeader.length);
     for (var i = 0; i < validExtensions.length; ++i) {
       var e = validExtensions[i];
+      if (!e.location) continue;
       var itemLocation = e.location.getItemLocation(e.id).QueryInterface(nsILocalFile);
       var descriptor = getAbsoluteDescriptor(itemLocation);
       var line = "Extension" + i + "=" + descriptor + "\r\n";
       fos.write(line, line.length);
     }
