--- mozilla/content/html/content/src/nsHTMLIFrameElement.cpp.orig	2008-06-21 13:31:06.000000000 +0300
+++ mozilla/content/html/content/src/nsHTMLIFrameElement.cpp	2008-06-21 13:39:12.000000000 +0300
@@ -34,13 +34,15 @@
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 #include "nsIDOMHTMLIFrameElement.h"
 #include "nsGenericHTMLElement.h"
-#include "nsIDOMGetSVGDocument.h"
 #include "nsIDOMDocument.h"
+#ifdef MOZ_SVG
+#include "nsIDOMGetSVGDocument.h"
 #include "nsIDOMSVGDocument.h"
+#endif
 #include "nsGkAtoms.h"
 #include "nsPresContext.h"
 #include "nsIPresShell.h"
 #include "nsIDocument.h"
 #include "nsMappedAttributes.h"
--- mozilla/content/html/content/src/nsHTMLObjectElement.cpp.orig	2008-06-21 13:31:06.000000000 +0300
+++ mozilla/content/html/content/src/nsHTMLObjectElement.cpp	2008-06-21 13:40:01.000000000 +0300
@@ -40,13 +40,15 @@
 #include "nsObjectLoadingContent.h"
 #include "nsGkAtoms.h"
 #include "nsDOMError.h"
 #include "nsIDocument.h"
 #include "nsIDOMDocument.h"
+#ifdef MOZ_SVG
 #include "nsIDOMSVGDocument.h"
-#include "nsIDOMHTMLObjectElement.h"
 #include "nsIDOMGetSVGDocument.h"
+#endif
+#include "nsIDOMHTMLObjectElement.h"
 #include "nsIFormSubmission.h"
 #include "nsIObjectFrame.h"
 #include "nsIPluginInstance.h"
 #include "nsIPluginInstanceInternal.h"
 
