# https://bugzilla.mozilla.org/show_bug.cgi?id=469756
diff -r e8695d16e25e widget/src/gtk2/nsWindow.cpp
--- a/widget/src/gtk2/nsWindow.cpp	Tue Dec 23 11:09:31 2008 +0200
+++ b/widget/src/gtk2/nsWindow.cpp	Tue Dec 23 11:09:53 2008 +0200
@@ -3963,16 +3963,20 @@ nsWindow::NativeCreate(nsIWidget        
     g_object_set_data(G_OBJECT(mDrawingarea->inner_window), "nsWindow",
                       this);
 
     g_object_set_data(G_OBJECT(mDrawingarea->clip_window), "mozdrawingarea",
                       mDrawingarea);
     g_object_set_data(G_OBJECT(mDrawingarea->inner_window), "mozdrawingarea",
                       mDrawingarea);
 
+    gdk_input_set_extension_events(mDrawingarea->inner_window,
+                                   gdk_window_get_events(mDrawingarea->inner_window),
+                                   GDK_EXTENSION_EVENTS_CURSOR);
+
     if (mContainer)
         g_object_set_data(G_OBJECT(mContainer), "nsWindow", this);
 
     if (mShell)
         g_object_set_data(G_OBJECT(mShell), "nsWindow", this);
 
     // attach listeners for events
     if (mShell) {
