--- m.orig/layout/generic/nsObjectFrame.cpp.orig	2008-09-11 12:17:26.000000000 +0300
+++ m/layout/generic/nsObjectFrame.cpp	2008-09-11 12:31:33.000000000 +0300
@@ -1805,23 +1805,26 @@ MatchPluginName(nsPluginInstanceOwner *a
                  aPluginName,
                  strlen(aPluginName)) == 0;
 }
 
 static PRBool
 DoDelayedStop(nsPluginInstanceOwner *aInstanceOwner, PRBool aDelayedStop)
 {
   // Don't delay stopping QuickTime (bug 425157), Flip4Mac (bug 426524),
-  // XStandard (bug 430219), CMISS Zinc (bug 429604).
+  // XStandard (bug 430219), CMISS Zinc (bug 429604). ARM Flash (454756)
   if (aDelayedStop
 #ifndef XP_WIN
       && !::MatchPluginName(aInstanceOwner, "QuickTime")
       && !::MatchPluginName(aInstanceOwner, "Flip4Mac")
       && !::MatchPluginName(aInstanceOwner, "XStandard plugin")
       && !::MatchPluginName(aInstanceOwner, "CMISS Zinc Plugin")
+#ifdef USE_ARM_KUSER
+      && !::MatchPluginName(aInstanceOwner, "Shockwave Flash")
+#endif
 #endif
       ) {
     nsCOMPtr<nsIRunnable> evt = new nsStopPluginRunnable(aInstanceOwner);
     NS_DispatchToCurrentThread(evt);
     return PR_TRUE;
   }
   return PR_FALSE;
 }
