Index: mozilla/security/manager/pki/resources/content/certViewer.xul
===================================================================
--- mozilla.orig/security/manager/pki/resources/content/certViewer.xul
+++ mozilla/security/manager/pki/resources/content/certViewer.xul
@@ -53,17 +53,17 @@
   buttonlabelaccept="&certmgr.close.label;"
   buttonaccesskeyaccept="&certmgr.close.accesskey;"
   onload="setWindowName();">
 
 <script type="application/x-javascript" src="chrome://global/content/strres.js"/>
 <script type="application/x-javascript" src="chrome://pippki/content/viewCertDetails.js"/>
 <script type="application/x-javascript" src="chrome://pippki/content/pippki.js"/>
 
-  <tabbox>
+  <tabbox flex="1" style="overflow: auto">
     <tabs>
       <tab id="general_tab" label="&certmgr.detail.general_tab.title;"
            accesskey="&certmgr.detail.general_tab.accesskey;"/>
       <tab id="prettyprint_tab" label="&certmgr.detail.prettyprint_tab.title;"
            accesskey="&certmgr.detail.prettyprint_tab.accesskey;"/>
     </tabs>
     <tabpanels>
       <vbox id="general_info"/>
