--- mozilla/security/manager/pki/resources/content/certViewer.xul.orig	2008-09-02 10:16:18.000000000 -0500
+++ mozilla/security/manager/pki/resources/content/certViewer.xul	2008-09-19 15:04:32.000000000 -0500
@@ -56,11 +56,11 @@
 
 <script type="application/x-javascript" src="chrome://global/content/strres.js"/>
 <script type="application/x-javascript" src="chrome://pippki/content/viewCertDetails.js"/>
 <script type="application/x-javascript" src="chrome://pippki/content/pippki.js"/>
 
-  <tabbox>
+  <tabbox flex="1" style="overflow: auto">
     <tabs>
       <tab id="general_tab" label="&certmgr.detail.general_tab.title;"
            accesskey="&certmgr.detail.general_tab.accesskey;"/>
       <tab id="prettyprint_tab" label="&certmgr.detail.prettyprint_tab.title;"
            accesskey="&certmgr.detail.prettyprint_tab.accesskey;"/>
