Index: mozilla/layout/base/nsCSSRendering.cpp
===================================================================
--- mozilla.orig/layout/base/nsCSSRendering.cpp
+++ mozilla/layout/base/nsCSSRendering.cpp
@@ -1425,17 +1425,18 @@ SetupBackgroundClip(gfxContext *aCtx, PR
                     gfxContextAutoSaveRestore* aAutoSR,
                     /* OUT: */
                     nsRect* aBGClipArea, nsRect* aDirtyRect,
                     gfxRect* aDirtyRectGfx)
 {
   *aBGClipArea = aBorderArea;
   PRBool radiiAreOuter = PR_TRUE;
   gfxCornerSizes clippedRadii = aBGRadii;
-  if (aBackgroundClip != NS_STYLE_BG_CLIP_BORDER) {
+  if (aForFrame->GetStylePosition()->mBoxSizing != NS_STYLE_BOX_SIZING_BORDER &&
+      aBackgroundClip != NS_STYLE_BG_CLIP_BORDER) {
     NS_ASSERTION(aBackgroundClip == NS_STYLE_BG_CLIP_PADDING,
                  "unexpected background-clip");
     nsMargin border = aForFrame->GetUsedBorder();
     aForFrame->ApplySkipSides(border);
     aBGClipArea->Deflate(border);
 
     if (aHaveRoundedCorners) {
       gfxFloat borderSizes[4] = {
