diff -r 04cecb0ec24c js/src/jscntxt.cpp
--- a/js/src/jscntxt.cpp	Sat Nov 22 01:53:37 2008 +0100
+++ b/js/src/jscntxt.cpp	Sat Nov 22 11:20:28 2008 +0100
@@ -1374,20 +1374,18 @@ js_GetErrorMessage(void *userRef, const 
 
 JSBool
 js_ResetOperationCount(JSContext *cx)
 {
     JS_ASSERT(cx->operationCount <= 0);
 
 #if JS_HAS_OPERATION_COUNT
     JSScript *script;
 
-    JS_ASSERT(cx->operationLimit > 0);
-
     cx->operationCount = (int32) cx->operationLimit;
     if (cx->operationCallbackIsSet)
         return cx->operationCallback(cx);
 
     if (cx->operationCallback) {
         /*
          * Invoke the deprecated branch callback. It may be called only when
          * the top-most frame is scripted or JSOPTION_NATIVE_BRANCH_CALLBACK
          * is set.
