#BMO 507592, NB#128061
diff -r ac2a11be9695 content/html/content/src/nsHTMLInputElement.cpp
--- a/content/html/content/src/nsHTMLInputElement.cpp	Sat Aug 01 17:12:29 2009 +0300
+++ b/content/html/content/src/nsHTMLInputElement.cpp	Sat Aug 01 22:52:53 2009 +0300
@@ -1399,8 +1399,6 @@ nsHTMLInputElement::Select()
 
   nsCOMPtr<nsPresContext> presContext = GetPresContext();
   if (state == eInactiveWindow) {
-    if (fm)
-      fm->SetFocus(this, nsIFocusManager::FLAG_NOSCROLL);
     SelectAll(presContext);
     return NS_OK;
   }
diff -r ac2a11be9695 content/html/content/src/nsHTMLTextAreaElement.cpp
--- a/content/html/content/src/nsHTMLTextAreaElement.cpp	Sat Aug 01 17:12:29 2009 +0300
+++ b/content/html/content/src/nsHTMLTextAreaElement.cpp	Sat Aug 01 22:52:53 2009 +0300
@@ -313,8 +313,6 @@ nsHTMLTextAreaElement::Select()
 
   nsCOMPtr<nsPresContext> presContext = GetPresContext();
   if (state == eInactiveWindow) {
-    if (fm)
-      fm->SetFocus(this, nsIFocusManager::FLAG_NOSCROLL);
     SelectAll(presContext);
     return NS_OK;
   }
