#NB 127717 BMO 507811
diff -r a5753f35f143 layout/generic/nsObjectFrame.cpp
--- a/layout/generic/nsObjectFrame.cpp	Sat Aug 01 12:17:20 2009 +0300
+++ b/layout/generic/nsObjectFrame.cpp	Sat Aug 01 16:50:35 2009 +0300
@@ -445,7 +445,8 @@ public:
 #ifdef XP_WIN
     return MatchPluginName("Shockwave Flash");
 #else
-    return PR_FALSE;
+    // return TRUE for windowless plugins
+    return (!(mPluginWindow && mPluginWindow->type != nsPluginWindowType_Drawable));
 #endif
   }
 
@@ -3890,6 +3891,7 @@ nsEventStatus nsPluginInstanceOwner::Pro
           switch (anEvent.message)
             {
             case NS_KEY_DOWN:
+            case NS_KEY_PRESS:
               event.type = XKeyPress;
               break;
             case NS_KEY_UP:
@@ -4209,7 +4211,7 @@ nsEventStatus nsPluginInstanceOwner::Pro
           event.state = gdkEvent->state;
           switch (anEvent.message)
             {
-            case NS_KEY_DOWN:
+            case NS_KEY_PRESS:
               event.type = XKeyPress;
               break;
             case NS_KEY_UP:
