#workaround for bug 503784
diff -r c7d11c7900f8 gfx/src/thebes/nsThebesDeviceContext.cpp
--- a/gfx/src/thebes/nsThebesDeviceContext.cpp	Mon Jul 13 08:57:23 2009 +0300
+++ b/gfx/src/thebes/nsThebesDeviceContext.cpp	Mon Jul 13 09:06:52 2009 +0300
@@ -353,7 +353,7 @@ NS_IMETHODIMP
 nsThebesDeviceContext::Observe(nsISupports* aSubject, const char* aTopic, const PRUnichar* aSomeData)
 {
     if (mFontCache && !nsCRT::strcmp(aTopic, "memory-pressure")) {
-        mFontCache->Compact();
+        mFontCache->Flush();
     }
     return NS_OK;
 }
