--- mozilla/gfx/thebes/src/gfxQtPlatform.cpp.orig	2008-08-10 18:06:50.000000000 +0300
+++ mozilla/gfx/thebes/src/gfxQtPlatform.cpp	2008-08-10 18:26:14.000000000 +0300
@@ -54,13 +54,15 @@
 
 #include "nsUnicharUtils.h"
 
 #include <fontconfig/fontconfig.h>
 
+#ifndef MOZ_NO_LCMS
 #include "nsMathUtils.h"
 
 #include "lcms.h"
+#endif
 
 #include <ft2build.h>
 #include FT_FREETYPE_H
 
 PRInt32 gfxQtPlatform::sDPI = -1;
--- mozilla/embedding/browser/qt/src/EmbedWindow.cpp.orig	2008-08-08 12:54:56.000000000 +0300
+++ mozilla/embedding/browser/qt/src/EmbedWindow.cpp	2008-08-10 19:36:54.000000000 +0300
@@ -157,12 +157,16 @@ NS_IMPL_ADDREF(EmbedWindow)
     NS_INTERFACE_MAP_BEGIN(EmbedWindow)
     NS_INTERFACE_MAP_ENTRY_AMBIGUOUS(nsISupports, nsIWebBrowserChrome)
     NS_INTERFACE_MAP_ENTRY(nsIWebBrowserChrome)
     NS_INTERFACE_MAP_ENTRY(nsIWebBrowserChromeFocus)
     NS_INTERFACE_MAP_ENTRY(nsIEmbeddingSiteWindow)
+#ifndef MOZ_NO_TOOLTIPS
     NS_INTERFACE_MAP_ENTRY(nsITooltipListener)
+#endif
+#ifndef MOZ_NO_CONTEXTMENU
     NS_INTERFACE_MAP_ENTRY(nsIContextMenuListener)
+#endif
     NS_INTERFACE_MAP_ENTRY(nsIInterfaceRequestor)
     NS_INTERFACE_MAP_END
 
 // nsIWebBrowserChrome
 
@@ -367,11 +371,11 @@ EmbedWindow::SetVisibility(PRBool aVisib
 
     return NS_OK;
 }
 
 // nsITooltipListener
-
+#ifndef MOZ_NO_TOOLTIPS
 NS_IMETHODIMP
 EmbedWindow::OnShowTooltip(PRInt32 aXCoords, PRInt32 aYCoords, const PRUnichar *aTipText)
 {
     QString tipText = QString::fromUtf16(aTipText);
 
@@ -408,12 +412,13 @@ EmbedWindow::OnHideTooltip(void)
 {
     if (tooltip)
         tooltip->hide();
     return NS_OK;
 }
+#endif
 
-
+#ifndef MOZ_NO_CONTEXTMENU
 NS_IMETHODIMP
 EmbedWindow::OnShowContextMenu(PRUint32 aContextFlags, nsIDOMEvent *aEvent, nsIDOMNode *aNode)
 {
 //     if (!aEvent->type == NS_CONTEXTMENU)
 //         return NS_OK;
@@ -445,10 +450,11 @@ EmbedWindow::OnShowContextMenu(PRUint32 
     }
 
     emit mOwner->showContextMenu(QCursor::pos(), url);
     return NS_OK;
 }
+#endif
 
 // nsIInterfaceRequestor
 
 NS_IMETHODIMP
 EmbedWindow::GetInterface(const nsIID &aIID, void** aInstancePtr)
--- mozilla/embedding/browser/qt/src/EmbedWindow.h.orig	2008-08-08 12:54:56.000000000 +0300
+++ mozilla/embedding/browser/qt/src/EmbedWindow.h	2008-08-10 19:36:20.000000000 +0300
@@ -42,12 +42,16 @@
 
 #include <nsStringGlue.h>
 #include <nsIWebBrowserChrome.h>
 #include <nsIWebBrowserChromeFocus.h>
 #include <nsIEmbeddingSiteWindow.h>
+#ifndef MOZ_NO_TOOLTIPS
 #include <nsITooltipListener.h>
+#endif
+#ifndef MOZ_NO_CONTEXTMENU
 #include <nsIContextMenuListener.h>
+#endif
 #include <nsISupports.h>
 #include <nsIWebBrowser.h>
 #include <nsIBaseWindow.h>
 #include <nsIInterfaceRequestor.h>
 #include <nsCOMPtr.h>
@@ -57,12 +61,16 @@ class QGeckoEmbed;
 class MozTipLabel;
 
 class EmbedWindow : public nsIWebBrowserChrome,
                     public nsIWebBrowserChromeFocus,
                     public nsIEmbeddingSiteWindow,
+#ifndef MOZ_NO_TOOLTIPS
                     public nsITooltipListener,
+#endif
+#ifndef MOZ_NO_CONTEXTMENU
                     public nsIContextMenuListener,
+#endif
                     public nsIInterfaceRequestor
 {
 public:
     EmbedWindow();
     ~EmbedWindow();
@@ -77,13 +85,17 @@ public:
 
     NS_DECL_NSIWEBBROWSERCHROMEFOCUS
 
     NS_DECL_NSIEMBEDDINGSITEWINDOW
 
+#ifndef MOZ_NO_TOOLTIPS
     NS_DECL_NSITOOLTIPLISTENER
+#endif
 
+#ifndef MOZ_NO_CONTEXTMENU
     NS_DECL_NSICONTEXTMENULISTENER
+#endif
 
     NS_DECL_NSIINTERFACEREQUESTOR
 
     nsString                 mTitle;
     nsString                 mJSStatus;
