--- mozilla/gfx/thebes/src/gfxQtPlatform.cpp.orig	2008-08-10 18:06:50.000000000 +0300
+++ mozilla/gfx/thebes/src/gfxQtPlatform.cpp	2008-08-10 18:26:14.000000000 +0300
@@ -54,13 +54,15 @@
 
 #include "nsUnicharUtils.h"
 
 #include <fontconfig/fontconfig.h>
 
+#ifndef MOZ_NO_LCMS
 #include "nsMathUtils.h"
 
 #include "lcms.h"
+#endif
 
 #include <ft2build.h>
 #include FT_FREETYPE_H
 
 PRInt32 gfxQtPlatform::sDPI = -1;
