Index: mozilla/widget/src/qt/nsClipboard.cpp
===================================================================
--- mozilla.orig/widget/src/qt/nsClipboard.cpp
+++ mozilla/widget/src/qt/nsClipboard.cpp
@@ -130,17 +130,22 @@ nsClipboard::SetNativeClipboardData( nsI
                     return NS_ERROR_FAILURE;
                 }
 
                 nsAutoString utf16string;
                 wideString->GetData(utf16string);
                 QString str = QString::fromUtf16(utf16string.get());
 
                 // Set the date to the clipboard
-                cb->setText(str, clipboardMode);
+                // (as text and as UTF-8 explicitly because Qt declares the text only as text/plain)
+                QMimeData* newData = new QMimeData();
+                newData->setText(str);
+                newData->setData("text/plain;charset=utf-8", str.toUtf8() );
+
+                cb->setMimeData(newData, clipboardMode);
                 break;
             }
 
             if ( !strcmp(flavorStr.get(), kNativeImageMime)
               || !strcmp(flavorStr.get(), kPNGImageMime)
               || !strcmp(flavorStr.get(), kJPEGImageMime)
               || !strcmp(flavorStr.get(), kGIFImageMime))
             {
