Index: mozilla/widget/src/qt/nsQtKeyUtils.cpp
===================================================================
--- mozilla.orig/widget/src/qt/nsQtKeyUtils.cpp
+++ mozilla/widget/src/qt/nsQtKeyUtils.cpp
@@ -49,16 +49,17 @@ struct nsKeyConverter
     int keysym; // Qt key code
 };
 
 static struct nsKeyConverter nsKeycodes[] =
 {
 //  { NS_VK_CANCEL,        Qt::Key_Cancel },
     { NS_VK_BACK,          Qt::Key_Backspace },
     { NS_VK_TAB,           Qt::Key_Tab },
+    { NS_VK_TAB,           Qt::Key_Backtab },
 //  { NS_VK_CLEAR,         Qt::Key_Clear },
     { NS_VK_RETURN,        Qt::Key_Return },
     { NS_VK_RETURN,        Qt::Key_Enter },
     { NS_VK_SHIFT,         Qt::Key_Shift },
     { NS_VK_CONTROL,       Qt::Key_Control },
     { NS_VK_ALT,           Qt::Key_Alt },
     { NS_VK_PAUSE,         Qt::Key_Pause },
     { NS_VK_CAPS_LOCK,     Qt::Key_CapsLock },
