--- mozilla/gfx/cairo/cairo/src/cairo-qpainter-surface.cpp.orig	2008-11-06 19:33:22.000000000 +0200
+++ mozilla/gfx/cairo/cairo/src/cairo-qpainter-surface.cpp	2008-11-09 05:39:47.000000000 +0200
@@ -1382,11 +1386,11 @@ _cairo_qpainter_surface_show_glyphs (voi
 	    qs->xlib_has_clipping = qs->has_clipping;
 	    qs->xlib_clip_bounds = qs->clip_bounds;
 	}
 
         return (cairo_int_status_t)
-               _cairo_surface_show_text_glyphs (qs->xlib_equiv, op, source, NULL, 0, glyphs, num_glyphs, NULL, 0, FALSE, scaled_font);
+               _cairo_surface_show_text_glyphs (qs->xlib_equiv, op, source, NULL, 0, glyphs, num_glyphs, NULL, 0, (cairo_text_cluster_flags_t)0, scaled_font);
     }
 #endif
 
     return CAIRO_INT_STATUS_UNSUPPORTED;
 }
