--- mozilla/widget/src/qt/nsWindow.cpp	2008-10-09 12:49:17.000000000 +0200
+++ mozilla/widget/src/qt/nsWindowChanged.cpp	2008-11-06 16:44:40.000000000 +0100
@@ -1811,6 +1811,12 @@
 
     Initialize(mDrawingArea);
 
+    // disable focus handling for secondary windows (problems with mouse selection and NS_ACTIVATE)
+    if (aParent != nsnull)
+    {
+        mDrawingArea->setFocusPolicy(Qt::NoFocus);
+    }
+
     LOG(("Create: nsWindow [%p] [%p]\n", (void *)this, (void *)mDrawingArea));
 
     // resize so that everything is set to the right dimensions
