--- mozilla/widget/src/qt/mozqwidget.cpp.orig	2009-01-16 14:03:39.000000000 +0100
+++ mozilla/widget/src/qt/mozqwidget.cpp	2009-01-16 14:04:38.000000000 +0100
@@ -12,6 +12,11 @@
 
 bool MozQWidget::event(QEvent *e)
 {
+    // always handle (delayed) delete requests triggered by
+    // calling deleteLater() on this widget:
+    if (e->type() == QEvent::DeferredDelete)
+        return QObject::event(e);
+
     nsEventStatus status = nsEventStatus_eIgnore;
     bool handled = true;
 
