# Bug 386854  Bustage build fix MOZ_XUL disabled
--- mozilla/content/base/src/nsGenericElement.cpp.orig	2007-05-28 15:43:19.000000000 +0300
+++ mozilla/content/base/src/nsGenericElement.cpp	2007-05-28 15:51:43.000000000 +0300
@@ -110,21 +110,23 @@
 #include "nsNodeInfoManager.h"
 #include "nsICategoryManager.h"
 #include "nsIDOMNSFeatureFactory.h"
 #include "nsIDOMDocumentType.h"
 #include "nsIDOMUserDataHandler.h"
 #include "nsEventDispatcher.h"
 #include "nsContentCreatorFunctions.h"
 #include "nsIFocusController.h"
 #include "nsIControllers.h"
 #include "nsXBLInsertionPoint.h"
+#ifdef MOZ_XUL
 #include "nsIXULDocument.h"
+#endif
 
 #include "nsCycleCollectionParticipant.h"
 #include "nsCCUncollectableMarker.h"
 
 #ifdef MOZ_SVG
 PRBool NS_SVG_TestFeature(const nsAString &fstr);
 #endif /* MOZ_SVG */
 
 #ifdef DEBUG_waterson
 
@@ -1714,36 +1716,40 @@ BindNodesInInsertPoints(nsXBLBinding* aB
                         nsIDocument* aDocument)
 {
   NS_PRECONDITION(aBinding && aInsertParent, "Missing arguments");
 
   nsresult rv;
   // These should be refcounted or otherwise protectable.
   nsInsertionPointList* inserts =
     aBinding->GetExistingInsertionPointsFor(aInsertParent);
   if (inserts) {
     PRBool allowScripts = aBinding->AllowScripts();
+#ifdef MOZ_XUL
     nsCOMPtr<nsIXULDocument> xulDoc = do_QueryInterface(aDocument);
+#endif
     PRUint32 i;
     for (i = 0; i < inserts->Length(); ++i) {
       nsCOMPtr<nsIContent> insertRoot =
         inserts->ElementAt(i)->GetDefaultContent();
       if (insertRoot) {
         PRUint32 j;
         for (j = 0; j < insertRoot->GetChildCount(); ++j) {
           nsCOMPtr<nsIContent> child = insertRoot->GetChildAt(j);
           rv = child->BindToTree(aDocument, aInsertParent,
                                  aBinding->GetBoundElement(), allowScripts);
           NS_ENSURE_SUCCESS(rv, rv);
 
+#ifdef MOZ_XUL
           if (xulDoc) {
             xulDoc->AddSubtreeToDocument(child);
           }
+#endif
         }
       }
     }
   }
 
   return NS_OK;
 }
 
 nsresult
 nsGenericElement::BindToTree(nsIDocument* aDocument, nsIContent* aParent,
