Index: mozilla/widget/src/gtk2/nsWindow.cpp
===================================================================
--- mozilla.orig/widget/src/gtk2/nsWindow.cpp
+++ mozilla/widget/src/gtk2/nsWindow.cpp
@@ -1759,19 +1759,21 @@
             
                     nsRefPtr<gfxContext> imgCtx = new gfxContext(img);
                     if (imgCtx) {
                         imgCtx->SetPattern(pattern);
                         imgCtx->SetOperator(gfxContext::OPERATOR_SOURCE);
                         imgCtx->Paint();
                     }
 
+#ifdef MOZ_XUL
                     UpdateTranslucentWindowAlphaInternal(nsRect(boundsRect.x, boundsRect.y,
                                                                 boundsRect.width, boundsRect.height),
                                                          img->Data(), img->Stride());
+#endif
                 }
             } else {
 #ifdef MOZ_ENABLE_GLITZ
                 ctx->PopGroupToSource();
                 ctx->Paint();
 #else // MOZ_ENABLE_GLITZ
                 if (bufferPixmapSurface) {
                     ctx->SetSource(bufferPixmapSurface);
