#Bug 384868  Wrong ifdef for rdf building
--- mozilla/toolkit/toolkit-tiers.mk.orig	2007-06-18 12:53:11.000000000 +0300
+++ mozilla/toolkit/toolkit-tiers.mk	2007-06-18 13:16:49.000000000 +0300
@@ -99,21 +99,21 @@ ifdef MOZ_PERMISSIONS
 tier_gecko_dirs += \
 		extensions/cookie \
 		extensions/permissions \
 		$(NULL)
 endif
 
 ifdef MOZ_STORAGE
 tier_gecko_dirs += storage
 endif
 
-ifdef MOZ_XUL
+ifdef MOZ_RDF
 tier_gecko_dirs += rdf
 endif
 
 ifdef MOZ_JSDEBUGGER
 tier_gecko_dirs += js/jsd
 endif
 
 tier_gecko_dirs	+= \
 		uriloader \
 		modules/libimg \
