--- mozilla/security/coreconf/UNIX.mk.orig	2007-06-24 17:00:03.000000000 +0300
+++ mozilla/security/coreconf/UNIX.mk	2007-06-24 17:44:16.000000000 +0300
@@ -37,20 +37,23 @@
 
 XP_DEFINE  += -DXP_UNIX
 LIB_SUFFIX  = a
 DLL_SUFFIX  = so
 AR          = ar cr $@
 LDOPTS     += -L$(SOURCE_LIB_DIR)
 
 ifdef BUILD_OPT
 	OPTIMIZER  += -O
 	DEFINES    += -UDEBUG -DNDEBUG
+ifndef BUILD_STRIP
+	OPTIMIZER  += -g
+endif
 else
 	OPTIMIZER  += -g
 	DEFINES    += -DDEBUG -UNDEBUG -DDEBUG_$(shell whoami)
 endif
 
 ifdef BUILD_TREE
 NSINSTALL_DIR  = $(BUILD_TREE)/nss
 NSINSTALL      = $(BUILD_TREE)/nss/nsinstall
 else
 NSINSTALL_DIR  = $(CORE_DEPTH)/coreconf/nsinstall
--- mozilla/security/coreconf/Linux.mk.orig	2007-06-24 17:42:36.000000000 +0300
+++ mozilla/security/coreconf/Linux.mk	2007-06-24 17:43:01.000000000 +0300
@@ -144,20 +144,24 @@ ifeq ($(OS_RELEASE),2.0)
 		MKSHLIB += -Wl,--version-script,$(MAPFILE)
 	endif
 	PROCESS_MAP_FILE = grep -v ';-' $< | \
          sed -e 's,;+,,' -e 's; DATA ;;' -e 's,;;,,' -e 's,;.*,;,' > $@
 endif
 
 ifdef BUILD_OPT
 	OPTIMIZER	= -O2
 endif
 
+ifndef BUILD_STRIP
+	OPTIMIZER	+= -g
+endif
+
 ifeq ($(USE_PTHREADS),1)
 OS_PTHREAD = -lpthread 
 endif
 
 OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) $(ARCHFLAG) -ansi -Wall -pipe -DLINUX -Dlinux -D_POSIX_SOURCE -D_BSD_SOURCE -DHAVE_STRERROR
 OS_LIBS			= $(OS_PTHREAD) -ldl -lc
 
 ifdef USE_PTHREADS
 	DEFINES		+= -D_REENTRANT
 endif
--- mozilla/security/manager/Makefile.in.orig	2007-06-24 17:05:59.000000000 +0300
+++ mozilla/security/manager/Makefile.in	2007-06-24 17:41:45.000000000 +0300
@@ -149,20 +149,23 @@ DEFAULT_GMAKE_FLAGS += NSPR_LIB_DIR=$(NS
 DEFAULT_GMAKE_FLAGS += MOZILLA_CLIENT=1
 DEFAULT_GMAKE_FLAGS += NO_MDUPDATE=1
 DEFAULT_GMAKE_FLAGS += NSS_ENABLE_ECC=1
 ABS_topsrcdir   := $(shell cd $(topsrcdir); pwd)
 ifneq ($(ABS_topsrcdir),$(MOZ_BUILD_ROOT))
 DEFAULT_GMAKE_FLAGS += BUILD_TREE=$(MOZ_BUILD_ROOT)
 endif
 ifndef MOZ_DEBUG
 DEFAULT_GMAKE_FLAGS += BUILD_OPT=1
 endif
+ifdef ENABLE_STRIP
+DEFAULT_GMAKE_FLAGS += BUILD_STRIP=1
+endif
 ifdef GNU_CC
 DEFAULT_GMAKE_FLAGS += NS_USE_GCC=1 NS_USE_NATIVE=
 else
 DEFAULT_GMAKE_FLAGS += NS_USE_GCC= NS_USE_NATIVE=1
 endif
 ifdef USE_N32
 # It is not really necessary to specify USE_PTHREADS=1.  USE_PTHREADS
 # merely adds _PTH to coreconf's OBJDIR name.
 DEFAULT_GMAKE_FLAGS += USE_N32=1 USE_PTHREADS=1
 endif
