Index: mozilla/content/html/document/src/nsHTMLDocument.cpp
===================================================================
--- mozilla.orig/content/html/document/src/nsHTMLDocument.cpp
+++ mozilla/content/html/document/src/nsHTMLDocument.cpp
@@ -3535,16 +3535,17 @@
     if ((tag == nsGkAtoms::embed  ||
          tag == nsGkAtoms::img    ||
          tag == nsGkAtoms::object ||
          tag == nsGkAtoms::applet) &&
         (!aForm || nsContentUtils::BelongsInForm(aForm, e))) {
       NS_ADDREF(*aResult = e);
     }
   }
+  }
 
   return NS_OK;
 }
 
 //----------------------------
 
 PRBool
 nsHTMLDocument::GetBodyContent()
