#386799-canvas-zero-crash
#
diff -r e0360a08d634 layout/generic/nsHTMLCanvasFrame.cpp
--- a/layout/generic/nsHTMLCanvasFrame.cpp	Fri Jul 06 11:03:59 2007 +0200
+++ b/layout/generic/nsHTMLCanvasFrame.cpp	Fri Jul 06 15:30:19 2007 +0200
@@ -188,6 +188,10 @@ nsHTMLCanvasFrame::PaintCanvas(nsIRender
   if (!canvas)
     return;
 
+  // anything to do?
+  if (inner.width == 0 || inner.height == 0)
+    return;
+
   nsSize canvasSize = GetCanvasSize();
   nsSize sizeAppUnits(PresContext()->DevPixelsToAppUnits(canvasSize.width),
                       PresContext()->DevPixelsToAppUnits(canvasSize.height));
diff -r e0360a08d634 widget/src/cocoa/nsNativeThemeCocoa.mm
--- a/widget/src/cocoa/nsNativeThemeCocoa.mm	Fri Jul 06 11:03:59 2007 +0200
+++ b/widget/src/cocoa/nsNativeThemeCocoa.mm	Fri Jul 06 15:30:19 2007 +0200
@@ -517,9 +517,13 @@ nsNativeThemeCocoa::DrawWidgetBackground
 
   double offsetX = 0.0, offsetY = 0.0;
   nsRefPtr<gfxASurface> thebesSurface = thebesCtx->CurrentSurface(&offsetX, &offsetY);
+  if (thebesSurface->CairoStatus() != 0) {
+    NS_WARNING("Got Cairo surface with nonzero error status");
+    return NS_ERROR_FAILURE;
+  }
+
   if (thebesSurface->GetType() != gfxASurface::SurfaceTypeQuartz) {
-    fprintf(stderr, "Expected surface of type Quartz, got %d\n",
-            thebesSurface->GetType());
+    NS_WARNING("Expected surface of type Quartz, got somthing else");
     return NS_ERROR_FAILURE;
   }
 
@@ -529,9 +533,9 @@ nsNativeThemeCocoa::DrawWidgetBackground
 
   //fprintf (stderr, "surface: %p cgContext: %p\n", quartzSurf, cgContext);
 
-  if (cgContext == nsnull ||
-      ((((unsigned long)cgContext) & 0xffff) == 0x3f3f)) {
-    fprintf (stderr, "********** Invalid CGContext!\n");
+  if (cgContext == nsnull) {
+    NS_WARNING("Invalid CGContext!");
+    return NS_ERROR_FAILURE;
   }
 
   // Eventually we can just do a GetCTM and restore it with SetCTM,
