Index: src/nsImageDocument.cpp
===================================================================
RCS file: /cvsroot/mozilla/content/html/document/src/nsImageDocument.cpp,v
retrieving revision 1.160
diff -p -u -1 -0 -r1.160 nsImageDocument.cpp
--- mozilla/content/html/document/src/nsImageDocument.cpp	1 May 2007 22:24:20 -0000	1.160
+++ mozilla/content/html/document/src/nsImageDocument.cpp	26 Jul 2007 18:39:29 -0000
@@ -593,47 +593,53 @@ nsImageDocument::CreateSyntheticDocument
 
   body->AppendChildTo(mImageContent, PR_FALSE);
   imageLoader->SetLoadingEnabled(PR_TRUE);
 
   return NS_OK;
 }
 
 nsresult
 nsImageDocument::CheckOverflowing(PRBool changeState)
 {
-  nsIPresShell *shell = GetPrimaryShell();
-  if (!shell) {
-    return NS_OK;
-  }
+  /* Create a scope so that the style context gets destroyed before we might
+   * call ClearStyleDataAndReflow.  Also, holding onto pointers to the
+   * presentatation through style resolution is potentially dangerous.
+   */
+  {
+    nsIPresShell *shell = GetPrimaryShell();
+    if (!shell) {
+      return NS_OK;
+    }
 
-  nsPresContext *context = shell->GetPresContext();
-  nsRect visibleArea = context->GetVisibleArea();
+    nsPresContext *context = shell->GetPresContext();
+    nsRect visibleArea = context->GetVisibleArea();
 
-  nsCOMPtr<nsIContent> content = do_QueryInterface(mBodyContent);
-  if (!content) {
-    NS_WARNING("no body on image document!");
-    return NS_ERROR_FAILURE;
-  }
+    nsCOMPtr<nsIContent> content = do_QueryInterface(mBodyContent);
+    if (!content) {
+      NS_WARNING("no body on image document!");
+      return NS_ERROR_FAILURE;
+    }
 
-  nsRefPtr<nsStyleContext> styleContext =
-    context->StyleSet()->ResolveStyleFor(content, nsnull);
+    nsRefPtr<nsStyleContext> styleContext =
+      context->StyleSet()->ResolveStyleFor(content, nsnull);
 
-  nsMargin m;
-  if (styleContext->GetStyleMargin()->GetMargin(m))
-    visibleArea.Deflate(m);
-  m = styleContext->GetStyleBorder()->GetBorder();
-  visibleArea.Deflate(m);
-  if (styleContext->GetStylePadding()->GetPadding(m))
+    nsMargin m;
+    if (styleContext->GetStyleMargin()->GetMargin(m))
+      visibleArea.Deflate(m);
+    m = styleContext->GetStyleBorder()->GetBorder();
     visibleArea.Deflate(m);
+    if (styleContext->GetStylePadding()->GetPadding(m))
+      visibleArea.Deflate(m);
 
-  mVisibleWidth = nsPresContext::AppUnitsToIntCSSPixels(visibleArea.width);
-  mVisibleHeight = nsPresContext::AppUnitsToIntCSSPixels(visibleArea.height);
+    mVisibleWidth = nsPresContext::AppUnitsToIntCSSPixels(visibleArea.width);
+    mVisibleHeight = nsPresContext::AppUnitsToIntCSSPixels(visibleArea.height);
+  }
 
   PRBool imageWasOverflowing = mImageIsOverflowing;
   mImageIsOverflowing =
     mImageWidth > mVisibleWidth || mImageHeight > mVisibleHeight;
   PRBool windowBecameBigEnough = imageWasOverflowing && !mImageIsOverflowing;
 
   if (changeState || mShouldResize || mFirstResize ||
       windowBecameBigEnough) {
     if (mImageIsOverflowing && (changeState || mShouldResize)) {
       ShrinkToFit();
