Index: editor/libeditor/base/nsEditor.cpp
===================================================================
RCS file: /cvsroot/mozilla/editor/libeditor/base/nsEditor.cpp,v
retrieving revision 1.496
diff -u -8 -p -r1.496 nsEditor.cpp
--- a/editor/libeditor/base/nsEditor.cpp	2 Jul 2007 14:01:27 -0000	1.496
+++ a/editor/libeditor/base/nsEditor.cpp	4 Jul 2007 17:29:17 -0000
@@ -957,18 +957,16 @@ nsEditor::BeginPlaceHolderTransaction(ns
 NS_IMETHODIMP 
 nsEditor::EndPlaceHolderTransaction()
 {
   NS_PRECONDITION(mPlaceHolderBatch > 0, "zero or negative placeholder batch count when ending batch!");
   if (mPlaceHolderBatch == 1)
   {
     nsCOMPtr<nsISelection>selection;
     nsresult rv = GetSelection(getter_AddRefs(selection));
-    if (NS_FAILED(rv))
-      return rv;
 
     nsCOMPtr<nsISelectionPrivate>selPrivate(do_QueryInterface(selection));
 
    // By making the assumption that no reflow happens during the calls
    // to EndUpdateViewBatch and ScrollSelectionIntoView, we are able to
    // allow the selection to cache a frame offset which is used by the
    // caret drawing code. We only enable this cache here; at other times,
    // we have no way to know whether reflow invalidates it
