Index: editor/libeditor/base/nsEditor.cpp
===================================================================
RCS file: /cvsroot/mozilla/editor/libeditor/base/nsEditor.cpp,v
retrieving revision 1.496
diff -u -8 -p -r1.496 nsEditor.cpp
--- a/editor/libeditor/base/nsEditor.cpp	2 Jul 2007 14:01:27 -0000	1.496
+++ a/editor/libeditor/base/nsEditor.cpp	4 Jul 2007 14:17:59 -0000
@@ -346,18 +346,23 @@ nsEditor::InstallEventListeners()
                                       sysGroup);
     NS_ASSERTION(NS_SUCCEEDED(rv),
                  "failed to register key listener in system group");
   }
 
   rv |= piTarget->AddEventListenerByIID(mMouseListenerP,
                                         NS_GET_IID(nsIDOMMouseListener));
 
-  rv |= piTarget->AddEventListenerByIID(mFocusListenerP,
-                                        NS_GET_IID(nsIDOMFocusListener));
+  if (elmP) {
+    // Focus event doesn't bubble so adding the listener to capturing phase.
+    // Make sure this works after bug 235441 gets fixed.
+    rv |= elmP->AddEventListenerByIID(mFocusListenerP,
+                                      NS_GET_IID(nsIDOMFocusListener),
+                                      NS_EVENT_FLAG_CAPTURE);
+  }
 
   rv |= piTarget->AddEventListenerByIID(mTextListenerP,
                                         NS_GET_IID(nsIDOMTextListener));
 
   rv |= piTarget->AddEventListenerByIID(mCompositionListenerP,
                                         NS_GET_IID(nsIDOMCompositionListener));
 
   rv |= piTarget->AddEventListenerByIID(mDragListenerP,
@@ -380,44 +385,44 @@ nsEditor::RemoveEventListeners()
   {
     return;
   }
 
   nsCOMPtr<nsPIDOMEventTarget> piTarget = GetPIDOMEventTarget();
 
   if (piTarget)
   {
-    // unregister the event listeners with the DOM event reveiver
-
+    // unregister the event listeners with the DOM event target
+    nsCOMPtr<nsIEventListenerManager> elmP;
+    piTarget->GetListenerManager(PR_TRUE, getter_AddRefs(elmP));
     if (mKeyListenerP)
     {
       nsCOMPtr<nsIDOMEventGroup> sysGroup;
       piTarget->GetSystemEventGroup(getter_AddRefs(sysGroup));
-      nsCOMPtr<nsIEventListenerManager> elmP;
-      piTarget->GetListenerManager(PR_TRUE, getter_AddRefs(elmP));
       if (sysGroup && elmP)
       {
         elmP->RemoveEventListenerByType(mKeyListenerP,
                                         NS_LITERAL_STRING("keypress"),
                                         NS_EVENT_FLAG_BUBBLE |
                                         NS_PRIV_EVENT_UNTRUSTED_PERMITTED,
                                         sysGroup);
       }
     }
 
     if (mMouseListenerP)
     {
       piTarget->RemoveEventListenerByIID(mMouseListenerP,
                                          NS_GET_IID(nsIDOMMouseListener));
     }
 
-    if (mFocusListenerP)
+    if (mFocusListenerP && elmP)
     {
-      piTarget->RemoveEventListenerByIID(mFocusListenerP,
-                                         NS_GET_IID(nsIDOMFocusListener));
+      elmP->RemoveEventListenerByIID(mFocusListenerP,
+                                     NS_GET_IID(nsIDOMFocusListener),
+                                     NS_EVENT_FLAG_CAPTURE);
     }
 
     if (mTextListenerP)
     {
       piTarget->RemoveEventListenerByIID(mTextListenerP,
                                          NS_GET_IID(nsIDOMTextListener));
     }
 
