Index: nsSocketTransportService2.cpp
===================================================================
RCS file: /cvsroot/mozilla/netwerk/base/src/nsSocketTransportService2.cpp,v
retrieving revision 1.25
diff -u -p -r1.25 nsSocketTransportService2.cpp
--- mozilla/netwerk/base/src/nsSocketTransportService2.cpp	30 May 2007 17:30:39 -0000	1.25
+++ mozilla/netwerk/base/src/nsSocketTransportService2.cpp	4 Jul 2007 04:55:10 -0000
@@ -365,6 +365,9 @@ nsSocketTransportService::Init()
     if (mInitialized)
         return NS_OK;
 
+    if (mShuttingDown)
+        return NS_ERROR_NOT_AVAILABLE;
+
     if (!mThreadEvent) {
         mThreadEvent = PR_NewPollableEvent();
         //
@@ -401,6 +404,9 @@ nsSocketTransportService::Shutdown()
     if (!mInitialized)
         return NS_OK;
 
+    if (mShuttingDown)
+        return NS_ERROR_FAILURE;
+
     {
         nsAutoLock lock(mLock);
 
