diff -Naur mozilla.old/gfx/src/gtk/Makefile.in mozilla/gfx/src/gtk/Makefile.in
--- mozilla.old/gfx/src/gtk/Makefile.in	2007-04-24 09:23:52.000000000 +0000
+++ mozilla/gfx/src/gtk/Makefile.in	2007-04-24 09:38:15.000000000 +0000
@@ -142,6 +142,11 @@
 
 include $(topsrcdir)/config/rules.mk
 
+ifdef GNU_CC
+CFLAGS   += -fno-unit-at-a-time
+CXXFLAGS += -fno-unit-at-a-time
+endif
+
 ifdef MOZ_ENABLE_XINERAMA
 GFX_XINERAMA_LIBS += $(MOZ_XINERAMA_LIBS)
 endif

