# Codesourcery gcc 3.4.4 compiler workarounds
Index: mozilla/embedding/browser/gtk/src/Makefile.in
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/Makefile.in
+++ mozilla/embedding/browser/gtk/src/Makefile.in
@@ -213,16 +213,21 @@
 CXXFLAGS        += $(MOZ_GTK2_CFLAGS)
 CFLAGS          += $(MOZ_GTK2_CFLAGS)
 ifdef MOZ_GNOMEVFS_CFLAGS
 CXXFLAGS        += $(MOZ_GNOMEVFS_CFLAGS)
 CFLAGS          += $(MOZ_GNOMEVFS_CFLAGS)
 endif
 DEFINES         += -D_IMPL_GTKMOZEMBED
 
+ifdef GNU_CC
+CFLAGS   += -fno-unit-at-a-time
+CXXFLAGS += -fno-unit-at-a-time
+endif
+
 
 MARSHAL_FILE = gtkmozembedmarshal
 MARSHAL_PREFIX = gtkmozembed
 
 $(MARSHAL_FILE).h: $(MARSHAL_FILE).list
 	glib-genmarshal --prefix=$(MARSHAL_PREFIX) $(srcdir)/$(MARSHAL_FILE).list --skip-source --header > $(MARSHAL_FILE).h
 
 $(MARSHAL_FILE).c: $(MARSHAL_FILE).list $(MARSHAL_FILE).h
Index: mozilla/extensions/ftv/src/Makefile.in
===================================================================
--- mozilla.orig/extensions/ftv/src/Makefile.in
+++ mozilla/extensions/ftv/src/Makefile.in
@@ -93,16 +93,21 @@
 FINAL_TARGET=$(DIST)/xpi-stage/$(XPI_NAME)/platform/$(OS_TARGET)_$(TARGET_XPCOM_ABI)
 endif
 endif
 
 #EXTRA_COMPONENTS = $(srcdir)/SimpleTest.js
 
 include $(topsrcdir)/config/rules.mk
 
+ifdef GNU_CC
+CFLAGS   += -fno-unit-at-a-time
+CXXFLAGS += -fno-unit-at-a-time
+endif
+
 FTV_CSS_SRC = $(srcdir)/ftv.css
 PREPROCESS_FTV_CSS = $(PERL) $(MOZILLA_DIR)/config/preprocessor.pl --marker=% $(DEFINES) $(ACDEFINES) $(FTV_CSS_SRC)
 
 $(DIST)/bin/res/ftv.css $(DESTDIR)$(mozappdir)/res/ftv.css: $(FTV_CSS_SRC) Makefile
 	if test ! -d $(@D); then rm -rf $(@D); $(NSINSTALL) -D $(@D); else true; fi
 	rm -f $@
 	$(PREPROCESS_FTV_CSS) > $@
 
Index: mozilla/extensions/spatialnavigation/src/Makefile.in
===================================================================
--- mozilla.orig/extensions/spatialnavigation/src/Makefile.in
+++ mozilla/extensions/spatialnavigation/src/Makefile.in
@@ -121,15 +121,20 @@
 ifdef MOZ_XUL_APP
 ifdef TARGET_XPCOM_ABI
 FINAL_TARGET=$(DIST)/xpi-stage/$(XPI_NAME)/platform/$(OS_TARGET)_$(TARGET_XPCOM_ABI)
 endif
 endif
 
 include $(topsrcdir)/config/rules.mk
 
+ifdef GNU_CC
+CFLAGS   += -fno-unit-at-a-time
+CXXFLAGS += -fno-unit-at-a-time
+endif
+
 # Copy the nsRect and nsFont source files here and make sure
 # they can be clobbered.
 export::
 	$(NSINSTALL) $(topsrcdir)/gfx/src/nsRect.cpp .
 	$(NSINSTALL) $(topsrcdir)/gfx/src/nsFont.cpp .
 
 GARBAGE += nsRect.cpp nsFont.cpp
Index: mozilla/extensions/spellcheck/src/Makefile.in
===================================================================
--- mozilla.orig/extensions/spellcheck/src/Makefile.in
+++ mozilla/extensions/spellcheck/src/Makefile.in
@@ -86,16 +86,21 @@
 EXTRA_DSO_LDOPTS = \
 		$(LIBS_DIR) \
 		$(MOZ_COMPONENT_LIBS) \
 		$(MOZ_UNICHARUTIL_LIBS) \
 		$(NULL)
 
 include $(topsrcdir)/config/rules.mk
 
+ifdef GNU_CC
+CFLAGS   += -fno-unit-at-a-time
+CXXFLAGS += -fno-unit-at-a-time
+endif
+
 LOCAL_INCLUDES += \
 	-I$(topsrcdir)/xpcom/io \
 	$(NULL)
 
 ifeq (cocoa,$(MOZ_WIDGET_TOOLKIT))
 LDFLAGS	+= \
 	-framework Cocoa \
 	-framework Foundation \
