#Bug 360789  White/brown/gray sporadic squares when refreshing a page
#
Index: mozilla/widget/src/gtk2/nsWindow.cpp
===================================================================
--- mozilla.orig/widget/src/gtk2/nsWindow.cpp
+++ mozilla/widget/src/gtk2/nsWindow.cpp
@@ -1238,19 +1238,16 @@
     // Update bounds on our child windows
     for (nsIWidget* kid = mFirstChild; kid; kid = kid->GetNextSibling()) {
         nsRect bounds;
         kid->GetBounds(bounds);
         bounds.x += aDx;
         bounds.y += aDy;
         static_cast<nsBaseWidget*>(kid)->SetBounds(bounds);
     }
-
-    // Process all updates so that everything is drawn.
-    gdk_window_process_all_updates();
     return NS_OK;
 }
 
 NS_IMETHODIMP
 nsWindow::ScrollWidgets(PRInt32 aDx,
                         PRInt32 aDy)
 {
     if (!mDrawingarea)
Index: mozilla/widget/src/windows/nsWindow.cpp
===================================================================
--- mozilla.orig/widget/src/windows/nsWindow.cpp
+++ mozilla/widget/src/windows/nsWindow.cpp
@@ -3284,17 +3284,16 @@
     trect.left = aClipRect->x;
     trect.top = aClipRect->y;
     trect.right = aClipRect->XMost();
     trect.bottom = aClipRect->YMost();
   }
 
   ::ScrollWindowEx(mWnd, aDx, aDy, NULL, (nsnull != aClipRect) ? &trect : NULL,
                    NULL, NULL, SW_INVALIDATE | SW_SCROLLCHILDREN);
-  ::UpdateWindow(mWnd);
   return NS_OK;
 }
 
 NS_IMETHODIMP nsWindow::ScrollWidgets(PRInt32 aDx, PRInt32 aDy)
 {
   // Scroll the entire contents of the window + change the offset of any child windows
   ::ScrollWindowEx(mWnd, aDx, aDy, NULL, NULL, NULL,
                    NULL, SW_INVALIDATE | SW_SCROLLCHILDREN);
