Index: mozilla/js/src/jsnum.c
===================================================================
--- mozilla.orig/js/src/jsnum.c
+++ mozilla/js/src/jsnum.c
@@ -110,31 +110,38 @@
     return js_NewNumberValue(cx, d, rval);
 }
 
 /* See ECMA 15.1.2.2. */
 static JSBool
 num_parseInt(JSContext *cx, JSObject *obj, uintN argc, jsval *argv, jsval *rval)
 {
     jsint radix;
+    jsint value;
     JSString *str;
     jsdouble d;
     const jschar *bp, *ep;
 
     if (argc > 1) {
         if (!js_ValueToECMAInt32(cx, argv[1], &radix))
             return JS_FALSE;
     } else {
         radix = 0;
     }
     if (radix != 0 && (radix < 2 || radix > 36)) {
         *rval = DOUBLE_TO_JSVAL(cx->runtime->jsNaN);
         return JS_TRUE;
     }
 
+    value = vp[2];
+    if (JSVAL_IS_INT(value)) {
+        JS_SET_RVAL(cx, vp, value);
+        return JS_TRUE;
+    }
+
     str = js_ValueToString(cx, argv[0]);
     if (!str)
         return JS_FALSE;
     /* XXXbe js_strtointeger shouldn't require NUL termination */
     bp = js_UndependString(cx, str);
     if (!bp)
         return JS_FALSE;
     if (!js_strtointeger(cx, bp, &ep, radix, &d))
