--- mozilla/js/src/Makefile.in.orig	2007-06-04 12:23:47.000000000 +0300
+++ mozilla/js/src/Makefile.in	2007-06-05 09:57:26.000000000 +0300
@@ -228,20 +228,27 @@ endif
 GARBAGE		+= jscpucfg.o jsautocfg.h jsautocfg.tmp jscpucfg
 
 ifneq (,$(CROSS_COMPILE)$(filter-out WINNT,$(OS_ARCH)))
 TARGETS		+= jscpucfg$(HOST_BIN_SUFFIX)
 endif
 
 ifdef JS_SAFE_ARENA
 DEFINES		+= -DJS_USE_SAFE_ARENA
 endif
 
+JS_USE_FDLIBM_MATH = 1
+
+ifdef JS_USE_FDLIBM_MATH
+DEFINES		+= -DJS_USE_FDLIBM_MATH=1
+LDFLAGS		+= fdlibm/libfdm.a
+endif
+
 ifdef JS_THREADSAFE
 DEFINES		+= -DJS_THREADSAFE
 endif
 
 ifdef JS_NO_THIN_LOCKS
 DEFINES		+= -DJS_USE_ONLY_NSPR_LOCKS
 endif
 
 ifdef JS_VERSION
 DEFINES		+= -DJS_VERSION=$(JS_VERSION)
