#https://bugzilla.mozilla.org/show_bug.cgi?id=532621
diff -r 0656a43e60de configure.in
--- a/configure.in	Tue Aug 04 21:15:44 2009 +0300
+++ b/configure.in	Tue Aug 04 23:44:32 2009 +0300
@@ -5221,6 +5221,12 @@ then
             fi
             MOZ_ENABLE_GNOMEVFS=
         ])
+    else
+        if test `echo "$MOZ_EXTENSIONS_DEFAULT" | grep -c gnomevfs` -ne 0; then
+            PKG_CHECK_MODULES(MOZ_GNOMEVFS, gnome-vfs-2.0 >= $GNOMEVFS_VERSION gnome-vfs-module-2.0 >= $GNOMEVFS_VERSION,[
+              MOZ_GNOMEVFS_LIBS=`echo $MOZ_GNOMEVFS_LIBS | sed 's/-llinc\>//'`
+            ])
+        fi
     fi
 
     AC_SUBST(MOZ_ENABLE_GNOMEVFS)
@@ -5788,7 +5794,7 @@ MOZ_ARG_ENABLE_STRING(extensions,
 done],
     MOZ_EXTENSIONS="$MOZ_EXTENSIONS_DEFAULT")
 
-if test -z "$MOZ_ENABLE_GNOMEVFS" && test `echo "$MOZ_EXTENSIONS" | grep -c gnomevfs` -ne 0; then
+if test -z "$MOZ_ENABLE_GNOMEVFS" && test -z "$MOZ_GNOMEVFS_LIBS" && test `echo "$MOZ_EXTENSIONS" | grep -c gnomevfs` -ne 0; then
     # Suppress warning on non-X11 platforms
     if test -n "$MOZ_X11"; then
         AC_MSG_WARN([Cannot build gnomevfs without required libraries. Removing gnomevfs from MOZ_EXTENSIONS.])
