# Broken linkage with PERF Metrics tools in XULRUNNER Configuration
Index: toolkit/library/libxul-config.mk
===================================================================
--- toolkit/library/libxul-config.mk.orig
+++ toolkit/library/libxul-config.mk
@@ -125,16 +125,20 @@
 ifdef MOZ_XPFE_COMPONENTS
 DEFINES += -DMOZ_XPFE_COMPONENTS
 COMPONENT_LIBS += \
 	mozfind \
 	appcomps \
 	$(NULL)
 endif
 
+ifdef MOZ_PERF_METRICS
+EXTRA_DSO_LIBS  += mozutil_s
+endif
+
 ifdef MOZ_XPINSTALL
 DEFINES += -DMOZ_XPINSTALL
 COMPONENT_LIBS += \
 	xpinstall \
 	$(NULL)
 endif
 
 ifdef MOZ_JSDEBUGGER
