#Mozilla hangs loading this page (4Mb in size)
Index: layout/generic/nsBRFrame.cpp
===================================================================
--- layout/generic/nsBRFrame.cpp.orig
+++ layout/generic/nsBRFrame.cpp
@@ -170,17 +170,17 @@
 
     // Return our reflow status
     PRUint32 breakType = aReflowState.mStyleDisplay->mBreakType;
     if (NS_STYLE_CLEAR_NONE == breakType) {
       breakType = NS_STYLE_CLEAR_LINE;
     }
 
     aStatus = NS_INLINE_BREAK | NS_INLINE_BREAK_AFTER |
-      NS_INLINE_MAKE_BREAK_TYPE(breakType);
+      NS_INLINE_MAKE_BREAK_TYPE(breakType) | NS_FRAME_COMPLETE;
     ll->SetLineEndsInBR(PR_TRUE);
   }
   else {
     aStatus = NS_FRAME_COMPLETE;
   }
 
   NS_FRAME_SET_TRUNCATION(aStatus, aReflowState, aMetrics);
   return NS_OK;
