#339358
#Shrink nsStyleSVG slightly
Index: layout/style/nsStyleStruct.h
===================================================================
--- layout/style/nsStyleStruct.h.orig
+++ layout/style/nsStyleStruct.h
@@ -1286,26 +1286,26 @@
   nsStyleCoord     mStrokeDashoffset; // [inherited]
   nsStyleCoord     mStrokeWidth;      // [inherited]
 
   float            mFillOpacity;      // [inherited]
   float            mStrokeMiterlimit; // [inherited]
   float            mStrokeOpacity;    // [inherited]
 
   PRUint32         mStrokeDasharrayLength;
-  PRUint8          mClipRule;         // [inherited]
+  PRUint8          mClipRule:1;       // [inherited]
   PRUint8          mColorInterpolation; // [inherited] see nsStyleConsts.h
   PRUint8          mColorInterpolationFilters; // [inherited] see nsStyleConsts.h
-  PRUint8          mFillRule;         // [inherited] see nsStyleConsts.h
-  PRUint8          mPointerEvents;    // [inherited] see nsStyleConsts.h
-  PRUint8          mShapeRendering;   // [inherited] see nsStyleConsts.h
-  PRUint8          mStrokeLinecap;    // [inherited] see nsStyleConsts.h
-  PRUint8          mStrokeLinejoin;   // [inherited] see nsStyleConsts.h
-  PRUint8          mTextAnchor;       // [inherited] see nsStyleConsts.h
-  PRUint8          mTextRendering;    // [inherited] see nsStyleConsts.h
+  PRUint8          mFillRule:1;       // [inherited] see nsStyleConsts.h
+  PRUint8          mPointerEvents:4;  // [inherited] see nsStyleConsts.h
+  PRUint8          mShapeRendering:2; // [inherited] see nsStyleConsts.h
+  PRUint8          mStrokeLinecap:2;  // [inherited] see nsStyleConsts.h
+  PRUint8          mStrokeLinejoin:2; // [inherited] see nsStyleConsts.h
+  PRUint8          mTextAnchor:2;     // [inherited] see nsStyleConsts.h
+  PRUint8          mTextRendering:2;  // [inherited] see nsStyleConsts.h
 };
 
 struct nsStyleSVGReset : public nsStyleStruct {
   nsStyleSVGReset();
   nsStyleSVGReset(const nsStyleSVGReset& aSource);
   ~nsStyleSVGReset();
 
   void* operator new(size_t sz, nsPresContext* aContext) CPP_THROW_NEW {
