Index: mozilla/uriloader/exthandler/nsExternalHelperAppService.cpp
===================================================================
--- mozilla.orig/uriloader/exthandler/nsExternalHelperAppService.cpp
+++ mozilla/uriloader/exthandler/nsExternalHelperAppService.cpp
@@ -2371,17 +2371,18 @@
   if (mFinalFileDestination && !mStopRequestIssued)
   {
     nsCOMPtr<nsIFile> movedFile;
     mFinalFileDestination->Clone(getter_AddRefs(movedFile));
     if (movedFile) {
       // Get the old leaf name and append .part to it
       nsAutoString name;
       mFinalFileDestination->GetLeafName(name);
-      name.AppendLiteral(".part");
+      // we have no need of this appending
+      //name.AppendLiteral(".part");
       movedFile->SetLeafName(name);
 
       nsCOMPtr<nsIFile> dir;
       movedFile->GetParent(getter_AddRefs(dir));
 
       mOutStream->Close();
 
       rv = mTempFile->MoveTo(dir, name);
