#Fixed window title problem with XML pretty-print.
#
diff -ruN -p -u10 mozilla.orig/content/xslt/src/xslt/txMozillaXMLOutput.cpp mozilla/content/xslt/src/xslt/txMozillaXMLOutput.cpp
--- mozilla.orig/content/xslt/src/xslt/txMozillaXMLOutput.cpp	2007-02-21 11:22:14.000000000 +0200
+++ mozilla/content/xslt/src/xslt/txMozillaXMLOutput.cpp	2007-02-21 17:50:38.000000000 +0200
@@ -784,22 +784,21 @@ txMozillaXMLOutput::endHTMLElement(nsICo
 
         mCurrentNode = mCurrentNodeStack.SafeObjectAt(last);
         mCurrentNodeStack.RemoveObjectAt(last);
         mTableState = NS_STATIC_CAST(TableState,
                                      NS_PTR_TO_INT32(mTableStateStack.pop()));
 
         return NS_OK;
     }
 
     // Set document title
-    else if (mCreatingNewDocument &&
-             atom == txHTMLAtoms::title && !mHaveTitleElement) {
+    else if (atom == txHTMLAtoms::title && !mHaveTitleElement) {
         // The first title wins
         mHaveTitleElement = PR_TRUE;
         nsCOMPtr<nsIDOMNSDocument> domDoc = do_QueryInterface(mDocument);
 
         nsAutoString title;
         nsContentUtils::GetNodeTextContent(aElement, PR_TRUE, title);
 
         if (domDoc) {
             title.CompressWhitespace();
             domDoc->SetTitle(title);
