Index: mozilla/gfx/cairo/cairo/src/cairo-xlib-surface.c
===================================================================
--- mozilla.orig/gfx/cairo/cairo/src/cairo-xlib-surface.c
+++ mozilla/gfx/cairo/cairo/src/cairo-xlib-surface.c
@@ -1139,18 +1139,16 @@ typedef enum {
  */
 static composite_operation_t
 _categorize_composite_operation (cairo_xlib_surface_t *dst,
 				 cairo_operator_t      op,
 				 cairo_pattern_t      *src_pattern,
 				 cairo_bool_t	       have_mask)
 
 {
-    if (!dst->buggy_repeat)
-	return DO_RENDER;
 
     if (src_pattern->type == CAIRO_PATTERN_TYPE_SURFACE)
     {
 	cairo_surface_pattern_t *surface_pattern = (cairo_surface_pattern_t *)src_pattern;
 
 	if (_cairo_matrix_is_integer_translation (&src_pattern->matrix, NULL, NULL) &&
 	    src_pattern->extend == CAIRO_EXTEND_REPEAT)
 	{
@@ -1213,19 +1211,16 @@ _recategorize_composite_operation (cairo
 	is_integer_translation &&
 	src_attr->extend == CAIRO_EXTEND_NONE &&
 	!needs_alpha_composite &&
 	_surfaces_compatible(src, dst))
     {
 	return DO_XCOPYAREA;
     }
 
-    if (!dst->buggy_repeat)
-	return DO_RENDER;
-
     if (is_integer_translation &&
 	src_attr->extend == CAIRO_EXTEND_REPEAT &&
 	(src->width != 1 || src->height != 1))
     {
 	if (!have_mask &&
 	    !needs_alpha_composite &&
 	    _surfaces_compatible (dst, src))
 	{
Index: mozilla/gfx/cairo/cairo/src/Makefile.in
===================================================================
--- mozilla.orig/gfx/cairo/cairo/src/Makefile.in
+++ mozilla/gfx/cairo/cairo/src/Makefile.in
@@ -175,13 +175,13 @@ endif
 LOCAL_INCLUDES  += -I$(srcdir)
 
 FORCE_STATIC_LIB = 1
 # This library is used by other shared libs in a static build
 FORCE_USE_PIC = 1
 
 include $(topsrcdir)/config/rules.mk
 
-DEFINES += -DPACKAGE_VERSION="\"moz\"" -DPACKAGE_BUGREPORT="\"http://bugzilla.mozilla.org/\""
+DEFINES += -DPACKAGE_VERSION="\"moz\"" -DPACKAGE_BUGREPORT="\"http://bugzilla.mozilla.org/\"" -DCAIRO_HAS_XLIB_SURFACE=1 -DCAIRO_HAS_XLIB_XRENDER_SURFACE=1
 
 ifeq ($(MOZ_WIDGET_TOOLKIT),windows)
 DEFINES += -DCAIRO_WIN32_STATIC_BUILD
 endif
