Index: mozilla/config/rules.mk
===================================================================
--- mozilla.orig/config/rules.mk
+++ mozilla/config/rules.mk
@@ -1737,17 +1737,17 @@ ifndef NO_DIST_INSTALL
 	  $(PERL) -I$(MOZILLA_DIR)/config $(MOZILLA_DIR)/config/make-jars.pl \
 	    -d $(MAKE_JARS_TARGET)/chrome -j $(FINAL_TARGET)/chrome \
 	    $(MAKE_JARS_FLAGS) -- "$(XULPPFLAGS) $(DEFINES) $(ACDEFINES)"; \
 	fi
 endif
 
 install:: $(CHROME_DEPS)
 ifneq (,$(filter flat symlink,$(MOZ_CHROME_FILE_FORMAT)))
-	$(error Flat chrome is for debugging only, and should not be used with the install target.)
+	$(warning Flat chrome is for debugging only, and should not be used with the install target.)
 endif
 ifndef NO_INSTALL
 	@$(EXIT_ON_ERROR) \
 	if test -f $(JAR_MANIFEST); then \
 	  if test ! -d $(DESTDIR)$(mozappdir)/chrome; then $(NSINSTALL) -D $(DESTDIR)$(mozappdir)/chrome; fi; \
 	  if test ! -d $(MAKE_JARS_TARGET)/chrome; then $(NSINSTALL) -D $(MAKE_JARS_TARGET)/chrome; fi; \
 	  $(PYTHON) $(MOZILLA_DIR)/config/Preprocessor.py $(XULPPFLAGS) $(DEFINES) $(ACDEFINES) \
 	    $(JAR_MANIFEST) | \
