--- mozilla/dom/src/base/nsGlobalWindow.cpp.orig	2007-04-26 14:12:20.000000000 +0300
+++ mozilla/dom/src/base/nsGlobalWindow.cpp	2007-04-26 14:16:39.581299750 +0300
@@ -5907,21 +5907,21 @@ nsGlobalWindow::GetSessionStorage(nsIDOM
   }
 
   return docShell->GetSessionStorageForURI(codebase, aSessionStorage);
 }
 
 NS_IMETHODIMP
 nsGlobalWindow::GetGlobalStorage(nsIDOMStorageList ** aGlobalStorage)
 {
   NS_ENSURE_ARG_POINTER(aGlobalStorage);
 
-#ifdef MOZ_STORAGE
+#if defined(MOZ_STORAGE) || defined(MOZ_MICROB)
   if (!gGlobalStorageList) {
     nsresult rv = NS_NewDOMStorageList(getter_AddRefs(gGlobalStorageList));
     NS_ENSURE_SUCCESS(rv, rv);
   }
 
   *aGlobalStorage = gGlobalStorageList;
   NS_IF_ADDREF(*aGlobalStorage);
 
   return NS_OK;
 #else
