# Use special glyph instead of default rectangular glyph
--- mozilla/gfx/src/gtk/nsFontMetricsXft.cpp.orig	2007-04-26 13:56:41.000000000 +0300
+++ mozilla/gfx/src/gtk/nsFontMetricsXft.cpp	2007-04-26 14:05:58.000000000 +0300
@@ -1331,6 +1331,29 @@
     return NS_OK;
 }
 
+#ifndef useUnknownRectGlyph
+
+nsresult
+nsFontMetricsXft::DrawUnknownGlyph(FcChar32   aChar,
+                                   nscoord    aX,
+                                   nscoord    aY,
+                                   XftColor  *aColor,
+                                   XftDraw   *aDraw)
+{
+    int ndigit = (IS_NON_BMP(aChar)) ? 3 : 2;
+    // horizontal lines
+    XftDrawRect(aDraw, aColor,
+                aX, aY - mMiniFontHeight * 2 + mMiniFontPadding,
+                mMiniFontWidth * ndigit + mMiniFontPadding * ndigit, mMiniFontPadding);
+    XftDrawRect(aDraw, aColor,
+                aX, aY - mMiniFontPadding * 2,
+                mMiniFontWidth * ndigit + mMiniFontPadding * ndigit, mMiniFontPadding);
+
+    return NS_OK;
+}
+
+#else
+
 nsresult
 nsFontMetricsXft::DrawUnknownGlyph(FcChar32   aChar,
                                    nscoord    aX,
@@ -1422,6 +1445,7 @@
 
     return NS_OK;
 }
+#endif //useUnknownRectGlyph
 
 nsresult
 nsFontMetricsXft::EnumerateXftGlyphs(const FcChar32 *aString, PRUint32 aLen,
