--- mozilla/content/events/src/nsEventStateManager.cpp.orig	2007-09-24 19:22:48.000000000 +0300
+++ mozilla/content/events/src/nsEventStateManager.cpp	2007-09-24 19:24:27.000000000 +0300
@@ -2277,21 +2277,21 @@ nsEventStateManager::PostHandleEvent(nsP
               //Shift focus forward or back depending on shift key
               ShiftFocus(!((nsInputEvent*)aEvent)->isShift);
             } else {
               ShiftFocusByDoc(!((nsInputEvent*)aEvent)->isShift);
             }
             *aStatus = nsEventStatus_eConsumeNoDefault;
             break;
 
           case NS_VK_F6:
             //Shift focus forward or back depending on shift key
-            ShiftFocusByDoc(!((nsInputEvent*)aEvent)->isShift);
+            //ShiftFocusByDoc(!((nsInputEvent*)aEvent)->isShift); // Disabled F6 as focus move key
             *aStatus = nsEventStatus_eConsumeNoDefault;
             break;
 
 //the problem is that viewer does not have xul so we cannot completely eliminate these
 #if NON_KEYBINDING
           case NS_VK_PAGE_DOWN:
           case NS_VK_PAGE_UP:
             if (!mCurrentFocus) {
               nsIScrollableView* sv = nsLayoutUtils::GetNearestScrollingView(aView, nsLayoutUtils::eVertical);
               if (sv) {
