diff -ruN -p -u10 mozilla.orig/js/src/jsobj.c mozilla/js/src/jsobj.c
--- mozilla.orig/js/src/jsobj.c	2007-10-17 11:53:07.000000000 +0300
+++ mozilla/js/src/jsobj.c	2007-10-17 11:54:52.000000000 +0300
@@ -3858,20 +3858,21 @@ js_DeleteProperty(JSContext *cx, JSObjec
             OBJ_DROP_PROPERTY(cx, proto, prop);
             if (*rval == JSVAL_FALSE)
                 return JS_TRUE;
         }
 
         /*
          * If no property, or the property comes unshared or impermanent from
          * a prototype, call the class's delProperty hook, passing rval as the
          * result parameter.
          */
+        if(OBJ_GET_CLASS(cx, obj)->delProperty != 0x1)
         return OBJ_GET_CLASS(cx, obj)->delProperty(cx, obj, ID_TO_VALUE(id),
                                                    rval);
     }
 
     sprop = (JSScopeProperty *)prop;
     if (sprop->attrs & JSPROP_PERMANENT) {
         OBJ_DROP_PROPERTY(cx, obj, prop);
         *rval = JSVAL_FALSE;
         return JS_TRUE;
     }
