Index: mozilla/gfx/public/nsIImage.h
===================================================================
--- mozilla.orig/gfx/public/nsIImage.h
+++ mozilla/gfx/public/nsIImage.h
@@ -36,19 +36,19 @@
  * ***** END LICENSE BLOCK ***** */
 
 #ifndef nsIImage_h___
 #define nsIImage_h___
 
 #include "nsISupports.h"
 #include "nsIRenderingContext.h"
 #include "nsRect.h"
+#ifdef MOZ_CAIRO_GFX
 #include "gfxRect.h"
 
-#ifdef MOZ_CAIRO_GFX
 class gfxASurface;
 #endif
 
 class nsIDeviceContext;
 
 struct nsColorMap
 {
   //I lifted this from the image lib. The difference is that
Index: mozilla/configure.in
===================================================================
--- mozilla.orig/configure.in
+++ mozilla/configure.in
@@ -4428,16 +4428,17 @@
                             WinCE - windows
                             * - cairo-gtk2],
     [ _DEFAULT_TOOLKIT=$enableval ],
     [ _DEFAULT_TOOLKIT=$_PLATFORM_DEFAULT_TOOLKIT])
 
     if test "$_DEFAULT_TOOLKIT" = "photon" \
         -o "$_DEFAULT_TOOLKIT" = "cairo-windows" \
         -o "$_DEFAULT_TOOLKIT" = "cairo-gtk2" \
+        -o "$_DEFAULT_TOOLKIT" = "gtk2" \
         -o "$_DEFAULT_TOOLKIT" = "cairo-beos" \
         -o "$_DEFAULT_TOOLKIT" = "cairo-os2" \
         -o "$_DEFAULT_TOOLKIT" = "cairo-mac" \
         -o "$_DEFAULT_TOOLKIT" = "cairo-cocoa"
     then
         dnl nglayout only supports building with one toolkit,
         dnl so ignore everything after the first comma (",").
         MOZ_WIDGET_TOOLKIT=`echo "$_DEFAULT_TOOLKIT" | sed -e "s/,.*$//"`
@@ -4472,16 +4473,25 @@
     MOZ_GFX_TOOLKIT=cairo
     MOZ_ENABLE_CAIRO_GFX=1
     MOZ_ENABLE_GTK2=1
     MOZ_ENABLE_XREMOTE=1
     TK_CFLAGS='$(MOZ_GTK2_CFLAGS) $(MOZ_CAIRO_CFLAGS)'
     TK_LIBS='$(MOZ_GTK2_LIBS) $(MOZ_CAIRO_LIBS)'
     AC_DEFINE(MOZ_WIDGET_GTK2)
     ;;
+
+gtk2)
+    MOZ_ENABLE_GTK2=1
+    MOZ_ENABLE_COREXFONTS=${MOZ_ENABLE_COREXFONTS-}
+    TK_CFLAGS='$(MOZ_GTK2_CFLAGS)'
+    TK_LIBS='$(MOZ_GTK2_LIBS)'
+    AC_DEFINE(MOZ_WIDGET_GTK2)
+    ;;
+
 cairo-beos)
     MOZ_WIDGET_TOOLKIT=beos
     MOZ_GFX_TOOLKIT=cairo
     MOZ_ENABLE_CAIRO_GFX=1
     TK_CFLAGS='$(MOZ_CAIRO_CFLAGS)'
     TK_LIBS='$(MOZ_CAIRO_LIBS)'
     ;;
 
