Index: mozilla/dom/base/nsJSEnvironment.cpp
===================================================================
--- mozilla.orig/dom/base/nsJSEnvironment.cpp
+++ mozilla/dom/base/nsJSEnvironment.cpp
@@ -210,23 +210,28 @@ static JSGCCallback gOldJSGCCallback;
 
 static PRBool sIsInitialized;
 static PRBool sDidShutdown;
 
 static PRInt32 sContextCount;
 
 static PRTime sMaxScriptRunTime;
 static PRTime sMaxChromeScriptRunTime;
+static PRBool sWakeUpNativeLoopOnJS;
 
 static nsIScriptSecurityManager *sSecurityManager;
 
 static nsICollation *gCollation;
 
 static nsIUnicodeDecoder *gDecoder;
 
+#include "nsWidgetsCID.h"
+#include "nsIAppShell.h"
+static NS_DEFINE_CID(kAppShellCID, NS_APPSHELL_CID);
+
 // nsUserActivityObserver observes user-interaction-active and
 // user-interaction-inactive notifications. It counts the number of
 // notifications and if the number is bigger than NS_CC_SOFT_LIMIT_ACTIVE
 // (in case the current notification is user-interaction-active) or
 // NS_CC_SOFT_LIMIT_INACTIVE (current notification is user-interaction-inactive)
 // MaybeCC is called with aHigherParameter set to PR_TRUE, otherwise PR_FALSE.
 //
 // When moving from active state to inactive, nsJSContext::IntervalCC() is
@@ -898,16 +903,30 @@ nsJSContext::DOMOperationCallback(JSCont
   // Get the native context
   nsJSContext *ctx = static_cast<nsJSContext *>(::JS_GetContextPrivate(cx));
 
   if (!ctx) {
     // Can happen; see bug 355811
     return JS_TRUE;
   }
 
+  nsCOMPtr<nsIAppShell> appShell = do_GetService(kAppShellCID);
+  if (appShell && sWakeUpNativeLoopOnJS) {
+    nsCOMPtr<nsPIDOMWindow> win(do_QueryInterface(ctx->GetGlobalObject()));
+    // Check for existing docshell before possible window close event
+    PRBool docshell = (win && win->GetDocShell());
+
+    ctx->SetScriptsEnabled(PR_FALSE, PR_FALSE);
+    appShell->WakeupNative(PR_FALSE);
+    ctx->SetScriptsEnabled(PR_TRUE, PR_FALSE);
+
+    if (docshell && !win->GetDocShell())
+      return JS_FALSE;
+  }
+
   // XXX Save the operation callback time so we can restore it after the GC,
   // because GCing can cause JS to run on our context, causing our
   // ScriptEvaluated to be called, and clearing our operation callback time.
   // See bug 302333.
   PRTime callbackTime = ctx->mOperationCallbackTime;
   PRTime modalStateTime = ctx->mModalStateTime;
 
   MaybeGC(cx);
@@ -3853,16 +3872,23 @@ static int
 ReportAllJSExceptionsPrefChangedCallback(const char* aPrefName, void* aClosure)
 {
   PRBool reportAll = nsContentUtils::GetBoolPref(aPrefName, PR_FALSE);
   nsContentUtils::XPConnect()->SetReportAllJSExceptions(reportAll);
   return 0;
 }
 
 static int
+WakeUpNativeLoopOnJSPrefChangedCallback(const char* aPrefName, void* aClosure)
+{
+  sWakeUpNativeLoopOnJS = nsContentUtils::GetBoolPref(aPrefName, PR_FALSE);
+  return 0;
+}
+
+static int
 SetMemoryHighWaterMarkPrefChangedCallback(const char* aPrefName, void* aClosure)
 {
   PRInt32 highwatermark = nsContentUtils::GetIntPref(aPrefName, 32);
 
   if (highwatermark >= 32) {
     // There are two options of memory usage in tracemonkey. One is
     // to use malloc() and the other is to use memory for GC. (E.g.
     // js_NewGCThing()/RefillDoubleFreeList()).
@@ -3967,16 +3993,21 @@ nsJSRuntime::Init()
                                       nsnull);
 
   nsContentUtils::RegisterPrefCallback("dom.report_all_js_exceptions",
                                        ReportAllJSExceptionsPrefChangedCallback,
                                        nsnull);
   ReportAllJSExceptionsPrefChangedCallback("dom.report_all_js_exceptions",
                                            nsnull);
 
+  nsContentUtils::RegisterPrefCallback("dom.wakeup_native_on_js",
+                                       WakeUpNativeLoopOnJSPrefChangedCallback,
+                                       nsnull);
+  WakeUpNativeLoopOnJSPrefChangedCallback("dom.wakeup_native_on_js", nsnull);
+
   nsContentUtils::RegisterPrefCallback("javascript.options.mem.high_water_mark",
                                        SetMemoryHighWaterMarkPrefChangedCallback,
                                        nsnull);
   SetMemoryHighWaterMarkPrefChangedCallback("javascript.options.mem.high_water_mark",
                                             nsnull);
 
   nsContentUtils::RegisterPrefCallback("javascript.options.mem.gc_frequency",
                                        SetMemoryGCFrequencyPrefChangedCallback,
Index: mozilla/widget/public/nsIAppShell.idl
===================================================================
--- mozilla.orig/widget/public/nsIAppShell.idl
+++ mozilla/widget/public/nsIAppShell.idl
@@ -93,12 +93,17 @@ interface nsIAppShell : nsISupports
   /**
    * Resumes the use of additional platform-specific methods to run Gecko
    * events on the main application thread.  Calls to suspendNative() and
    * resumeNative() may be nested.  On some platforms this will be a no-op.
    */
   void resumeNative();
 
   /**
+   * Will call DoNextNativeEvent for proceeding native event loop
+   */
+  void wakeupNative(in boolean mayWait);
+
+  /**
    * The current event loop nesting level.
    */
   readonly attribute unsigned long eventloopNestingLevel;
 };
Index: mozilla/widget/src/xpwidgets/nsBaseAppShell.cpp
===================================================================
--- mozilla.orig/widget/src/xpwidgets/nsBaseAppShell.cpp
+++ mozilla/widget/src/xpwidgets/nsBaseAppShell.cpp
@@ -205,16 +205,22 @@ NS_IMETHODIMP
 nsBaseAppShell::ResumeNative()
 {
   --mSuspendNativeCount;
   NS_ASSERTION(mSuspendNativeCount >= 0, "Unbalanced call to nsBaseAppShell::ResumeNative!");
   return NS_OK;
 }
 
 NS_IMETHODIMP
+nsBaseAppShell::WakeupNative(PRBool mayWait)
+{
+  return OnProcessNextEvent(nsnull, PR_FALSE, 0);
+}
+
+NS_IMETHODIMP
 nsBaseAppShell::GetEventloopNestingLevel(PRUint32* aNestingLevelResult)
 {
   NS_ENSURE_ARG_POINTER(aNestingLevelResult);
 
   *aNestingLevelResult = mEventloopNestingLevel;
 
   return NS_OK;
 }
