--- mozilla/security/manager/pki/resources/content/exceptionDialog.js.orig	2008-08-07 11:42:34.000000000 +0300
+++ mozilla/security/manager/pki/resources/content/exceptionDialog.js	2008-08-06 15:16:09.000000000 +0300
@@ -75,35 +75,38 @@ function initExceptionDialog() {
   gDialog = document.documentElement;
   gBundleBrand = srGetStrBundle("chrome://branding/locale/brand.properties");
   gPKIBundle = srGetStrBundle("chrome://pippki/locale/pippki.properties");
 
   var brandName = gBundleBrand.GetStringFromName("brandShortName");
   
   setText("warningText", gPKIBundle.formatStringFromName("addExceptionBrandedWarning2",
                                                          [brandName], 1));
   gDialog.getButton("extra1").disabled = true;
   
-  var args = window.arguments;
-  if (args && args[0]) {
-    if (args[0].location) {
-      // We were pre-seeded with a location.
-      document.getElementById("locationTextBox").value = args[0].location;
+  var args = window.arguments[0];
+  var paramBlock = args.QueryInterface(Components.interfaces.nsIDialogParamBlock);
+  var prefetchCert = paramBlock.GetInt(1);
+  var location = paramBlock.GetString(0);
+  if (paramBlock) {
+    if (location) {
+      // Pre-seeded with a location.
+      document.getElementById('locationTextBox').value = location;
       document.getElementById('checkCertButton').disabled = false;
       
       // We can optionally pre-fetch the certificate too
-      if (args[0].prefetchCert)
+      if (prefetchCert)
         checkCert();
     }
-    
-    // Set out parameter to false by default
-    args[0].exceptionAdded = false; 
   }
+    
+  // Set out parameter "exceptionAdded" to false by default
+  paramBlock.SetInt(0, 0);
 }
 
 // returns true if found and global status could be set
 function findRecentBadCert(uri) {
   try {
     var recentCertsSvc = Components.classes["@mozilla.org/security/recentbadcerts;1"]
                          .getService(Components.interfaces.nsIRecentBadCertsService);
     if (!recentCertsSvc)
       return false;
 
@@ -342,16 +345,20 @@ function addException() {
   
   var permanentCheckbox = document.getElementById("permanent");
 
   var uri = getURI();
   overrideService.rememberValidityOverride(
     uri.asciiHost, uri.port,
     gCert,
     flags,
     !permanentCheckbox.checked);
   
-  var args = window.arguments;
-  if (args && args[0])
-    args[0].exceptionAdded = true;
+  var args = window.arguments[0];
+  var paramBlock = args.QueryInterface(Components.interfaces.nsIDialogParamBlock);
+  if (args && paramBlock)
+  {
+    // Set the exceptionAdded to true
+    paramBlock.SetInt(0, 1);
+  }
   
   gDialog.acceptDialog();
 }
