#https://bugzilla.mozilla.org/show_bug.cgi?id=532623
Index: mozilla/xpcom/base/nsMemoryImpl.cpp
===================================================================
--- mozilla.orig/xpcom/base/nsMemoryImpl.cpp
+++ mozilla/xpcom/base/nsMemoryImpl.cpp
@@ -56,16 +56,17 @@
 #if defined(XP_WIN)
 #include <windows.h>
 #endif
 
 #if defined(MOZ_PLATFORM_MAEMO)
 #include <osso-mem.h>
 #include <fcntl.h>
 #include <unistd.h>
+#include <time.h>
 static const char kHighMark[] = "/sys/kernel/high_watermark";
 #endif
 
 // Some platforms notify you when system memory is low, others do not.
 // In the case of those that do not, we want to post low memory
 // notifications from IsLowMemory().  For those that can notify us, that
 // code usually lives in toolkit.
 #ifdef WINCE
@@ -130,16 +131,24 @@ nsMemoryImpl::IsLowMemory(PRBool *result
     *result = PR_FALSE;
 #elif defined(XP_WIN)
     MEMORYSTATUSEX stat;
     stat.dwLength = sizeof stat;
     GlobalMemoryStatusEx(&stat);
     *result = (stat.ullAvailPageFile < kRequiredMemory) &&
         ((float)stat.ullAvailPageFile / stat.ullTotalPageFile) < 0.1;
 #elif defined(MOZ_PLATFORM_MAEMO)
+    static time_t sTimeLastCheck = 0;
+    static PRBool sLastLowMemValue = PR_FALSE;
+    time_t newTime = time(0);
+    if (newTime - sTimeLastCheck < 1) {
+      *result = sLastLowMemValue;
+      return NS_OK;
+    }
+    sTimeLastCheck = newTime;
     static int osso_highmark_fd = -1;
     if (osso_highmark_fd == -1) {
         osso_highmark_fd = open (kHighMark, O_RDONLY);
 
         if (osso_highmark_fd == -1) {
             NS_ERROR("can't find the osso highmark file");    
             *result = PR_FALSE;
             return NS_OK;
@@ -147,17 +156,18 @@ nsMemoryImpl::IsLowMemory(PRBool *result
     }
 
     // be kind, rewind.
     lseek(osso_highmark_fd, 0L, SEEK_SET);
 
     int c = 0;
     read (osso_highmark_fd, &c, 1);
 
-    *result = (c == '1');
+    sLastLowMemValue = (c == '1');
+    *result = sLastLowMemValue;
 #else
     *result = PR_FALSE;
 #endif
 
 #ifdef NOTIFY_LOW_MEMORY
     if (*result) {
         sGlobalMemory.FlushMemory(NS_LITERAL_STRING("low-memory").get(), PR_FALSE);
     }
