# XXX turn this off for minimo builds
# 
Index: mozilla/layout/style/nsCSSScanner.h
===================================================================
--- mozilla.orig/layout/style/nsCSSScanner.h
+++ mozilla/layout/style/nsCSSScanner.h
@@ -42,17 +42,17 @@
 #ifndef nsCSSScanner_h___
 #define nsCSSScanner_h___
 
 #include "nsString.h"
 #include "nsCOMPtr.h"
 class nsIUnicharInputStream;
 
 // XXX turn this off for minimo builds
-#define CSS_REPORT_PARSE_ERRORS
+//#define CSS_REPORT_PARSE_ERRORS
 
 #define CSS_BUFFER_SIZE 256
 
 // for #ifdef CSS_REPORT_PARSE_ERRORS
 #include "nsXPIDLString.h"
 class nsIURI;
 
 // Token types
