diff -r 02ef5b6d0c3d layout/base/nsPresContext.cpp
--- a/layout/base/nsPresContext.cpp	Sat May 23 09:51:44 2009 +0300
+++ b/layout/base/nsPresContext.cpp	Thu May 28 23:13:52 2009 +0300
@@ -198,7 +198,7 @@ nsPresContext::nsPresContext(nsIDocument
                         NS_FONT_VARIANT_NORMAL, NS_FONT_WEIGHT_NORMAL,
                         NS_FONT_STRETCH_NORMAL, 0, 0),
     mCanPaginatedScroll(PR_FALSE),
-    mIsRootPaginatedDocument(PR_FALSE), mSupressResizeReflow(PR_FALSE)
+    mIsRootPaginatedDocument(PR_FALSE), mSupressResizeReflow(PR_FALSE), mResizeEventSuspended(PR_FALSE)
 {
   // NOTE! nsPresContext::operator new() zeroes out all members, so don't
   // bother initializing members to 0.
diff -r 02ef5b6d0c3d layout/base/nsPresContext.h
--- a/layout/base/nsPresContext.h	Sat May 23 09:51:44 2009 +0300
+++ b/layout/base/nsPresContext.h	Thu May 28 23:13:52 2009 +0300
@@ -807,6 +807,9 @@ public:
     mCrossDocDirtyRegion.SetEmpty();
   }
 
+  void SuspendResizeEvent(PRBool aSuspend) { mResizeEventSuspended = aSuspend; };
+  PRBool IsResizeEventSuspended(void) const { return mResizeEventSuspended; };
+
 protected:
   friend class nsRunnableMethod<nsPresContext>;
   NS_HIDDEN_(void) ThemeChangedInternal();
@@ -964,6 +967,7 @@ protected:
   PRBool                mInitialized;
 #endif
 
+  unsigned              mResizeEventSuspended : 1;
 
 protected:
 
diff -r 02ef5b6d0c3d layout/base/nsPresShell.cpp
--- a/layout/base/nsPresShell.cpp	Sat May 23 09:51:44 2009 +0300
+++ b/layout/base/nsPresShell.cpp	Thu May 28 23:13:52 2009 +0300
@@ -2716,7 +2716,7 @@ PresShell::ResizeReflow(nscoord aWidth, 
       nsRect(0, 0, aWidth, rootFrame->GetRect().height));
   }
 
-  if (!mIsDestroying && !mResizeEvent.IsPending()) {
+  if (!mIsDestroying && !mResizeEvent.IsPending() && !GetPresContext()->IsResizeEventSuspended()) {
     nsRefPtr<nsRunnableMethod<PresShell> > resizeEvent =
       NS_NEW_RUNNABLE_METHOD(PresShell, this, FireResizeEvent);
     if (NS_SUCCEEDED(NS_DispatchToCurrentThread(resizeEvent))) {
diff -r 11385e70f644 layout/generic/nsGfxScrollFrame.cpp
--- a/layout/generic/nsGfxScrollFrame.cpp	Sat Jun 13 16:21:32 2009 +0300
+++ b/layout/generic/nsGfxScrollFrame.cpp	Sat Jun 13 16:22:58 2009 +0300
@@ -1585,18 +1585,22 @@ nsGfxScrollFrameInner::FireScrollPortEve
   }
 
   nsScrollPortEvent event(PR_TRUE,
                           (orient == nsScrollPortEvent::horizontal ?
                            mHorizontalOverflow : mVerticalOverflow) ?
                             NS_SCROLLPORT_OVERFLOW : NS_SCROLLPORT_UNDERFLOW,
                           nsnull);
   event.orient = orient;
-  return nsEventDispatcher::Dispatch(mOuter->GetContent(),
-                                     mOuter->PresContext(), &event);
+  nsPresContext *ctx = mOuter->PresContext();
+  ctx->SuspendResizeEvent(PR_TRUE);
+  nsresult rv = nsEventDispatcher::Dispatch(mOuter->GetContent(),
+                                            mOuter->PresContext(), &event);
+  ctx->SuspendResizeEvent(PR_FALSE);
+  return rv;
 }
 
 void
 nsGfxScrollFrameInner::ReloadChildFrames()
 {
   mScrolledFrame = nsnull;
   mHScrollbarBox = nsnull;
   mVScrollbarBox = nsnull;
