diff -r -u mozilla.orig/gfx/thebes/src/gfxPlatformGtk.cpp mozilla/gfx/thebes/src/gfxPlatformGtk.cpp
--- mozilla.orig/gfx/thebes/src/gfxPlatformGtk.cpp	2008-07-22 20:07:46.000000000 +0300
+++ mozilla/gfx/thebes/src/gfxPlatformGtk.cpp	2008-07-22 20:12:42.000000000 +0300
@@ -145,8 +145,15 @@
 
     if (!UseGlitz()) {
         GdkPixmap* pixmap = nsnull;
-        XRenderPictFormat* xrenderFormat =
-            XRenderFindStandardFormat(display, xrenderFormatID);
+        XRenderPictFormat* xrenderFormat = nsnull;
+
+        if (xrenderFormatID == PictStandardRGB24) {
+            // Actually we want not strictly RGB24, but the fastest non-alpha
+            // format (XXX to be verified if it is actually safe)
+            xrenderFormat = XRenderFindVisualFormat(display, GDK_VISUAL_XVISUAL(gdk_visual_get_system()));
+        }
+
+        if (!xrenderFormat) xrenderFormat = XRenderFindStandardFormat(display, xrenderFormatID);
 
         if (xrenderFormat) {
             pixmap = gdk_pixmap_new(nsnull, size.width, size.height,
