# Hide all not used symbols from engine
Index: mozilla/toolkit/library/libxul-export.map
===================================================================
--- /dev/null
+++ mozilla/toolkit/library/libxul-export.map
@@ -0,0 +1,29 @@
+{
+  global:
+	_GLOBAL_OFFSET_TABLE_;
+	__exidx_end;
+	_bss_end__;
+	__bss_start__;
+	__dso_handle;
+	__libc_csu_fini;
+	__exidx_start;
+	_init;
+	__bss_end__;
+	_start;
+	__libc_csu_init;
+	__end__;
+	_fini;
+	_edata;
+	_end;
+	_IO_stdin_used;
+	__data_start;
+	NS_*;
+	CRMF_*;
+	XRE_*;
+   _ZTV14nsGetInterface;
+   _ZNK12nsIPresShell18GetRootScrollFrameE;
+   gtk_moz_*;
+   *gfx*;
+   *XRE*;
+  local: *;
+};
Index: mozilla/toolkit/library/libxul-rules.mk
===================================================================
--- mozilla.orig/toolkit/library/libxul-rules.mk
+++ mozilla/toolkit/library/libxul-rules.mk
@@ -89,14 +89,17 @@
 endif
 
 ifeq ($(MOZ_WIDGET_TOOLKIT),gtk2)
 EXTRA_DSO_LDOPTS += $(MOZ_PANGO_LIBS)
 endif
 
 EXTRA_DSO_LDOPTS += $(MOZ_CAIRO_LIBS)
 
+# Hack for Maemo: exporting only limited set of symbols for Linux
+EXTRA_DSO_LDOPTS += -Wl,--version-script=$(srcdir)/libxul-export.map
+
 export:: dlldeps.cpp
 
 dlldeps.cpp: $(topsrcdir)/xpcom/build/dlldeps.cpp
 	$(INSTALL) $^ .
 
 GARBAGE += dlldeps.cpp
