Index: mozilla/toolkit/components/passwordmgr/src/nsLoginManagerPrompter.js
===================================================================
--- mozilla.orig/toolkit/components/passwordmgr/src/nsLoginManagerPrompter.js
+++ mozilla/toolkit/components/passwordmgr/src/nsLoginManagerPrompter.js
@@ -425,16 +425,20 @@
             this.log("===== promptAuth called =====");
 
             // If the user submits a login but it fails, we need to remove the
             // notification bar that was displayed. Conveniently, the user will
             // be prompted for authentication again, which brings us here.
             var notifyBox = this._getNotifyBox();
             if (notifyBox)
                 this._removeLoginNotifications(notifyBox);
+            if (this._xulRememberDialog) {
+                this._xulRememberDialog.close();
+                this._xulRememberDialog = null;
+            }
 
             var [hostname, httpRealm] = this._getAuthTarget(aChannel, aAuthInfo);
 
 
             // Looks for existing logins to prefill the prompt with.
             var foundLogins = this._pwmgr.findLogins({},
                                         hostname, null, httpRealm);
             this.log("found " + foundLogins.length + " matching logins.");
@@ -550,20 +554,59 @@
      * promptToSavePassword
      *
      */
     promptToSavePassword : function (aLogin) {
         var notifyBox = this._getNotifyBox();
 
         if (notifyBox)
             this._showSaveLoginNotification(notifyBox, aLogin);
-        else
-            this._showSaveLoginDialog(aLogin);
+        else if (!this._showXulSaveLoginDialog(aLogin))
+            {
+                this._showSaveLoginDialog(aLogin);
+            }
+    },
+
+    /*
+     * xulSaveLoginDialog
+     *
+     */
+    _showXulSaveLoginDialog : function (aLogin) {
+        var ww = Components.classes["@mozilla.org/embedcomp/window-watcher;1"]
+                            .getService(Components.interfaces.nsIWindowWatcher);
+        var activeWindow = ww.activeWindow;
+
+        var brandShortName = this._brandBundle.GetStringFromName("brandShortName");
+        var params = {inn:{prompter:this, logindata:aLogin, brand:brandShortName}, out:null};
+
+        var ret = this._xulRememberDialog = activeWindow.openDialog("chrome://global/content/savePasswordDialog.xul",
+                        "a_window",
+                        "dialog",
+                        params);
+
+        return ret;
+    },
+
+    _xulRememberCb : function (aLogin) {
+        this.log("Saving login for " + aLogin.hostname);
+        this._pwmgr.addLogin(aLogin);
+        this._xulRememberDialog.close();
+        this._xulRememberDialog = null;
+    },
+
+    _xulNeverCb : function (aLogin) {
+        this.log("Disabling " + aLogin.hostname + " logins by request.");
+        this._pwmgr.setLoginSavingEnabled(aLogin.hostname, false);
+        this._xulRememberDialog.close();
+        this._xulRememberDialog = null;
     },
 
+    _xulCancelCb : function (aLogin) {
+        this._xulRememberDialog = null;
+    },
 
     /*
      * _showLoginNotification
      *
      * Displays a notification bar.
      *
      */
     _showLoginNotification : function (aNotifyBox, aName, aText, aButtons) {
Index: mozilla/toolkit/content/savePasswordDialog.js
===================================================================
--- /dev/null
+++ mozilla/toolkit/content/savePasswordDialog.js
@@ -0,0 +1,74 @@
+/* -*- Mode: Java; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+/* ***** BEGIN LICENSE BLOCK *****
+ * Version: MPL 1.1/GPL 2.0/LGPL 2.1
+ *
+ * The contents of this file are subject to the Mozilla Public License Version
+ * 1.1 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ * http://www.mozilla.org/MPL/
+ *
+ * Software distributed under the License is distributed on an "AS IS" basis,
+ * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
+ * for the specific language governing rights and limitations under the
+ * License.
+ *
+ * The Original Code is Mozilla Communicator client code.
+ *
+ * The Initial Developer of the Original Code is
+ * Netscape Communications Corporation.
+ * Portions created by the Initial Developer are Copyright (C) 1998
+ * the Initial Developer. All Rights Reserved.
+ *
+ * Contributor(s):
+ *
+ * Alternatively, the contents of this file may be used under the terms of
+ * either the GNU General Public License Version 2 or later (the "GPL"), or
+ * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
+ * in which case the provisions of the GPL or the LGPL are applicable instead
+ * of those above. If you wish to allow use of your version of this file only
+ * under the terms of either the GPL or the LGPL, and not to allow others to
+ * use your version of this file under the terms of the MPL, indicate your
+ * decision by deleting the provisions above and replace them with the notice
+ * and other provisions required by the GPL or the LGPL. If you do not delete
+ * the provisions above, a recipient may use your version of this file under
+ * the terms of any one of the MPL, the GPL or the LGPL.
+ *
+ * ***** END LICENSE BLOCK ***** */
+
+var param;
+var prompter;
+var logindata;
+
+function onLoad() {
+    prompter = window.arguments[0].inn.prompter;
+    logindata = window.arguments[0].inn.logindata;
+    var brand = window.arguments[0].inn.brand;
+    var dialog = document.documentElement;
+    var stringBundle = document.getElementById("strings");
+
+    // The following does not work, see Mozilla bug 316920,
+    // Dialog title not updated on-screen although dialog.title DOM attribute is changed
+    dialog.title = stringBundle.getString("savePasswordTitle");
+
+    dialog.getButton("extra1").label = stringBundle.getString("notifyBarNeverForSiteButtonText");
+    dialog.getButton("extra2").label = stringBundle.getString("notifyBarRememberButtonText");
+    dialog.getButton("cancel").label = stringBundle.getString("notifyBarNotNowButtonText");
+    var infotext = stringBundle.getFormattedString("savePasswordText", [brand]);
+    document.getElementById("info").setAttribute("value", infotext);
+
+    // This does nothing?
+    dialog.centerWindowOnScreen();
+}
+
+function doCancel() {
+        prompter._xulCancelCb(logindata);
+}
+
+function doRemember() {
+        prompter._xulRememberCb(logindata);
+}
+
+function doNever() {
+        prompter._xulNeverCb(logindata);
+}
+
Index: mozilla/toolkit/content/savePasswordDialog.xul
===================================================================
--- /dev/null
+++ mozilla/toolkit/content/savePasswordDialog.xul
@@ -0,0 +1,56 @@
+<?xml version="1.0"?>
+
+<!-- ***** BEGIN LICENSE BLOCK *****
+ Version: MPL 1.1/GPL 2.0/LGPL 2.1
+
+ The contents of this file are subject to the Mozilla Public License Version
+ 1.1 (the "License"); you may not use this file except in compliance with
+ the License. You may obtain a copy of the License at
+ http://www.mozilla.org/MPL/
+
+ Software distributed under the License is distributed on an "AS IS" basis,
+ WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
+ for the specific language governing rights and limitations under the
+ License.
+
+ The Original Code is Mozilla Communicator client code, released
+ March 31, 1998.
+
+ The Initial Developer of the Original Code is
+ Netscape Communications Corporation.
+ Portions created by the Initial Developer are Copyright (C) 1998-1999
+ the Initial Developer. All Rights Reserved.
+
+ Contributor(s):
+
+ Alternatively, the contents of this file may be used under the terms of
+ either the GNU General Public License Version 2 or later (the "GPL"), or
+ the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
+ in which case the provisions of the GPL or the LGPL are applicable instead
+ of those above. If you wish to allow use of your version of this file only
+ under the terms of either the GPL or the LGPL, and not to allow others to
+ use your version of this file under the terms of the MPL, indicate your
+ decision by deleting the provisions above and replace them with the notice
+ and other provisions required by the GPL or the LGPL. If you do not delete
+ the provisions above, a recipient may use your version of this file under
+ the terms of any one of the MPL, the GPL or the LGPL.
+
+ ***** END LICENSE BLOCK ***** -->
+
+<?xml-stylesheet href="chrome://global/skin/global.css" type="text/css"?>
+
+<dialog id="rememberpassword" title="Confirm?"
+   xmlns="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul"
+   buttons="extra1,extra2,cancel"
+   onload="onLoad();"
+   ondialogcancel="return doCancel();"
+   ondialogextra1="return doNever();"
+   ondialogextra2="return doRemember();">
+
+  <stringbundle id="strings"
+                src="chrome://passwordmgr/locale/passwordmgr.properties"/>
+
+   <script type="application/x-javascript" src="chrome://global/content/savePasswordDialog.js"/>
+
+   <label id="info"/>
+</dialog>
Index: mozilla/toolkit/content/jar.mn
===================================================================
--- mozilla.orig/toolkit/content/jar.mn
+++ mozilla/toolkit/content/jar.mn
@@ -30,16 +30,18 @@
 *+ content/global/findUtils.js                (findUtils.js)
 *+ content/global/globalOverlay.js            (globalOverlay.js)
 *+ content/global/inlineSpellCheckUI.js       (inlineSpellCheckUI.js)
 +  content/global/mozilla.xhtml               (mozilla.xhtml)
 *+ content/global/nsDragAndDrop.js            (nsDragAndDrop.js)
 +  content/global/selectDialog.js             (selectDialog.js)
 +  content/global/selectDialog.xul            (selectDialog.xul)
 *+ content/global/viewZoomOverlay.js          (viewZoomOverlay.js)
+*+ content/global/savePasswordDialog.js       (savePasswordDialog.js)
+*+ content/global/savePasswordDialog.xul       (savePasswordDialog.xul)
 *+ content/global/bindings/autocomplete.xml    (widgets/autocomplete.xml)
 *+ content/global/bindings/browser.xml         (widgets/browser.xml)
 *+ content/global/bindings/button.xml          (widgets/button.xml)
 *+ content/global/bindings/checkbox.xml        (widgets/checkbox.xml)
 *+ content/global/bindings/colorpicker.xml     (widgets/colorpicker.xml)
 *+ content/global/bindings/datetimepicker.xml  (widgets/datetimepicker.xml)
 *+ content/global/bindings/dialog.xml          (widgets/dialog.xml)
 *+ content/global/bindings/editor.xml          (widgets/editor.xml)
