#https://bugzilla.mozilla.org/show_bug.cgi?id=532614
--- mozilla/toolkit/system/dbus/nsDBusService.cpp.orig	2009-07-30 08:44:35.000000000 +0300
+++ mozilla/toolkit/system/dbus/nsDBusService.cpp	2009-07-29 15:44:11.000000000 +0300
@@ -129,20 +129,21 @@ void nsDBusService::DoTimerCallback(nsIT
     }
   }
 }
 
 static void TimerCallback(nsITimer *aTimer, void *aClosure) {
   static_cast<nsDBusService*>(aClosure)->DoTimerCallback(aTimer);
 }
 
 void nsDBusService::DropConnection() {
   if (mConnection) {
+    dbus_connection_remove_filter(mConnection, dbus_filter, this);
     if (mSingleClient) {
       mSingleClient->UnregisterWithConnection(mConnection);
     }
     dbus_connection_unref(mConnection);
     mConnection = nsnull;
   }
 }
 
 void nsDBusService::HandleDBusDisconnect() {
   DropConnection();
