# Removed unused headers mozilla/browser/components...
# 
Index: mozilla/browser/components/build/nsModule.cpp
===================================================================
--- mozilla.orig/browser/components/build/nsModule.cpp
+++ mozilla/browser/components/build/nsModule.cpp
@@ -67,17 +67,16 @@
 #include "nsOmniWebProfileMigrator.h"
 #include "nsMacIEProfileMigrator.h"
 #include "nsCaminoProfileMigrator.h"
 #include "nsICabProfileMigrator.h"
 #endif
 
 #endif // WINCE
 
-#include "rdf.h"
 #include "nsFeedSniffer.h"
 #include "nsAboutFeeds.h"
 #include "nsIAboutModule.h"
 
 #include "nsPrivateBrowsingServiceWrapper.h"
 #include "nsNetCID.h"
 
 /////////////////////////////////////////////////////////////////////////////
Index: mozilla/browser/components/migration/src/nsOperaProfileMigrator.cpp
===================================================================
--- mozilla.orig/browser/components/migration/src/nsOperaProfileMigrator.cpp
+++ mozilla/browser/components/migration/src/nsOperaProfileMigrator.cpp
@@ -51,18 +51,16 @@
 #include "nsILocalFile.h"
 #include "nsINIParser.h"
 #include "nsIObserverService.h"
 #include "nsIPermissionManager.h"
 #include "nsIPrefLocalizedString.h"
 #include "nsIPrefService.h"
 #include "nsIProfileMigrator.h"
 #include "nsIProperties.h"
-#include "nsIRDFContainer.h"
-#include "nsIRDFService.h"
 #include "nsIServiceManager.h"
 #include "nsIStringBundle.h"
 #include "nsISupportsPrimitives.h"
 #include "nsNetUtil.h"
 #include "nsOperaProfileMigrator.h"
 #include "nsToolkitCompsCID.h"
 #ifdef XP_WIN
 #include <windows.h>
Index: mozilla/browser/components/migration/src/nsBrowserProfileMigratorUtils.cpp
===================================================================
--- mozilla.orig/browser/components/migration/src/nsBrowserProfileMigratorUtils.cpp
+++ mozilla/browser/components/migration/src/nsBrowserProfileMigratorUtils.cpp
@@ -47,17 +47,16 @@
 #include "nsIProperties.h"
 #include "nsIProfileMigrator.h"
 
 #include "nsIURI.h"
 #include "nsNetUtil.h"
 #include "nsISupportsPrimitives.h"
 
 #include "nsAppDirectoryServiceDefs.h"
-#include "nsIRDFService.h"
 #include "nsIStringBundle.h"
 #include "nsISupportsArray.h"
 #include "nsXPCOMCID.h"
 
 #define MIGRATION_BUNDLE "chrome://browser/locale/migration/migration.properties"
 
 #define BOOKMARKS_FILE_NAME NS_LITERAL_STRING("bookmarks.html")
 
