--- mozilla/layout/generic/nsLineLayout.cpp.orig	2007-11-20 16:36:21.000000000 +0200
+++ mozilla/layout/generic/nsLineLayout.cpp	2007-11-21 03:34:30.000000000 +0200
@@ -755,17 +755,20 @@ IsPercentageAware(const nsIFrame* aFrame
         fType == nsGkAtoms::HTMLButtonControlFrame ||
         fType == nsGkAtoms::gfxButtonControlFrame ||
         fType == nsGkAtoms::fieldSetFrame ||
         fType == nsGkAtoms::comboboxDisplayFrame) {
       return PR_TRUE;
     }
 
     // Handle SVG, which doesn't map width/height into style
-    if ((fType == nsGkAtoms::svgOuterSVGFrame ||
+    if ((
+#ifdef MOZ_SVG
+         fType == nsGkAtoms::svgOuterSVGFrame ||
+#endif
          fType == nsGkAtoms::subDocumentFrame) &&
         const_cast<nsIFrame*>(aFrame)->GetIntrinsicSize().width.GetUnit() ==
         eStyleUnit_Percent) {
       return PR_TRUE;
     }
   }
 
   return PR_FALSE;
