--- mozilla/client.mk.orig	2008-05-16 22:15:53.000000000 +0300
+++ mozilla/client.mk	2008-05-20 15:34:09.000000000 +0300
@@ -89,15 +89,11 @@ ifeq (,$(wildcard client.mk))
 $(error Must run from the client.mk directory, or specify TOPSRCDIR)
 endif
 TOPSRCDIR = $(CWD)
 endif
 
-ifeq (Darwin,$(shell uname -s))
-AUTOCONF ?= autoconf213
-else
-AUTOCONF ?= autoconf-2.13
-endif
+AUTOCONF ?= $(shell which autoconf-2.13 autoconf2.13 autoconf213 | head -1)
 
 MKDIR := mkdir
 SH := /bin/sh
 ifndef MAKE
 MAKE := gmake
