#https://bugzilla.mozilla.org/show_bug.cgi?id=465170
# workarround
Index: mozilla/js/src/nanojit/LIR.cpp
===================================================================
--- mozilla.orig/js/src/nanojit/LIR.cpp
+++ mozilla/js/src/nanojit/LIR.cpp
@@ -892,21 +892,16 @@ namespace nanojit
 				v == LIR_xor || v == LIR_or || v == LIR_and ||
 				v == LIR_eq) {
 				// move const to rhs
 				LIns* t = oprnd2;
 				oprnd2 = oprnd1;
 				oprnd1 = t;
 			}
 			else if (v >= LIR_lt && v <= LIR_uge) {
-				NanoStaticAssert((LIR_lt ^ 1) == LIR_gt);
-				NanoStaticAssert((LIR_le ^ 1) == LIR_ge);
-				NanoStaticAssert((LIR_ult ^ 1) == LIR_ugt);
-				NanoStaticAssert((LIR_ule ^ 1) == LIR_uge);
-
 				// move const to rhs, swap the operator
 				LIns *t = oprnd2;
 				oprnd2 = oprnd1;
 				oprnd1 = t;
 				v = LOpcode(v^1);
 			}
 		}
 
