diff -r 27dbaf87bb59 dom/src/geolocation/MaemoLocationProvider.cpp
--- a/dom/src/geolocation/MaemoLocationProvider.cpp	Sat Nov 15 12:57:00 2008 -0500
+++ b/dom/src/geolocation/MaemoLocationProvider.cpp	Sat Nov 15 13:33:23 2008 -0500
@@ -33,16 +33,18 @@
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
 
 
 #include "MaemoLocationProvider.h"
 #include "nsGeolocation.h"
+#include "nsIClassInfo.h"
+#include "nsDOMClassInfoID.h"
 
 
 ////////////////////////////////////////////////////
 // nsGeoPosition
 ////////////////////////////////////////////////////
 
 /**
  * Simple object that holds a single point in space.
