diff -r 9a49d707e10d config/system-headers
--- a/config/system-headers	Wed May 06 08:43:50 2009 +0300
+++ b/config/system-headers	Wed May 06 09:21:55 2009 +0300
@@ -1008,6 +1008,7 @@ bzlib.h
 #endif
 #ifdef MOZ_PLATFORM_HILDON
 hildon-uri.h
+hildon-mime.h
 #endif
 #ifdef NS_OSSO
 libosso.h
diff -r 9a49d707e10d js/src/config/system-headers
--- a/js/src/config/system-headers	Wed May 06 08:43:50 2009 +0300
+++ b/js/src/config/system-headers	Wed May 06 09:21:55 2009 +0300
@@ -1008,6 +1008,7 @@ bzlib.h
 #endif
 #ifdef MOZ_PLATFORM_HILDON
 hildon-uri.h
+hildon-mime.h
 #endif
 #ifdef NS_OSSO
 libosso.h
diff -r 9a49d707e10d uriloader/exthandler/Makefile.in
--- a/uriloader/exthandler/Makefile.in	Wed May 06 08:43:50 2009 +0300
+++ b/uriloader/exthandler/Makefile.in	Wed May 06 09:21:55 2009 +0300
@@ -168,6 +168,8 @@ ifdef MOZ_ENABLE_GNOMEVFS
 LOCAL_INCLUDES   += $(MOZ_GNOMEVFS_CFLAGS)
 EXTRA_DSO_LDOPTS += $(MOZ_GNOMEVFS_LIBS)
 endif
+LOCAL_INCLUDES   += $(LIBHILDONMIME_CFLAGS)
+EXTRA_DSO_LDOPTS += $(LIBHILDONMIME_LIBS)
 endif
 
 ifeq ($(OS_ARCH),WINNT WINCE)
