# bug 442109
Index: mozilla/modules/plugin/base/src/nsPluginNativeWindowGtk2.cpp
===================================================================
--- mozilla.orig/modules/plugin/base/src/nsPluginNativeWindowGtk2.cpp
+++ mozilla/modules/plugin/base/src/nsPluginNativeWindowGtk2.cpp
@@ -55,16 +55,18 @@
 
 #ifdef MOZ_COMPOSITED_PLUGINS
 #include "nsPluginInstancePeer.h"
 
 extern "C" {
 #include <X11/extensions/Xdamage.h>
 #include <X11/extensions/Xcomposite.h>
 }
+#include "nsIPluginInstancePeer2.h"
+#include "nsPluginInstancePeer.h"
 #endif
 
 #include "gtk2xtbin.h"
 
 class nsPluginNativeWindowGtk2 : public nsPluginNativeWindow {
 public: 
   nsPluginNativeWindowGtk2();
   virtual ~nsPluginNativeWindowGtk2();
