# https://bugzilla.mozilla.org/show_bug.cgi?id=469667
#
--- mozilla/widget/src/gtk2/nsAppShell.cpp.orig	2008-12-12 16:46:07.000000000 +0200
+++ mozilla/widget/src/gtk2/nsAppShell.cpp	2008-12-12 16:47:20.000000000 +0200
@@ -155,8 +155,17 @@ nsAppShell::ScheduleNativeEventCallback(
     write(mPipeFDs[1], buf, 1);
 }
 
+static guint timerID;
+static gboolean watchdog_timeout(gpointer data)
+{
+    timerID = 0;
+    return FALSE;
+}
+
 PRBool
 nsAppShell::ProcessNextNativeEvent(PRBool mayWait)
 {
+    if (mayWait && !timerID && !g_main_context_pending(NULL))
+        timerID = g_timeout_add(1000, watchdog_timeout, 0);
     return g_main_context_iteration(NULL, mayWait);
 }
