# Build configuration for HOST libidl libraries
Index: mozilla/configure.in
===================================================================
--- mozilla.orig/configure.in
+++ mozilla/configure.in
@@ -7651,18 +7651,19 @@ int main() {
     fi
 fi
 
 if test -n "$CROSS_COMPILE"; then
      if test -z "$HOST_LIBIDL_CONFIG"; then
         HOST_LIBIDL_CONFIG="$LIBIDL_CONFIG"
     fi
     if test -n "$HOST_LIBIDL_CONFIG" && test "$HOST_LIBIDL_CONFIG" != "no"; then
-        HOST_LIBIDL_CFLAGS=`${HOST_LIBIDL_CONFIG} --cflags`
-        HOST_LIBIDL_LIBS=`${HOST_LIBIDL_CONFIG} --libs`
+        HOST_LIBIDL_CFLAGS=`PKG_CONFIG_PATH="${prefix}/lib/pkgconfig" ${HOST_LIBIDL_CONFIG} --cflags`
+        HOST_LIBIDL_LIBS=`PKG_CONFIG_PATH="${prefix}/lib/pkgconfig" ${HOST_LIBIDL_CONFIG} --libs`
+        HOST_LIBIDL_LIBS="$HOST_LIBIDL_LIBS -lpthread -lrt"
     else
         HOST_LIBIDL_CFLAGS="$LIBIDL_CFLAGS"
         HOST_LIBIDL_LIBS="$LIBIDL_LIBS"
     fi
 fi
 
 if test -z "$SKIP_PATH_CHECKS"; then
 if test -z "${GLIB_CFLAGS}" || test -z "${GLIB_LIBS}" ; then
