#Bug 376533  uClibc strings/index and childiterator.index conficts
Index: mozilla/layout/base/nsChildIterator.h
===================================================================
--- mozilla.orig/layout/base/nsChildIterator.h
+++ mozilla/layout/base/nsChildIterator.h
@@ -41,16 +41,20 @@
  * used to construct frames (normal children or something from XBL)
  */
 
 #include "nsCOMPtr.h"
 #include "nsIContent.h"
 #include "nsIDOMNodeList.h"
 #include "nsIDOMNode.h"
 
+#ifdef __UCLIBC__
+#undef index
+#endif
+
 /**
  * Helper class for iterating children during frame construction.
  * This class should always be used in lieu of the straight content
  * node APIs, since it handles XBL-generated anonymous content as
  * well.
  */
 class NS_STACK_CLASS ChildIterator
 {
