--- mozilla/build/autoconf/freetype2.m4.orig	2007-12-27 08:20:33.000000000 +0200
+++ mozilla/build/autoconf/freetype2.m4	2007-12-27 08:32:12.000000000 +0200
@@ -25,21 +25,26 @@ if test x$ft_config_exec_prefix != x ; t
   if test x${FT2_CONFIG+set} != xset ; then
     FT2_CONFIG=$ft_config_exec_prefix/bin/freetype-config
   fi
 fi
 if test x$ft_config_prefix != x ; then
   ft_config_args="$ft_config_args --prefix=$ft_config_prefix"
   if test x${FT2_CONFIG+set} != xset ; then
     FT2_CONFIG=$ft_config_prefix/bin/freetype-config
   fi
 fi
+if test x$FT2_CONFIG == x ; then
+  FT2_CONFIG=$prefix/bin/pkg-config
+  ft_config_args="$ft_config_args freetype2"
+else
 AC_PATH_PROG(FT2_CONFIG, freetype-config, no)
+fi
 
 min_ft_version=ifelse([$1], ,6.1.0,$1)
 AC_MSG_CHECKING(for FreeType - version >= $min_ft_version)
 no_ft=""
 if test "$FT2_CONFIG" = "no" ; then
   no_ft=yes
 else
   FT2_CFLAGS=`$FT2_CONFIG $ft_config_args --cflags`
   FT2_LIBS=`$FT2_CONFIG $ft_config_args --libs`
   ft_config_major_version=`$FT2_CONFIG $ft_config_args --version | \
