--- mozilla/security/nss/lib/freebl/mpi/mpcpucache.c.orig	2006-04-28 20:06:22.000000000 +0300
+++ mozilla/security/nss/lib/freebl/mpi/mpcpucache.c	2007-12-27 16:47:33.000000000 +0200
@@ -46,21 +46,21 @@
  *
  * mp_modexp uses this information to make sure that private key information
  * isn't being leaked through the cache.
  *
  * Currently the file returns good data for most modern x86 processors, and
  * reasonable data on 64-bit ppc processors. All other processors are assumed
  * to have a cache line size of 32 bytes unless modified by target.mk.
  * 
  */
 
-#if defined(i386) || defined(__i386) || defined(__X86__) || defined (_M_IX86) || defined(__x86_64__) || defined(__x86_64)
+#if (defined(i386) || defined(__i386) || defined(__X86__) || defined (_M_IX86) || defined(__x86_64__) || defined(__x86_64)) && !defined(__powerpc) && !defined(_MIPSEL)
 /* X86 processors have special instructions that tell us about the cache */
 #include "string.h"
 
 #if defined(__x86_64__) || defined(__x86_64)
 #define AMD_64 1
 #endif
 
 /* Generic CPUID function */
 #if defined(AMD_64)
 static void cpuid(unsigned long op, unsigned long *eax, 
