Index: mozilla/layout/style/nsStyleStruct.h
===================================================================
--- mozilla.orig/layout/style/nsStyleStruct.h
+++ mozilla/layout/style/nsStyleStruct.h
@@ -998,19 +998,22 @@ struct nsStyleDisplay {
            NS_STYLE_DISPLAY_LIST_ITEM == mDisplay ||
            NS_STYLE_DISPLAY_TABLE == mDisplay;
   }
 
   PRBool IsInlineOutside() const {
     return NS_STYLE_DISPLAY_INLINE == mDisplay ||
            NS_STYLE_DISPLAY_INLINE_BLOCK == mDisplay ||
            NS_STYLE_DISPLAY_INLINE_TABLE == mDisplay ||
-           NS_STYLE_DISPLAY_INLINE_BOX == mDisplay ||
-           NS_STYLE_DISPLAY_INLINE_GRID == mDisplay ||
-           NS_STYLE_DISPLAY_INLINE_STACK == mDisplay;
+           NS_STYLE_DISPLAY_INLINE_BOX == mDisplay
+#ifdef MOZ_XUL
+           || NS_STYLE_DISPLAY_INLINE_GRID == mDisplay ||
+           NS_STYLE_DISPLAY_INLINE_STACK == mDisplay
+#endif
+	   ;
   }
 
   PRBool IsFloating() const {
     return NS_STYLE_FLOAT_NONE != mFloats;
   }
 
   PRBool IsAbsolutelyPositioned() const {return (NS_STYLE_POSITION_ABSOLUTE == mPosition) ||
                                                 (NS_STYLE_POSITION_FIXED == mPosition);}
