#Bug 376256 – NSPR cross-compilation with dpkg-cross is fail
--- mozilla/nsprpub/config/Makefile.in.orig	2005-11-23 08:30:32.000000000 +0200
+++ mozilla/nsprpub/config/Makefile.in	2007-04-02 19:49:27.000000000 +0300
@@ -140,21 +140,31 @@ endef
 export:: $(TARGETS)
 	rm -f $(dist_bindir)/nspr-config
 
 ifdef WRAP_SYSTEM_INCLUDES
 export::
 	if test ! -d system_wrappers; then mkdir system_wrappers; fi
 	$(PERL) $(srcdir)/make-system-wrappers.pl system_wrappers < $(srcdir)/system-headers
 	$(INSTALL) system_wrappers $(dist_includedir)
 endif
 
+ifdef CROSS_COMPILE
+$(OBJDIR)/%.$(OBJ_SUFFIX): %.c
+	@$(MAKE_OBJDIR)
+	$(HOST_CC) -o $@ -c $(CFLAGS) $<
+endif
+
 $(OBJDIR)/%$(PROG_SUFFIX): $(OBJDIR)/%.$(OBJ_SUFFIX)
 	@$(MAKE_OBJDIR)
 ifeq ($(MOZ_OS2_TOOLS),VACPP)
 	$(LD) $(EXEFLAGS) $<
 else
+ifdef CROSS_COMPILE
+	$(HOST_CC) $(XCFLAGS) $< $(LDFLAGS) $(XLDOPTS) $(OUTOPTION)$@
+else
 	$(CC) $(XCFLAGS) $< $(LDFLAGS) $(XLDOPTS) $(OUTOPTION)$@
 endif
+endif
 
 install:: nspr.m4
 	$(NSINSTALL) -D $(DESTDIR)$(datadir)/aclocal
 	$(NSINSTALL) -t -m 0644 $< $(DESTDIR)$(datadir)/aclocal
