# https://bugzilla.mozilla.org/show_bug.cgi?id=408925
Index: EmbedPrivate.cpp
===================================================================
RCS file: /cvsroot/mozilla/embedding/browser/gtk/src/EmbedPrivate.cpp,v
retrieving revision 1.93
diff -p -U 8 -r1.93 mozilla/embedding/browser/gtk/src/EmbedPrivate.cpp
--- mozilla/embedding/browser/gtk/src/EmbedPrivate.cpp	30 Jan 2008 04:28:26 -0000	1.93
+++ mozilla/embedding/browser/gtk/src/EmbedPrivate.cpp	9 Jun 2008 14:56:25 -0000
@@ -132,17 +132,18 @@ GTKEmbedDirectoryProvider::GetFile(const
 {
   if (EmbedPrivate::sAppFileLocProvider) {
     nsresult rv = EmbedPrivate::sAppFileLocProvider->GetFile(aKey, aPersist,
                                                              aResult);
     if (NS_SUCCEEDED(rv))
       return rv;
   }
 
-  if (EmbedPrivate::sProfileDir && !strcmp(aKey, NS_APP_USER_PROFILE_50_DIR)) {
+  if (EmbedPrivate::sProfileDir && (!strcmp(aKey, NS_APP_USER_PROFILE_50_DIR)
+                                 || !strcmp(aKey, NS_APP_PROFILE_DIR_STARTUP))) {
     *aPersist = PR_TRUE;
     return EmbedPrivate::sProfileDir->Clone(aResult);
   }
 
   return NS_ERROR_FAILURE;
 }
 
 NS_IMETHODIMP
