--- mozilla/toolkit/xre/nsXREDirProvider.cpp.orig	2007-12-19 01:00:33.000000000 +0200
+++ mozilla/toolkit/xre/nsXREDirProvider.cpp	2007-12-19 01:49:54.000000000 +0200
@@ -98,17 +98,18 @@
 #endif
 
 #define PREF_OVERRIDE_DIRNAME "preferences"
 
 nsXREDirProvider* gDirServiceProvider = nsnull;
 
 nsXREDirProvider::nsXREDirProvider() :
   mProfileNotified(PR_FALSE),
-  mExtensionsLoaded(PR_FALSE)
+  mExtensionsLoaded(PR_FALSE),
+  mExtensionsIniTime(0)
 {
   gDirServiceProvider = this;
 }
 
 nsXREDirProvider::~nsXREDirProvider()
 {
   gDirServiceProvider = nsnull;
 }
@@ -534,49 +535,54 @@ LoadExtensionDirectories(nsINIParser &pa
     LoadPlatformDirectory(dir, aDirectories);
   }
   while (PR_TRUE);
 }
 
 void
 nsXREDirProvider::LoadBundleDirectories()
 {
-  if (mExtensionsLoaded)
-    return;
+  nsCOMPtr<nsIFile> extensionsINI;
+  if (mProfileDir && !gSafeMode) {
+    mProfileDir->Clone(getter_AddRefs(extensionsINI));
+    if (extensionsINI)
+      extensionsINI->AppendNative(NS_LITERAL_CSTRING("extensions.ini"));
+  }
+
+  if (mExtensionsLoaded) {
+    PRInt64 aLastModifiedTime = 0;
+    extensionsINI->GetLastModifiedTime(&aLastModifiedTime);
+    if (mExtensionsIniTime >= aLastModifiedTime)
+      return;
+  }
 
   mExtensionsLoaded = PR_TRUE;
 
   // first load distribution/bundles
   if (mXULAppDir) {
     LoadPlatformDirectory(mXULAppDir, mAppBundleDirectories);
 
     LoadAppBundleDirs();
   }
 
-  if (mProfileDir && !gSafeMode) {
-    nsCOMPtr<nsIFile> extensionsINI;
-    mProfileDir->Clone(getter_AddRefs(extensionsINI));
-    if (!extensionsINI)
-      return;
-
-    extensionsINI->AppendNative(NS_LITERAL_CSTRING("extensions.ini"));
+  if (!extensionsINI)
+    return;
 
-    nsCOMPtr<nsILocalFile> extensionsINILF =
-      do_QueryInterface(extensionsINI);
-    if (!extensionsINILF)
-      return;
+  nsCOMPtr<nsILocalFile> extensionsINILF =
+    do_QueryInterface(extensionsINI);
+  if (!extensionsINILF)
+    return;
 
-    nsINIParser parser;
-    nsresult rv = parser.Init(extensionsINILF);
-    if (NS_FAILED(rv))
-      return;
+  nsINIParser parser;
+  nsresult rv = parser.Init(extensionsINILF);
+  if (NS_FAILED(rv))
+    return;
 
-    LoadExtensionDirectories(parser, "ExtensionDirs", mExtensionDirectories);
-    LoadExtensionDirectories(parser, "ThemeDirs", mThemeDirectories);
-  }
+  LoadExtensionDirectories(parser, "ExtensionDirs", mExtensionDirectories);
+  LoadExtensionDirectories(parser, "ThemeDirs", mThemeDirectories);
 }
 
 void
 nsXREDirProvider::LoadAppBundleDirs()
 {
   if (!mXULAppDir)
     return;
 
--- mozilla/toolkit/xre/nsXREDirProvider.h.orig	2007-09-26 21:35:21.000000000 +0300
+++ mozilla/toolkit/xre/nsXREDirProvider.h	2007-12-19 01:40:07.000000000 +0200
@@ -136,14 +136,15 @@ protected:
 
   nsCOMPtr<nsIDirectoryServiceProvider> mAppProvider;
   nsCOMPtr<nsILocalFile> mGREDir;
   nsCOMPtr<nsIFile>      mXULAppDir;
   nsCOMPtr<nsIFile>      mProfileDir;
   nsCOMPtr<nsIFile>      mProfileLocalDir;
   PRPackedBool           mProfileNotified;
   PRPackedBool           mExtensionsLoaded;
+  PRTime                 mExtensionsIniTime;
   nsCOMArray<nsIFile>    mAppBundleDirectories;
   nsCOMArray<nsIFile>    mExtensionDirectories;
   nsCOMArray<nsIFile>    mThemeDirectories;
 };
 
 #endif
