#Bug 386862  Building GTK2 Linux embedding-profile=basic failed 
Index: mozilla/widget/src/gtk2/Makefile.in
===================================================================
--- mozilla.orig/widget/src/gtk2/Makefile.in
+++ mozilla/widget/src/gtk2/Makefile.in
@@ -103,20 +103,25 @@
 		nsSound.cpp \
 		nsNativeKeyBindings.cpp \
 		nsScreenGtk.cpp \
 		nsScreenManagerGtk.cpp \
 		nsDeviceContextSpecG.cpp \
 		nsPrintOptionsGTK.cpp \
 		nsImageToPixbuf.cpp \
 		nsAccessibilityHelper.cpp \
-		nsPrintJobFactoryGTK.cpp \
-		nsPrintJobGTK.cpp \
 		nsIdleServiceGTK.cpp \
-		$(NULL)
+	  $(NULL)
+
+ifdef MOZ_ENABLE_POSTSCRIPT
+CPPSRCS		+= \
+	  nsPrintJobFactoryGTK.cpp \
+	  nsPrintJobGTK.cpp \
+	  $(NULL)
+endif
 
 # build our subdirs, too
 ifdef ACCESSIBILITY
 REQUIRES	+= accessibility
 endif
 
 SHARED_LIBRARY_LIBS = ../xpwidgets/libxpwidgets_s.a
 
Index: mozilla/widget/src/gtk2/nsDeviceContextSpecG.cpp
===================================================================
--- mozilla.orig/widget/src/gtk2/nsDeviceContextSpecG.cpp
+++ mozilla/widget/src/gtk2/nsDeviceContextSpecG.cpp
@@ -407,18 +407,21 @@
   double width, height;
   mPrintSettings->GetEffectivePageSize(&width, &height);
   // convert twips to points
   width /= 20;
   height /= 20;
 
   DO_PR_DEBUG_LOG(("\"%s\", %f, %f\n", path, width, height));
 
-  nsresult rv = nsPrintJobFactoryGTK::CreatePrintJob(this, mPrintJob);
+  nsresult rv;
+#ifdef MOZ_ENABLE_POSTSCRIPT
+	rv = nsPrintJobFactoryGTK::CreatePrintJob(this, mPrintJob);
   if (NS_FAILED(rv))
+#endif
     return rv;
 
   nsCOMPtr<nsILocalFile> file;
   rv = mPrintJob->GetSpoolFile(getter_AddRefs(file));
   if (NS_FAILED(rv))
     return rv;
 
   nsCOMPtr<nsIFileOutputStream> stream = do_CreateInstance("@mozilla.org/network/file-output-stream;1");
