#Bug 386862  Building GTK2 Linux embedding-profile=basic failed 
Index: mozilla/widget/src/gtk2/Makefile.in
===================================================================
--- mozilla.orig/widget/src/gtk2/Makefile.in
+++ mozilla/widget/src/gtk2/Makefile.in
@@ -103,21 +103,26 @@
 		nsSound.cpp \
 		nsNativeKeyBindings.cpp \
 		nsScreenGtk.cpp \
 		nsScreenManagerGtk.cpp \
 		nsDeviceContextSpecG.cpp \
 		nsPrintOptionsGTK.cpp \
 		nsImageToPixbuf.cpp \
 		nsAccessibilityHelper.cpp \
-		nsPrintJobFactoryGTK.cpp \
-		nsPrintJobGTK.cpp \
 		nsIdleServiceGTK.cpp \
 		$(NULL)
 
+ifdef MOZ_ENABLE_POSTSCRIPT
+CPPSRCS		+= \
+	  nsPrintJobFactoryGTK.cpp \
+	  nsPrintJobGTK.cpp \
+		$(NULL)
+endif
+
 # build our subdirs, too
 ifdef ACCESSIBILITY
 REQUIRES	+= accessibility
 endif
 
 SHARED_LIBRARY_LIBS = ../xpwidgets/libxpwidgets_s.a
 
 EXTRA_DSO_LDOPTS += \
Index: mozilla/widget/src/gtk2/nsDeviceContextSpecG.cpp
===================================================================
--- mozilla.orig/widget/src/gtk2/nsDeviceContextSpecG.cpp
+++ mozilla/widget/src/gtk2/nsDeviceContextSpecG.cpp
@@ -407,17 +407,20 @@
   double width, height;
   mPrintSettings->GetEffectivePageSize(&width, &height);
   // convert twips to points
   width /= 20;
   height /= 20;
 
   DO_PR_DEBUG_LOG(("\"%s\", %f, %f\n", path, width, height));
 
-  nsresult rv = nsPrintJobFactoryGTK::CreatePrintJob(this, mPrintJob);
+  nsresult rv = NS_ERROR_UNEXPECTED;
+#if defined(USE_POSTSCRIPT)
+  rv = nsPrintJobFactoryGTK::CreatePrintJob(this, mPrintJob);
+#endif
   if (NS_FAILED(rv))
     return rv;
 
   nsCOMPtr<nsILocalFile> file;
   rv = mPrintJob->GetSpoolFile(getter_AddRefs(file));
   if (NS_FAILED(rv))
     return rv;
 
@@ -691,17 +694,21 @@
 
 NS_IMETHODIMP nsDeviceContextSpecGTK::BeginDocument(PRUnichar * aTitle, PRUnichar * aPrintToFileName, PRInt32 aStartPage, PRInt32 aEndPage)
 {
   return NS_OK;
 }
 
 NS_IMETHODIMP nsDeviceContextSpecGTK::EndDocument()
 {
+#if defined(USE_POSTSCRIPT)
   return mPrintJob->Submit();
+#else
+  return NS_ERROR_UNEXPECTED;
+#endif
 }
 
 /* Get prefs for printer
  * Search order:
  * - Get prefs per printer name and module name
  * - Get prefs per printer name
  * - Get prefs per module name
  * - Get prefs
