#Bug 386862  Building GTK2 Linux embedding-profile=basic failed 
Index: mozilla/widget/src/gtk2/Makefile.in
===================================================================
--- mozilla.orig/widget/src/gtk2/Makefile.in
+++ mozilla/widget/src/gtk2/Makefile.in
@@ -103,20 +103,25 @@
 		nsSound.cpp \
 		nsNativeKeyBindings.cpp \
 		nsScreenGtk.cpp \
 		nsScreenManagerGtk.cpp \
 		nsDeviceContextSpecG.cpp \
 		nsPrintOptionsGTK.cpp \
 		nsImageToPixbuf.cpp \
 		nsAccessibilityHelper.cpp \
-		nsPrintJobFactoryGTK.cpp \
-		nsPrintJobGTK.cpp \
 		nsIdleServiceGTK.cpp \
-		$(NULL)
+	  $(NULL)
+
+ifdef MOZ_ENABLE_POSTSCRIPT
+CPPSRCS		+= \
+	  nsPrintJobFactoryGTK.cpp \
+	  nsPrintJobGTK.cpp \
+	  $(NULL)
+endif
 
 # build our subdirs, too
 ifdef ACCESSIBILITY
 REQUIRES	+= accessibility
 endif
 
 SHARED_LIBRARY_LIBS = ../xpwidgets/libxpwidgets_s.a
 
Index: mozilla/widget/src/gtk2/nsDeviceContextSpecG.cpp
===================================================================
--- mozilla.orig/widget/src/gtk2/nsDeviceContextSpecG.cpp
+++ mozilla/widget/src/gtk2/nsDeviceContextSpecG.cpp
@@ -395,16 +395,17 @@
                    nsIDeviceContextSpec)
 
 //#define USE_PDF 1
 #include "gfxPDFSurface.h"
 #include "gfxPSSurface.h"
 #include "nsUnitConversion.h"
 NS_IMETHODIMP nsDeviceContextSpecGTK::GetSurfaceForPrinter(gfxASurface **aSurface)
 {
+#ifdef MOZ_ENABLE_POSTSCRIPT
   const char *path;
   GetPath(&path);
 
   double width, height;
   mPrintSettings->GetEffectivePageSize(&width, &height);
   // convert twips to points
   width /= 20;
   height /= 20;
@@ -431,16 +432,19 @@
   gfxPSSurface *surface = new gfxPSSurface(stream, gfxSize(width, height));
 #endif
 //  surface->SetDPI(600, 600);
   
   *aSurface = surface;
   NS_ADDREF(*aSurface);
 
   return NS_OK;
+#else
+  return NS_ERROR_FAILURE;
+#endif
 }
 
 /** -------------------------------------------------------
  *  Initialize the nsDeviceContextSpecGTK
  *  @update   dc 2/15/98
  *  @update   syd 3/2/99
  */
 NS_IMETHODIMP nsDeviceContextSpecGTK::Init(nsIWidget *aWidget,
