#https://bugzilla.mozilla.org/show_bug.cgi?id=532627
diff -r ba43b4ba7807 modules/plugin/base/src/nsPluginHost.cpp
--- a/modules/plugin/base/src/nsPluginHost.cpp	Mon Mar 30 22:14:17 2009 +0300
+++ b/modules/plugin/base/src/nsPluginHost.cpp	Mon Mar 30 23:09:51 2009 +0300
@@ -5336,8 +5336,10 @@ nsPluginHostImpl::WritePluginInfo()
     }
   }
 
-  if (fd)
+  if (fd) {
+    PR_Sync(fd);
     PR_Close(fd);
+  }
   return NS_OK;
 }
 
