--- /scratchbox/users/petejc/home/petejc/src/microb-engine/scratch_build/build-tree/mozilla/extensions/spatialnavigation/src/nsSpatialNavigation.cpp	2008-04-24 12:18:45.000000000 -0700
+++ mozilla/extensions/spatialnavigation/src/nsSpatialNavigation.cpp	2008-04-25 07:45:34.000000000 -0700
@@ -104,7 +104,27 @@
   } 
     else
   {
-    printf("<%s>\n", NS_ConvertUTF16toUTF8(str).get());
+    nsCOMPtr<nsIDOMHTMLInputElement> inputEl(do_QueryInterface(content));
+
+    // printf("inputEl (%p)\n", (void*) inputEl);
+
+    if (inputEl)
+    {
+      nsAutoString name, type, value;
+
+      inputEl->GetName(name);
+      inputEl->GetType(type);
+      inputEl->GetValue(value);
+
+      printf("<%s name=\"%s\" type=\"%s\" value=\"%s\">\n", NS_ConvertUTF16toUTF8(str).get(), 
+                                               NS_ConvertUTF16toUTF8(name).get(), 
+                                               NS_ConvertUTF16toUTF8(type).get(),
+                                               NS_ConvertUTF16toUTF8(value).get());
+    }
+      else
+    {
+      printf("<%s>\n", NS_ConvertUTF16toUTF8(str).get());
+    }
   }
 
 /*
--- /scratchbox/users/petejc/home/petejc/src/microb-engine/scratch_build/build-tree/mozilla/extensions/spatialnavigation/src/nsSpatialNavigationUtils.cpp	2008-04-15 09:12:21.000000000 -0700
+++ mozilla/extensions/spatialnavigation/src/nsSpatialNavigationUtils.cpp	2008-04-25 07:41:24.000000000 -0700
@@ -441,7 +441,19 @@
   nsAutoString inputType;
   nsCOMPtr<nsIDOMHTMLInputElement> inputElement = do_QueryInterface(currentContent);
   if (inputElement && NS_SUCCEEDED(inputElement->GetDisabled(&disabled)) && NS_SUCCEEDED(inputElement->GetType(inputType)))
-    return !disabled && (! inputType.LowerCaseEqualsLiteral("hidden")) && (!inputType.LowerCaseEqualsLiteral("button"));
+  {
+    if (inputType.LowerCaseEqualsLiteral("button"))
+    {
+      nsCOMPtr<nsIDOMNode> pn;
+      inputElement->GetParentNode(getter_AddRefs(pn));
+
+      nsCOMPtr<nsIDOMHTMLSelectElement> selectElement = do_QueryInterface(pn);
+
+      if (selectElement) return PR_FALSE;
+    }
+
+    return !disabled && (! inputType.LowerCaseEqualsLiteral("hidden")) && (!inputType.LowerCaseEqualsLiteral("file"));
+  }
 
   nsAutoString textareaType;
   nsCOMPtr<nsIDOMHTMLTextAreaElement> textareaElement = do_QueryInterface(currentContent);

