Index: mozilla/extensions/spatialnavigation/src/nsSpatialNavigation.cpp
===================================================================
--- mozilla.orig/extensions/spatialnavigation/src/nsSpatialNavigation.cpp
+++ mozilla/extensions/spatialnavigation/src/nsSpatialNavigation.cpp
@@ -671,28 +671,27 @@
     getContentInDirection(direction, presContext, focusedRect, focusedFrame, PR_FALSE, isAREA, getter_AddRefs(c));
   
   if (c) {
    
     nsIDocument* doc = c->GetDocument();
     if (!doc)
       return NS_ERROR_FAILURE;
     
-/*    nsIPresShell *presShell = doc->GetPrimaryShell();
+    nsIPresShell *presShell = doc->GetPrimaryShell();
 
     nsIFrame* cframe = presShell->GetPrimaryFrameFor(c);
     
     PRBool b = IsPartiallyVisible(presShell, cframe); 
     
     if (b)
       setFocusedContent(c);
     else
-      ScrollWindow(direction, getContentWindow());*/
+      ScrollWindow(direction, getContentWindow());
 
-    setFocusedContent(c);
     return NS_OK;
   }
 
   ///////////////////////////////////////////////////////////////////////////////////////////////////
   // do it all again at the parent document
   ///////////////////////////////////////////////////////////////////////////////////////////////////
 
   {
@@ -800,23 +799,23 @@
     NS_IF_ADDREF(*aContent = content);
     return;
   }
 
   contentWindow = getContentWindow();
   DisableJSScope foopy (contentWindow);
 
   //xxxx should/can we prevent it from going into chrome???
-  if (direction == eNavLeft || direction == eNavUp)
+//  if (direction == eNavLeft || direction == eNavUp)
     focusController->MoveFocus(PR_TRUE, nsnull);
-  else
-    focusController->MoveFocus(PR_FALSE, nsnull);
+//  else
+//    focusController->MoveFocus(PR_FALSE, nsnull);
   
   // so there is no focused content -- lets make some up, hightlight it and return.  
-  focusController->GetFocusedElement(getter_AddRefs(element));
+//  focusController->GetFocusedElement(getter_AddRefs(element));
 }
 
 void 
 nsSpatialNavigation::setFocusedContent(nsIContent* c)
 {
   if (!c)
     return;
 
