#55568 Accessing already freed memory
Index: mozilla/extensions/spatialnavigation/src/nsSpatialNavigationService.cpp
===================================================================
--- mozilla.orig/extensions/spatialnavigation/src/nsSpatialNavigationService.cpp
+++ mozilla/extensions/spatialnavigation/src/nsSpatialNavigationService.cpp
@@ -169,61 +169,61 @@
       mKeyCodeModifier = tempInt32;
     
     return NS_OK;
   }
   
   if (!strcmp(aTopic, NS_PREFBRANCH_PREFCHANGE_TOPIC_ID)) 
   {
     nsCOMPtr<nsIPrefBranch> prefBranch = do_QueryInterface(aSubject);
-    nsXPIDLCString cstr;
     
-    const char* pref = NS_ConvertUTF16toUTF8(aData).get();
+    nsCString prefc(NS_ConvertUTF16toUTF8(aData).get());
+    nsString prefs(aData);
 
-    if (!strcmp(pref, "snav.enabled"))
+    if (prefs.EqualsLiteral("snav.enabled"))
     {
-      prefBranch->GetBoolPref(pref, &mEnabled);
+      prefBranch->GetBoolPref(prefc.get(), &mEnabled);
     }
-    else if (!strcmp(pref, "snav.ignoreTextFields"))
+    else if (prefs.EqualsLiteral("snav.ignoreTextFields"))
     {
-      prefBranch->GetBoolPref(pref, &mIgnoreTextFields);
+      prefBranch->GetBoolPref(prefc.get(), &mIgnoreTextFields);
     }
-    else if (!strcmp(pref, "snav.directionalBias"))
+    else if (prefs.EqualsLiteral("snav.directionalBias"))
     {
-      prefBranch->GetIntPref(pref, &gDirectionalBias);
+      prefBranch->GetIntPref(prefc.get(), &gDirectionalBias);
       if (gDirectionalBias == 0)
         gDirectionalBias = 1;
     }
-    else if (!strcmp(pref, "snav.disableJS"))
+    else if (prefs.EqualsLiteral("snav.disableJS"))
     {
-      prefBranch->GetBoolPref(pref, &mDisableJSWhenFocusing);
+      prefBranch->GetBoolPref(prefc.get(), &mDisableJSWhenFocusing);
     }
-    else if (!strcmp(pref, "snav.rectFudge"))
+    else if (prefs.EqualsLiteral("snav.rectFudge"))
     {
-      prefBranch->GetIntPref(pref, &gRectFudge);
+      prefBranch->GetIntPref(prefc.get(), &gRectFudge);
     }
-    else if (!strcmp(pref, "snav.keyCode.left"))
+    else if (prefs.EqualsLiteral("snav.keyCode.left"))
     {
-      prefBranch->GetIntPref(pref, &mKeyCodeLeft);
+      prefBranch->GetIntPref(prefc.get(), &mKeyCodeLeft);
     }
-    else if (!strcmp(pref, "snav.keyCode.right"))
+    else if (prefs.EqualsLiteral("snav.keyCode.right"))
     {
-      prefBranch->GetIntPref(pref, &mKeyCodeRight);
+      prefBranch->GetIntPref(prefc.get(), &mKeyCodeRight);
     }
-    else if (!strcmp(pref, "snav.keyCode.up"))
+    else if (prefs.EqualsLiteral("snav.keyCode.up"))
     {
-      prefBranch->GetIntPref(pref, &mKeyCodeUp);
+      prefBranch->GetIntPref(prefc.get(), &mKeyCodeUp);
     }    
-    else if (!strcmp(pref, "snav.keyCode.down"))
+    else if (prefs.EqualsLiteral("snav.keyCode.down"))
     {
-      prefBranch->GetIntPref(pref, &mKeyCodeDown);
+      prefBranch->GetIntPref(prefc.get(), &mKeyCodeDown);
     }    
-    else if (!strcmp(pref, "snav.keyCode.modifier"))
+    else if (prefs.EqualsLiteral("snav.keyCode.modifier"))
     {
-      prefBranch->GetIntPref(pref, &mKeyCodeModifier);
+      prefBranch->GetIntPref(prefc.get(), &mKeyCodeModifier);
     }
     
     return NS_OK;
   }
   
   return NS_OK;
 }
 
