--- mozilla/extensions/spatialnavigation/src/nsSpatialNavigation.cpp.orig	2007-09-25 18:03:48.000000000 +0300
+++ mozilla/extensions/spatialnavigation/src/nsSpatialNavigation.cpp	2007-09-26 18:34:36.000000000 +0300
@@ -244,20 +244,32 @@ nsSpatialNavigation::HandleKey(nsIDOMEve
   PRUint32 keyCode;
   PRBool isModifier;
   nsCOMPtr<nsIDOMKeyEvent> keyEvent(do_QueryInterface(aEvent));
   
   if (!keyEvent)
     return NS_ERROR_FAILURE;
 
   if (NS_FAILED(keyEvent->GetKeyCode(&keyCode)))
 	return NS_ERROR_FAILURE;
   
+  if (keyCode == nsIDOMKeyEvent::DOM_VK_RETURN && !aReal) {
+    nsCOMPtr<nsIContent> focusedContent;
+    getFocusedContent(0, getter_AddRefs(focusedContent));
+    if (focusedContent && !isContentInsideWindow(focusedContent, mTopWindow)) {
+      nsCOMPtr<nsIDOMNSHTMLElement> nsElement = do_QueryInterface(focusedContent);
+      if (nsElement) {
+        nsElement->Blur();
+        return NS_OK;
+      }
+    }
+  }
+  
   // figure out what modifier to use  
   
   /************************************************
     Value of the keyCodeModifier is
   
     SHIFT          = 0x00100000
     CONTROL        = 0x00001100
     ALT            = 0x00000012
   *************************************************/
   
