diff -ruN -p -U10 mozilla/extensions/spatialnavigation/src.orig/nsSpatialNavigation.cpp mozilla/extensions/spatialnavigation/src/nsSpatialNavigation.cpp
--- mozilla/extensions/spatialnavigation/src.orig/nsSpatialNavigation.cpp	2007-08-16 17:22:16.000000000 +0300
+++ mozilla/extensions/spatialnavigation/src/nsSpatialNavigation.cpp	2007-08-16 17:28:00.000000000 +0300
@@ -115,20 +115,33 @@ nsSpatialNavigation::KeyDown(nsIDOMEvent
   if (!nsEvent)
     return NS_ERROR_FAILURE;
   
   nsCOMPtr<nsIDOMEventTarget> domEventTarget;
   nsEvent->GetOriginalTarget(getter_AddRefs(domEventTarget));
   
   nsCOMPtr<nsIContent> targetContent = do_QueryInterface(domEventTarget);
 
   if (targetContent->IsNodeOfType(nsINode::eXUL)) 
     return NS_OK;
+
+  PRBool designMode = PR_FALSE;
+  nsCOMPtr<nsIDOMNode> eventNode = do_QueryInterface(domEventTarget);
+  nsCOMPtr<nsIDOMDocument> domDoc;
+  if (eventNode)
+    eventNode->GetOwnerDocument(getter_AddRefs(domDoc));
+  nsCOMPtr<nsIDOMNSHTMLDocument> htmlDoc = do_QueryInterface(domDoc);
+  if (htmlDoc) {
+    nsString DMode;
+    htmlDoc->GetDesignMode(DMode);
+    designMode = DMode.EqualsLiteral("on");
+    return NS_OK;
+  }
   
   if (targetContent->IsNodeOfType(nsINode::eHTML_FORM_CONTROL)) 
   {
       nsCOMPtr<nsIFormControl> formControl(do_QueryInterface(targetContent));
       formControlType = formControl->GetType();
       
       if (mService->mIgnoreTextFields)
       {
         if (formControlType == NS_FORM_TEXTAREA ||
             formControlType == NS_FORM_INPUT_TEXT ||
diff -ruN -p -U10 mozilla/extensions/spatialnavigation/src.orig/nsSpatialNavigationPrivate.h mozilla/extensions/spatialnavigation/src/nsSpatialNavigationPrivate.h
--- mozilla/extensions/spatialnavigation/src.orig/nsSpatialNavigationPrivate.h	2007-08-16 17:22:16.000000000 +0300
+++ mozilla/extensions/spatialnavigation/src/nsSpatialNavigationPrivate.h	2007-08-16 17:23:39.000000000 +0300
@@ -80,20 +80,21 @@
 #include "nsIDOMHTMLMapElement.h"
 #include "nsIDOMHTMLOptionElement.h"
 #include "nsIDOMHTMLSelectElement.h"
 #include "nsIDOMKeyEvent.h"
 #include "nsIDOMKeyListener.h"
 #include "nsIDOMNSEvent.h"
 #include "nsIDOMNSHTMLElement.h"
 #include "nsIDOMNSUIEvent.h"
 #include "nsIDOMNSHTMLInputElement.h"
 #include "nsIDOMNSHTMLTextAreaElement.h"
+#include "nsIDOMNSHTMLDocument.h"
 #include "nsIDOMNode.h"
 #include "nsIDOMWindow.h"
 #include "nsIDOMWindowInternal.h"
 #include "nsIDocShell.h"
 #include "nsIDocShellTreeItem.h"
 #include "nsIDocShellTreeOwner.h"
 #include "nsIDocument.h"
 #include "nsIDocumentViewer.h"
 #include "nsIEventStateManager.h"
 #include "nsIFile.h"
