Index: jsinterp.c
===================================================================
RCS file: /cvsroot/mozilla/js/src/jsinterp.c,v
retrieving revision 3.375
diff -p -u -8 -r3.375 jsinterp.c
--- mozilla/js/src/jsinterp.c	19 Sep 2007 21:10:58 -0000	3.375
+++ mozilla/js/src/jsinterp.c	20 Sep 2007 21:37:44 -0000
@@ -3058,18 +3058,18 @@ interrupt:
             if (!ok)
                 goto out;
             sp--;
             STORE_OPND(-1, rval);
           END_CASE(JSOP_SETNAME)
 
 #define INTEGER_OP(OP, EXTRA_CODE)                                            \
     JS_BEGIN_MACRO                                                            \
-        FETCH_INT(cx, -1, j);                                                 \
         FETCH_INT(cx, -2, i);                                                 \
+        FETCH_INT(cx, -1, j);                                                 \
         EXTRA_CODE                                                            \
         i = i OP j;                                                           \
         sp--;                                                                 \
         STORE_INT(cx, -1, i);                                                 \
     JS_END_MACRO
 
 #define BITWISE_OP(OP)          INTEGER_OP(OP, (void) 0;)
 #define SIGNED_SHIFT_OP(OP)     INTEGER_OP(OP, j &= 31;)
@@ -3289,18 +3289,18 @@ interrupt:
           BEGIN_CASE(JSOP_RSH)
             SIGNED_SHIFT_OP(>>);
           END_CASE(JSOP_RSH)
 
           BEGIN_CASE(JSOP_URSH)
           {
             uint32 u;
 
-            FETCH_INT(cx, -1, j);
             FETCH_UINT(cx, -2, u);
+            FETCH_INT(cx, -1, j);
             u >>= j & 31;
             sp--;
             STORE_UINT(cx, -1, u);
           }
           END_CASE(JSOP_URSH)
 
 #undef INTEGER_OP
 #undef BITWISE_OP
