Index: content/base/src/nsGenericElement.cpp
===================================================================
RCS file: /cvsroot/mozilla/content/base/src/nsGenericElement.cpp,v
retrieving revision 3.582
diff -u -p -d -1 -0 -r3.582 nsGenericElement.cpp
--- mozilla/content/base/src/nsGenericElement.cpp	12 Jul 2007 20:05:45 -0000	3.582
+++ mozilla/content/base/src/nsGenericElement.cpp	26 Jul 2007 20:04:44 -0000
@@ -2074,21 +2074,22 @@ nsGenericElement::BindToTree(nsIDocument
   return NS_OK;
 }
 
 void
 nsGenericElement::UnbindFromTree(PRBool aDeep, PRBool aNullParent)
 {
   NS_PRECONDITION(aDeep || (!GetCurrentDoc() && !GetBindingParent()),
                   "Shallow unbind won't clear document and binding parent on "
                   "kids!");
   // Make sure to unbind this node before doing the kids
-  nsIDocument *document = GetCurrentDoc();
+  nsIDocument *document =
+    HasFlag(NODE_FORCE_XBL_BINDINGS) ? GetOwnerDoc() : GetCurrentDoc();
   if (document) {
     // Notify XBL- & nsIAnonymousContentCreator-generated
     // anonymous content that the document is changing.
     document->BindingManager()->ChangeDocumentFor(this, document, nsnull);
 
     if (HasAttr(kNameSpaceID_XLink, nsGkAtoms::href)) {
       document->ForgetLink(this);
     }
 
     document->ClearBoxObjectFor(this);
