# Add support for some rss types
# works XMLPrettyPrint hacks
Index: mozilla/layout/build/nsLayoutModule.cpp
===================================================================
--- mozilla.orig/layout/build/nsLayoutModule.cpp
+++ mozilla/layout/build/nsLayoutModule.cpp
@@ -1030,16 +1030,26 @@ static const nsModuleComponentInfo gComp
   { "Canvas 2D Rendering Context",
     NS_CANVASRENDERINGCONTEXT2D_CID,
     "@mozilla.org/content/canvas-rendering-context;1?id=2d",
     CreateCanvasRenderingContext2D },
 #endif
 
   { "XML document encoder",
     NS_TEXT_ENCODER_CID,
+    NS_DOC_ENCODER_CONTRACTID_BASE "application/atom+xml",
+    CreateTextEncoder },
+
+  { "XML document encoder",
+    NS_TEXT_ENCODER_CID,
+    NS_DOC_ENCODER_CONTRACTID_BASE "application/rss+xml",
+    CreateTextEncoder },
+
+  { "XML document encoder",
+    NS_TEXT_ENCODER_CID,
     NS_DOC_ENCODER_CONTRACTID_BASE "text/xml",
     CreateTextEncoder },
 
   { "XML document encoder",
     NS_TEXT_ENCODER_CID,
     NS_DOC_ENCODER_CONTRACTID_BASE "application/xml",
     CreateTextEncoder },
 
@@ -1067,16 +1077,26 @@ static const nsModuleComponentInfo gComp
 
   { "HTML copy encoder",
     NS_HTMLCOPY_TEXT_ENCODER_CID,
     NS_HTMLCOPY_ENCODER_CONTRACTID,
     CreateHTMLCopyTextEncoder },
 
   { "XML content serializer",
     NS_XMLCONTENTSERIALIZER_CID,
+    NS_CONTENTSERIALIZER_CONTRACTID_PREFIX "application/atom+xml",
+    CreateXMLContentSerializer },
+
+  { "XML content serializer",
+    NS_XMLCONTENTSERIALIZER_CID,
+    NS_CONTENTSERIALIZER_CONTRACTID_PREFIX "application/rss+xml",
+    CreateXMLContentSerializer },
+
+  { "XML content serializer",
+    NS_XMLCONTENTSERIALIZER_CID,
     NS_CONTENTSERIALIZER_CONTRACTID_PREFIX "text/xml",
     CreateXMLContentSerializer },
 
   { "XML content serializer",
     NS_XMLCONTENTSERIALIZER_CID,
     NS_CONTENTSERIALIZER_CONTRACTID_PREFIX "application/xml",
     CreateXMLContentSerializer },
 
Index: mozilla/layout/build/nsContentDLF.cpp
===================================================================
--- mozilla.orig/layout/build/nsContentDLF.cpp
+++ mozilla/layout/build/nsContentDLF.cpp
@@ -103,16 +103,18 @@ static const char* const gHTMLTypes[] = 
 #endif
   "application/xhtml+xml",
   0
 };
   
 static const char* const gXMLTypes[] = {
   "text/xml",
   "application/xml",
+  "application/atom+xml",
+  "application/rss+xml",
   "application/rdf+xml",
   "text/rdf",
   0
 };
 
 #ifdef MOZ_SVG
 static const char* const gSVGTypes[] = {
   "image/svg+xml",
