# Add support for some rss types
# works XMLPrettyPrint hacks
--- mozilla/layout/build/nsLayoutModule.cpp.orig	2007-04-26 02:00:44.000000000 +0300
+++ mozilla/layout/build/nsLayoutModule.cpp	2007-04-26 16:33:43.799281250 +0300
@@ -1044,20 +1044,30 @@ static const nsModuleComponentInfo gComp
     NS_DOC_ENCODER_CONTRACTID_BASE "text/xml",
     CreateTextEncoder },
 
   { "XML document encoder",
     NS_TEXT_ENCODER_CID,
     NS_DOC_ENCODER_CONTRACTID_BASE "application/xml",
     CreateTextEncoder },
 
   { "XML document encoder",
     NS_TEXT_ENCODER_CID,
+    NS_DOC_ENCODER_CONTRACTID_BASE "application/atom+xml",
+    CreateTextEncoder },
+
+  { "XML document encoder",
+    NS_TEXT_ENCODER_CID,
+    NS_DOC_ENCODER_CONTRACTID_BASE "application/rss+xml",
+    CreateTextEncoder },
+
+  { "XML document encoder",
+    NS_TEXT_ENCODER_CID,
     NS_DOC_ENCODER_CONTRACTID_BASE "application/xhtml+xml",
     CreateTextEncoder },
 
 #ifdef MOZ_SVG
   { "SVG document encoder",
     NS_TEXT_ENCODER_CID,
     NS_DOC_ENCODER_CONTRACTID_BASE "image/svg+xml",
     CreateTextEncoder },
 #endif
 
@@ -1081,20 +1091,30 @@ static const nsModuleComponentInfo gComp
     NS_CONTENTSERIALIZER_CONTRACTID_PREFIX "text/xml",
     CreateXMLContentSerializer },
 
   { "XML content serializer",
     NS_XMLCONTENTSERIALIZER_CID,
     NS_CONTENTSERIALIZER_CONTRACTID_PREFIX "application/xml",
     CreateXMLContentSerializer },
 
   { "XML content serializer",
     NS_XMLCONTENTSERIALIZER_CID,
+    NS_CONTENTSERIALIZER_CONTRACTID_PREFIX "application/atom+xml",
+    CreateXMLContentSerializer },
+
+  { "XML content serializer",
+    NS_XMLCONTENTSERIALIZER_CID,
+    NS_CONTENTSERIALIZER_CONTRACTID_PREFIX "application/rss+xml",
+    CreateXMLContentSerializer },
+
+  { "XML content serializer",
+    NS_XMLCONTENTSERIALIZER_CID,
     NS_CONTENTSERIALIZER_CONTRACTID_PREFIX "application/xhtml+xml",
     CreateXMLContentSerializer },
 
 #ifdef MOZ_SVG
   { "SVG content serializer",
     NS_XMLCONTENTSERIALIZER_CID,
     NS_CONTENTSERIALIZER_CONTRACTID_PREFIX "image/svg+xml",
     CreateXMLContentSerializer },
 #endif
 
--- mozilla/layout/build/nsContentDLF.cpp.orig	2007-04-26 01:59:47.000000000 +0300
+++ mozilla/layout/build/nsContentDLF.cpp	2007-04-26 16:32:36.747090750 +0300
@@ -97,20 +97,22 @@ static const char* const gHTMLTypes[] = 
 #ifdef MOZ_VIEW_SOURCE
   "application/x-view-source", //XXX I wish I could just use nsMimeTypes.h here
 #endif
   "application/xhtml+xml",
   0
 };
   
 static const char* const gXMLTypes[] = {
   "text/xml",
   "application/xml",
+  "application/atom+xml",
+  "application/rss+xml",
   "application/rdf+xml",
   "text/rdf",
   0
 };
 
 #ifdef MOZ_SVG
 static const char* const gSVGTypes[] = {
   "image/svg+xml",
   0
 };
