#56960
#[crash] Browser crashed during page loading and scrolling down
--- mozilla/layout/generic/nsLineBox.cpp.orig	2007-05-24 16:59:58.000000000 +0300
+++ mozilla/layout/generic/nsLineBox.cpp	2007-05-24 17:15:09.000000000 +0300
@@ -264,20 +264,21 @@ nsLineBox::IsLastChild(nsIFrame* aFrame)
 
 PRInt32
 nsLineBox::IndexOf(nsIFrame* aFrame) const
 {
   PRInt32 i, n = GetChildCount();
   nsIFrame* frame = mFirstChild;
   for (i = 0; i < n; i++) {
     if (frame == aFrame) {
       return i;
     }
+    if (!frame) return -1;
     frame = frame->GetNextSibling();
   }
   return -1;
 }
 
 PRBool
 nsLineBox::ContainsAfter(nsIFrame* aFrameInLine,
                          nsIFrame* aFrameToFind,
                          nsLineList::iterator aLineIter,
                          const nsLineList::iterator& aEndLines) const
