# Added check for string to avoid crash situation
--- mozilla/js/src/jsopcode.c.orig	2007-09-24 14:54:19.000000000 +0300
+++ mozilla/js/src/jsopcode.c	2007-09-24 16:03:32.000000000 +0300
@@ -1694,21 +1694,23 @@ Decompile(SprintStack *ss, jsbytecode *p
                      JSSTRING_LENGTH(ATOM_TO_STRING(atom))) != TOK_EOF)
 
 /*
  * Given an atom already fetched from jp->script's atom map, quote/escape its
  * string appropriately into rval, and select fmt from the quoted and unquoted
  * alternatives.
  */
 #define GET_QUOTE_AND_FMT(qfmt, ufmt, rval)                                   \
     JS_BEGIN_MACRO                                                            \
         jschar quote_;                                                        \
-        if (!ATOM_IS_IDENTIFIER(atom)) {                                      \
+        JSString *str = ATOM_TO_STRING(atom);                                 \
+        if (!str) return NULL;                                                \
+        if (!js_IsIdentifier(str)) {                                          \
             quote_ = '\'';                                                    \
             fmt = qfmt;                                                       \
         } else {                                                              \
             quote_ = 0;                                                       \
             fmt = ufmt;                                                       \
         }                                                                     \
         rval = QuoteString(&ss->sprinter, ATOM_TO_STRING(atom), quote_);      \
         if (!rval)                                                            \
             return NULL;                                                      \
     JS_END_MACRO
--- mozilla/js/src/jsstr.c.orig	2007-09-24 11:52:07.000000000 +0300
+++ mozilla/js/src/jsstr.c	2007-09-24 16:20:48.000000000 +0300
@@ -2657,21 +2657,21 @@ js_ValueToPrintable(JSContext *cx, jsval
 }
 
 JS_FRIEND_API(JSString *)
 js_ValueToString(JSContext *cx, jsval v)
 {
     JSObject *obj;
     JSString *str;
 
     if (JSVAL_IS_OBJECT(v)) {
         obj = JSVAL_TO_OBJECT(v);
-        if (!obj)
+        if (!obj || !(obj)->map->ops)
             return ATOM_TO_STRING(cx->runtime->atomState.nullAtom);
         if (!OBJ_DEFAULT_VALUE(cx, obj, JSTYPE_STRING, &v))
             return NULL;
     }
     if (JSVAL_IS_STRING(v)) {
         str = JSVAL_TO_STRING(v);
     } else if (JSVAL_IS_INT(v)) {
         str = js_NumberToString(cx, JSVAL_TO_INT(v));
     } else if (JSVAL_IS_DOUBLE(v)) {
         str = js_NumberToString(cx, *JSVAL_TO_DOUBLE(v));
