--- mozilla/modules/libimg/png/pngpread.c 2007/04/12 18:34:543.14
+++ mozilla/modules/libimg/png/pngpread.c 2007/04/12 20:11:503.15
@@ -1732,11 +1732,11 @@
 png_progressive_combine_row (png_structp png_ptr,
    png_bytep old_row, png_bytep new_row)
 {
-   if(png_ptr == NULL) return;
 #ifdef PNG_USE_LOCAL_ARRAYS
    const int FARDATA png_pass_dsp_mask[7] =
       {0xff, 0x0f, 0xff, 0x33, 0xff, 0x55, 0xff};
 #endif
+   if(png_ptr == NULL) return;
    if (new_row != NULL)    /* new_row must == png_ptr->row_buf here. */
       png_combine_row(png_ptr, old_row, png_pass_dsp_mask[png_ptr->pass]);
 }