Index: mozilla/modules/libpr0n/decoders/jpeg/nsJPEGDecoder.cpp
===================================================================
--- mozilla.orig/modules/libpr0n/decoders/jpeg/nsJPEGDecoder.cpp
+++ mozilla/modules/libpr0n/decoders/jpeg/nsJPEGDecoder.cpp
@@ -44,17 +44,19 @@
 #include "nsIComponentManager.h"
 #include "nsIInputStream.h"
 
 #include "nspr.h"
 #include "nsCRT.h"
 #include "ImageLogging.h"
 #include "nsIImage.h"
 #include "nsIInterfaceRequestorUtils.h"
+#ifdef MOZ_CAIRO_GFX
 #include "gfxColor.h"
+#endif
 
 #include "jerror.h"
 
 NS_IMPL_ISUPPORTS1(nsJPEGDecoder, imgIDecoder)
 
 #if defined(PR_LOGGING)
 PRLogModuleInfo *gJPEGlog = PR_NewLogModule("JPEGDecoder");
 #else
