# Bug 376446  spurious white pixels in animated gifs
Index: modules/libpr0n/src/imgContainer.cpp
===================================================================
RCS file: /cvsroot/mozilla/modules/libpr0n/src/imgContainer.cpp,v
retrieving revision 1.43
diff -u -8 -p -r1.43 imgContainer.cpp
--- modules/libpr0n/src/imgContainer.cpp	20 Mar 2007 23:56:49 -0000	1.43
+++ modules/libpr0n/src/imgContainer.cpp	1 May 2007 14:30:16 -0000
@@ -438,18 +438,17 @@ NS_IMETHODIMP imgContainer::Notify(nsITi
   PRInt32 nextFrameIndex = mAnim->currentAnimationFrameIndex + 1;
   PRInt32 timeout = 0;
 
   // If we're done decoding the next frame, go ahead and display it now and
   // reinit the timer with the next frame's delay time.
   // currentDecodingFrameIndex is not set until the second frame has
   // finished decoding (see EndFrameDecode)
   if (mAnim->doneDecoding || 
-      (numFrames == 2 && nextFrameIndex < 2) ||
-      (numFrames > 2 && nextFrameIndex < mAnim->currentDecodingFrameIndex)) {
+      (nextFrameIndex < mAnim->currentDecodingFrameIndex)) {
     if (numFrames == nextFrameIndex) {
       // End of Animation
 
       // If animation mode is "loop once", it's time to stop animating
       if (mAnimationMode == kLoopOnceAnimMode || mLoopCount == 0) {
         StopAnimation();
         return NS_OK;
       } else {
@@ -467,18 +466,17 @@ NS_IMETHODIMP imgContainer::Notify(nsITi
     if (!(nextFrame = mFrames[nextFrameIndex])) {
       // something wrong with the next frame, skip it
       mAnim->currentAnimationFrameIndex = nextFrameIndex;
       mAnim->timer->SetDelay(100);
       return NS_OK;
     }
     nextFrame->GetTimeout(&timeout);
 
-  } else if ((numFrames == 2 && nextFrameIndex == 2) ||
-             (numFrames > 2 && nextFrameIndex == mAnim->currentDecodingFrameIndex)) {
+  } else if (nextFrameIndex == mAnim->currentDecodingFrameIndex) {
     // Uh oh, the frame we want to show is currently being decoded (partial)
     // Wait a bit and try again
     mAnim->timer->SetDelay(100);
     return NS_OK;
   } else { //  (nextFrameIndex > currentDecodingFrameIndex)
     // We shouldn't get here. However, if we are requesting a frame
     // that hasn't been decoded yet, go back to the last frame decoded
     NS_WARNING("imgContainer::Notify()  Frame is passed decoded frame");
