diff -r fa3561adc02e toolkit/mozapps/extensions/content/extensions.js
--- a/toolkit/mozapps/extensions/content/extensions.js	Fri Aug 07 13:48:19 2009 +0300
+++ b/toolkit/mozapps/extensions/content/extensions.js	Fri Aug 07 13:50:48 2009 +0300
@@ -2009,16 +2009,17 @@ const gAddonsMsgObserver = {
   observe: function (aSubject, aTopic, aData)
   {
     switch (aData) {
     case "addons-enable-xpinstall":
       gPref.setBoolPref("xpinstall.enabled", true);
       break;
     case "addons-enable-compatibility":
       gPref.clearUserPref(PREF_EM_CHECK_COMPATIBILITY);
+      gPref.savePrefFile(null);
       gCheckCompat = true;
       break;
     case "addons-enable-updatesecurity":
       gPref.clearUserPref(PREF_EM_CHECK_UPDATE_SECURITY);
       gCheckUpdateSecurity = true;
       break;
     case "addons-no-updates":
       var children = gExtensionsView.children;
