diff -ruN -p -U 8 mozilla.orig/configure.in mozilla/configure.in
--- mozilla.orig/configure.in	2010-02-15 16:50:46.000000000 +0200
+++ mozilla/configure.in	2010-02-15 16:51:07.000000000 +0200
@@ -6436,16 +6436,24 @@ if test "$X11_COMPOSITED_PLUGINS" = "yes
     dnl if we have Xcomposite we should also have Xdamage and Xfixes
     AC_CHECK_HEADERS([X11/extensions/Xdamage.h], [],
                      [AC_MSG_ERROR([Couldn't find X11/extentsions/Xdamage.h which is required for composited plugins.])])
     AC_CHECK_LIB(Xcomposite, XCompositeRedirectWindow, [XCOMPOSITE_LIBS="-lXcomposite -lXdamage -lXfixes"],
                  [MISSING_X="$MISSING_X -lXcomposite"], $XLIBS)
 fi
 AC_SUBST(XCOMPOSITE_LIBS)
 
+MOZ_ARG_WITHOUT_BOOL(arm-thumb2,
+                     [  --without-arm-thumb2        don't use arm thumb2 instructions)],
+                         AVOID_ARM_THUMB2="yes")
+if test "$AVOID_ARM_THUMB2" = "yes"; then
+    NJ_FORCE_NO_ARM_THUMB2=1
+    AC_DEFINE(NJ_FORCE_NO_ARM_THUMB2)
+fi
+
 PKG_CHECK_MODULES(LIBOSSO,libosso,
                   NS_OSSO=1,
                   NS_OSSO=)
 
 if test $NS_OSSO; then
     if test -z "$MOZ_ENABLE_DBUS"; then
         AC_MSG_ERROR([DBus is required when building for OSSO])
     fi
