#https://bugzilla.mozilla.org/show_bug.cgi?id=532609
Index: mozilla/xpcom/build/Makefile.in
===================================================================
--- mozilla.orig/xpcom/build/Makefile.in
+++ mozilla/xpcom/build/Makefile.in
@@ -188,16 +188,17 @@ endif
 ifeq ($(OS_ARCH),BeOS)
 EXTRA_DSO_LDOPTS += -lbe
 endif
 
 ifneq (,$(filter gtk2,$(MOZ_WIDGET_TOOLKIT)))
 EXTRA_DSO_LDOPTS += $(MOZ_GTK2_LIBS)
 ifdef MOZ_PLATFORM_MAEMO
 EXTRA_DSO_LDOPTS += $(MOZ_PLATFORM_MAEMO_LIBS)
+EXTRA_DSO_LDOPTS += $(LIBHILDONMIME_LIBS)
 endif
 endif
 
 ifeq ($(OS_ARCH),WINNT)
 EXTRA_DSO_LDOPTS += $(call EXPAND_LIBNAME,shell32 ole32 uuid version)
 ifneq (,$(MOZ_DEBUG)$(NS_TRACE_MALLOC))
 EXTRA_DSO_LDOPTS += $(call EXPAND_LIBNAME,imagehlp)
 endif
