#Make suspectCurrent assertion not fire constantly after we Fault.  b=386664
diff --git a/xpcom/base/nsCycleCollector.cpp b/xpcom/base/nsCycleCollector.cpp
--- a/xpcom/base/nsCycleCollector.cpp
+++ b/xpcom/base/nsCycleCollector.cpp
@@ -2391,7 +2391,8 @@ nsCycleCollector_suspectCurrent(nsISuppo
 {
     if (sCollector) {
         PRBool res = sCollector->Suspect(n, PR_TRUE);
-        NS_ASSERTION(res, "suspectCurrent should not fail");
+        NS_ASSERTION(res || sCollector->mParams.mDoNothing,
+                     "suspectCurrent should not fail");
     }
 }
 
