Index: netwerk/base/src/nsProxyAutoConfig.js
===================================================================
RCS file: /cvsroot/mozilla/netwerk/base/src/nsProxyAutoConfig.js,v
retrieving revision 1.54
diff -r1.54 netwerk/base/src/nsProxyAutoConfig.js
--- netwerk/base/src/nsProxyAutoConfig.js Mon Feb 25 22:44:59 2008
+++ netwerk/base/src/nsProxyAutoConfig.js    Mon Feb 25 22:45:18 2008
@@ -243,15 +243,14 @@ var pacUtils = 
 "   return newRe.test(url);\n" +
 "}\n" +
 
-"var wdays = new Array('SUN', 'MON', 'TUE', 'WED', 'THU', 'FRI', 'SAT');\n" +
+"var wdays = {SUN: 0, MON: 1, TUE: 2, WED: 3, THU: 4, FRI: 5, SAT: 6};\n" +
 
-"var monthes = new Array('JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN', 'JUL', 'AUG', 'SEP', 'OCT', 'NOV', 'DEC');\n"+
+"var months = {JAN: 0, FEB: 1, MAR: 2, APR: 3, MAY: 4, JUN: 5, JUL: 6, AUG: 7, SEP: 8, OCT: 9, NOV: 10, DEC: 11};\n"+
 
 "function weekdayRange() {\n" +
 "    function getDay(weekday) {\n" +
-"        for (var i = 0; i < 6; i++) {\n" +
-"            if (weekday == wdays[i]) \n" +
-"                return i;\n" +
+"        if (weekday in wdays) {\n" +
+"            return wdays[weekday];\n" +
 "        }\n" +
 "        return -1;\n" +
 "    }\n" +
@@ -274,9 +273,8 @@ var pacUtils = 
 
 "function dateRange() {\n" +
 "    function getMonth(name) {\n" +
-"        for (var i = 0; i < 6; i++) {\n" +
-"            if (name == monthes[i])\n" +
-"                return i;\n" +
+"        if (name in months) {\n" +
+"            return months[name];\n" +
 "        }\n" +
 "        return -1;\n" +
 "    }\n" +
