Index: content/base/src/nsXMLHttpRequest.cpp
===================================================================
RCS file: /cvsroot/mozilla/content/base/src/nsXMLHttpRequest.cpp,v
retrieving revision 1.184
diff -u -8 -p -r1.184 nsXMLHttpRequest.cpp
--- mozilla/content/base/src/nsXMLHttpRequest.cpp	2 Jul 2007 13:20:01 -0000	1.184
+++ mozilla/content/base/src/nsXMLHttpRequest.cpp	3 Jul 2007 12:07:41 -0000
@@ -1588,16 +1588,17 @@ nsXMLHttpRequest::Send(nsIVariant *aBody
           nsCOMPtr<nsISupportsString> wstr(do_QueryInterface(supports));
           if (wstr) {
             wstr->GetData(serial);
           } else {
             // stream?
             nsCOMPtr<nsIInputStream> stream(do_QueryInterface(supports));
             if (stream) {
               postDataStream = stream;
+              charset.Truncate();
             }
           }
         }
       }
       break;
     case nsIDataType::VTYPE_VOID:
     case nsIDataType::VTYPE_EMPTY:
       // Makes us act as if !aBody, don't upload anything
@@ -1637,18 +1638,28 @@ nsXMLHttpRequest::Send(nsIVariant *aBody
       nsCAutoString contentType;
       if (NS_FAILED(httpChannel->
                       GetRequestHeader(NS_LITERAL_CSTRING("Content-Type"),
                                        contentType)) ||
           contentType.IsEmpty()) {
         contentType = NS_LITERAL_CSTRING("application/xml");
       }
 
-      contentType.AppendLiteral(";charset=");
-      contentType.Append(charset);
+      // We don't want to set a charset for streams.
+      if (!charset.IsEmpty()) {
+        nsCAutoString actualType, dummy;
+        rv = NS_ParseContentType(contentType, actualType, dummy);
+        if (NS_FAILED(rv)) {
+          actualType.AssignLiteral("application/xml");
+        }
+
+        contentType.Assign(actualType);
+        contentType.AppendLiteral(";charset=");
+        contentType.Append(charset);
+      }
 
       rv = uploadChannel->SetUploadStream(postDataStream, contentType, -1);
       // Reset the method to its original value
       if (httpChannel) {
         httpChannel->SetRequestMethod(method);
       }
     }
   }

