Index: mozilla/dom/src/base/nsFocusController.cpp
===================================================================
--- mozilla.orig/dom/src/base/nsFocusController.cpp
+++ mozilla/dom/src/base/nsFocusController.cpp
@@ -272,36 +272,33 @@
     }
     else if (mCurrentWindow) {
       nsCOMPtr<nsIDOMDocument> domDoc;
       mCurrentWindow->GetDocument(getter_AddRefs(domDoc));
       doc = do_QueryInterface(domDoc);
     }
   }
 
-  if (!doc)
+  if (!doc) {
     // No way to obtain an event state manager.  Give up.
-    return NS_OK;
+    return NS_ERROR_FAILURE;
+  }
 
-
-  // Obtain a presentation context
-  nsIPresShell *shell = doc->GetPrimaryShell();
-  if (!shell)
-    return NS_OK;
+	nsIPresShell *shell = doc->GetPrimaryShell();
+	if (!shell)
+		return NS_ERROR_FAILURE;
 
   // Make sure frames have been constructed before shifting focus, bug 273092.
   shell->FlushPendingNotifications(Flush_Frames);
 
   // Retrieve the context
   nsCOMPtr<nsPresContext> presContext = shell->GetPresContext();
 
   // Make this ESM shift the focus per our instructions.
-  presContext->EventStateManager()->ShiftFocus(aForward, content);
-
-  return NS_OK;
+  return presContext->EventStateManager()->ShiftFocus(aForward, content);
 }
 
 /////
 // nsIDOMFocusListener
 /////
 
 nsresult 
 nsFocusController::Focus(nsIDOMEvent* aEvent)
