# bug 463933
Index: mozilla/gfx/src/thebes/nsSystemFontsGTK2.cpp
===================================================================
--- mozilla.orig/gfx/src/thebes/nsSystemFontsGTK2.cpp
+++ mozilla/gfx/src/thebes/nsSystemFontsGTK2.cpp
@@ -50,36 +50,45 @@
 #include <pango/pango.h>
 #include <pango/pango-fontmap.h>
 #endif
 
 #include <fontconfig/fontconfig.h>
 #include "nsSystemFontsGTK2.h"
 #include "gfxPlatformGtk.h"
 
+#include "nsIPref.h"
+#include "nsIServiceManager.h"
+
 // Glue to avoid build/runtime dependencies on Pango > 1.6
 #if defined(MOZ_PANGO) && !defined(THEBES_USE_PANGO_CAIRO)
 static gboolean
 (* PTR_pango_font_description_get_size_is_absolute)(PangoFontDescription*)
     = nsnull;
 
+static PRInt32 sSysFontDpiAdd = 0;
+
 static void InitPangoLib()
 {
     static PRBool initialized = PR_FALSE;
     if (initialized)
         return;
     initialized = PR_TRUE;
 
     PRLibrary *pangoLib = nsnull;
     PTR_pango_font_description_get_size_is_absolute =
         (gboolean (*)(PangoFontDescription*))
         PR_FindFunctionSymbolAndLibrary("pango_font_description_get_size_is_absolute",
                                         &pangoLib);
     if (pangoLib)
         PR_UnloadLibrary(pangoLib);
+
+    nsCOMPtr<nsIPref> prefs(do_GetService(NS_PREF_CONTRACTID));
+    if (prefs)
+        prefs->GetIntPref("system.font.dpi.add", &sSysFontDpiAdd);
 }
 
 static void
 ShutdownPangoLib()
 {
 }
 
 static gboolean
@@ -216,17 +225,17 @@ nsSystemFontsGTK2::GetSystemFontInfo(Gtk
     aFontStyle->stretch = NS_FONT_STRETCH_NORMAL;
 
     float size = float(pango_font_description_get_size(desc)) / PANGO_SCALE;
 
     // |size| is now either pixels or pango-points (not Mozilla-points!)
 
     if (!MOZ_pango_font_description_get_size_is_absolute(desc)) {
         // |size| is in pango-points, so convert to pixels.
-        size *= gfxPlatformGtk::DPI() / POINTS_PER_INCH_FLOAT;
+        size *= (gfxPlatformGtk::DPI() + sSysFontDpiAdd) / POINTS_PER_INCH_FLOAT;
     }
 
     // |size| is now pixels
 
     aFontStyle->size = size;
   
     pango_font_description_free(desc);
 
