# bug 463933
diff --git a/gfx/src/thebes/nsSystemFontsGTK2.cpp b/gfx/src/thebes/nsSystemFontsGTK2.cpp
--- a/gfx/src/thebes/nsSystemFontsGTK2.cpp
+++ b/gfx/src/thebes/nsSystemFontsGTK2.cpp
@@ -55,11 +55,16 @@
 #include "nsSystemFontsGTK2.h"
 #include "gfxPlatformGtk.h"
 
+#include "nsIPref.h"
+#include "nsIServiceManager.h"
+
 // Glue to avoid build/runtime dependencies on Pango > 1.6
 #if defined(MOZ_PANGO) && !defined(THEBES_USE_PANGO_CAIRO)
 static gboolean
 (* PTR_pango_font_description_get_size_is_absolute)(PangoFontDescription*)
     = nsnull;
+
+static PRInt32 sSysFontDpiAdd = 0;
 
 static void InitPangoLib()
 {
@@ -75,6 +80,10 @@
                                         &pangoLib);
     if (pangoLib)
         PR_UnloadLibrary(pangoLib);
+
+    nsCOMPtr<nsIPref> prefs(do_GetService(NS_PREF_CONTRACTID));
+    if (prefs)
+        prefs->GetIntPref("system.font.dpi.add", &sSysFontDpiAdd);
 }
 
 static void
@@ -219,7 +228,7 @@
 
     if (!MOZ_pango_font_description_get_size_is_absolute(desc)) {
         // |size| is in pango-points, so convert to pixels.
-        size *= gfxPlatformGtk::DPI() / POINTS_PER_INCH_FLOAT;
+        size *= (gfxPlatformGtk::DPI() + sSysFontDpiAdd) / POINTS_PER_INCH_FLOAT;
     }
 
     // |size| is now pixels
