--- mozilla/intl/locale/src/unix/nsPosixLocale.cpp.orig	2010-03-09 07:35:15.000000000 +0200
+++ mozilla/intl/locale/src/unix/nsPosixLocale.cpp	2010-03-09 07:36:33.000000000 +0200
@@ -37,20 +37,24 @@
 
 #include "nsISupports.h"
 #include "nscore.h"
 #include "nsString.h"
 #include "nsPosixLocale.h"
 #include "nsLocaleCID.h"
 #include "prprf.h"
 #include "plstr.h"
 #include "nsReadableUtils.h"
 
+#ifdef MOZ_PLATFORM_HILDON
+#define MAEMO_CHANGES
+#endif
+
 /* nsPosixLocale ISupports */
 NS_IMPL_ISUPPORTS1(nsPosixLocale, nsIPosixLocale)
 
 nsPosixLocale::nsPosixLocale(void)
 {
 }
 
 nsPosixLocale::~nsPosixLocale(void)
 {
 
@@ -108,27 +112,28 @@ nsPosixLocale::GetXPLocale(const char* p
     if (strcmp(posixLocale,"C")==0 || strcmp(posixLocale,"POSIX")==0) {
       locale.AssignLiteral("en-US");
       return NS_OK;
     }
     if (!ParseLocaleString(posixLocale,lang_code,country_code,extra,'_')) {
 //      * locale = "x-user-defined";
       // use posix if parse failed
       CopyASCIItoUTF16(nsDependentCString(posixLocale), locale);
       return NS_OK;
     }
-
+// In Hildon we have only no-NO, not nb-NO and nn-NO
+#ifndef MAEMO_CHANGES
     // Special case: substitute "nb" (Norwegian Bokmal) for macrolanguage
     // code "no" (Norwegian)
     if (nsDependentCString(lang_code).LowerCaseEqualsLiteral("no")) {
       lang_code[1] = 'b';
     }
-
+#endif
     if (*country_code) {
       PR_snprintf(posix_locale,sizeof(posix_locale),"%s-%s",lang_code,country_code);
     } 
     else {
       PR_snprintf(posix_locale,sizeof(posix_locale),"%s",lang_code);
     }
 
     CopyASCIItoUTF16(nsDependentCString(posix_locale), locale);
     return NS_OK;
 
