# Bug 373008  [spatial] [@ nsSpatialNavigation::getContentInDirection - nsSpatialNavigation::handleMove]
#
Index: mozilla/extensions/spatialnavigation/src/nsSpatialNavigation.cpp
===================================================================
--- mozilla.orig/extensions/spatialnavigation/src/nsSpatialNavigation.cpp
+++ mozilla/extensions/spatialnavigation/src/nsSpatialNavigation.cpp
@@ -577,17 +577,17 @@ nsresult
 nsSpatialNavigation::getContentInDirection(int aDirection, 
                                            nsPresContext* aPresContext, 
                                            nsRect& aFocusedRect, 
                                            nsIFrame* aFocusedFrame, 
                                            PRBool aIsAREA,
                                            PRBool aFocusDocuments,
                                            nsIContent** aContent)
 {  
-
+  NS_ENSURE_ARG_POINTER(aFocusedFrame);
   // Check to see if we should decend into subdoc
   nsIContent* subFrameContent = aFocusedFrame->GetContent();
   nsCOMPtr<nsIDOMHTMLHtmlElement> hhElement = do_QueryInterface(subFrameContent);
   nsCOMPtr<nsIDOMHTMLIFrameElement> iFrameElement = do_QueryInterface(subFrameContent);
 
   if ( (hhElement || iFrameElement) && mNavigationFramesState)
   {
     aPresContext = getPresContext(subFrameContent);
@@ -648,30 +648,32 @@ nsSpatialNavigation::handleMove(int dire
   nsIFrame* focusedFrame;
   getFrameForContent(focusedContent, &focusedFrame);
 
   nsRect focusedRect;
   PRBool isAREA = isArea(focusedContent);
   if (!isAREA) 
   {
     // RECT !!
-    focusedRect = makeRectRelativeToGlobalView(focusedFrame);
+		if (focusedFrame)
+	 		focusedRect = makeRectRelativeToGlobalView(focusedFrame);
 
     // deflate the rect to avoid overlapping with other
     // rects.
     focusedRect.Deflate(gRectFudge, gRectFudge);
   }
   else
   {
     nsCOMPtr<nsIDOMHTMLAreaElement> e = do_QueryInterface(focusedContent);
     getRectOfAreaElement(focusedFrame, e, &focusedRect);
   }
 
   nsCOMPtr<nsIContent> c;
-  getContentInDirection(direction, presContext, focusedRect, focusedFrame, PR_FALSE, isAREA, getter_AddRefs(c));
+  if (focusedFrame)
+    getContentInDirection(direction, presContext, focusedRect, focusedFrame, PR_FALSE, isAREA, getter_AddRefs(c));
   
   if (c) {
    
     nsIDocument* doc = c->GetDocument();
     if (!doc)
       return NS_ERROR_FAILURE;
     
 /*    nsIPresShell *presShell = doc->GetPrimaryShell();
@@ -730,16 +732,17 @@ nsSpatialNavigation::handleMove(int dire
     if (!subdoc) return NS_OK;
 
     nsIPresShell *subdocShell = subdoc->GetPrimaryShell();
     if (!subdocShell) return NS_OK;
   
     nsPresContext *subdocPresContext = subdocShell->GetPresContext();
 
     nsIFrame* subdocFrame = subdocShell->GetRootFrame();
+    NS_ENSURE_TRUE(subdocFrame, NS_ERROR_FAILURE);
 
     nsRect subdocRect = subdocFrame->GetRect();
 
     nsPoint frame_offset = subdocFrame->GetOffsetToExternal(parentFrame);
 	
 	subdocRect.x = frame_offset.x;
 	subdocRect.y = frame_offset.y;
 
