#Disable bilinear scaling for images
#bug 99283
Index: mozilla/gfx/src/thebes/nsThebesImage.cpp
===================================================================
--- mozilla.orig/gfx/src/thebes/nsThebesImage.cpp
+++ mozilla/gfx/src/thebes/nsThebesImage.cpp
@@ -681,16 +681,22 @@ nsThebesImage::Draw(gfxContext*        a
         }
     }
 
     if ((op == gfxContext::OPERATOR_OVER || pushedGroup) &&
         format == gfxASurface::ImageFormatRGB24) {
         aContext->SetOperator(gfxContext::OPERATOR_SOURCE);
     }
 
+#ifdef __arm__ // Disable bilinear scaling on arm, because pixman is slow
+
+    if (currentTarget->GetType() == gfxASurface::SurfaceTypeImage)
+      pattern->SetFilter(gfxPattern::FILTER_NEAREST);
+#endif
+
     // Phew! Now we can actually draw this image
     aContext->NewPath();
     aContext->SetPattern(pattern);
     aContext->Rectangle(fill);
     aContext->Fill();
   
     aContext->SetOperator(op);
     if (pushedGroup) {
