#Disable bilinear scaling for images
#bug 99283
diff -r b5574246655a gfx/src/thebes/nsThebesImage.cpp
--- a/gfx/src/thebes/nsThebesImage.cpp	Fri Feb 06 12:00:37 2009 +0200
+++ b/gfx/src/thebes/nsThebesImage.cpp	Sat Feb 07 16:36:19 2009 +0200
@@ -676,6 +676,12 @@ nsThebesImage::Draw(gfxContext*        a
         aContext->SetOperator(gfxContext::OPERATOR_SOURCE);
     }
 
+#ifdef __arm__ // Disable bilinear scaling on arm, because pixman is slow
+
+    if (currentTarget->GetType() == gfxASurface::SurfaceTypeImage)
+      pattern->SetFilter(3);
+#endif
+
     // Phew! Now we can actually draw this image
     aContext->NewPath();
     aContext->SetPattern(pattern);
