diff -ruN -p -U5 src.orig/EmbedProgress.cpp src/EmbedProgress.cpp
--- mozilla/embedding/browser/gtk/src.orig/EmbedProgress.cpp	2007-12-23 12:37:05.000000000 +0200
+++ mozilla/embedding/browser/gtk/src/EmbedProgress.cpp	2007-12-23 12:37:31.000000000 +0200
@@ -40,11 +40,10 @@
 #include "nsIChannel.h"
 #include "nsIWebProgress.h"
 #include "nsIDOMWindow.h"
 
 #include "nsIURI.h"
-#include "nsCRTGlue.h"
 
 EmbedProgress::EmbedProgress(void)
 {
   mOwner = nsnull;
 }
@@ -175,21 +174,15 @@ NS_IMETHODIMP
 EmbedProgress::OnStatusChange(nsIWebProgress  *aWebProgress,
 			      nsIRequest      *aRequest,
 			      nsresult         aStatus,
 			      const PRUnichar *aMessage)
 {
-  // need to make a copy so we can safely cast to a void *
-  PRUnichar *tmpString = NS_strdup(aMessage);
-
   g_signal_emit(G_OBJECT(mOwner->mOwningWidget),
 		  moz_embed_signals[STATUS_CHANGE], 0,
 		  static_cast<void *>(aRequest),
 		  static_cast<int>(aStatus),
-		  static_cast<void *>(tmpString));
-
-  NS_Free(tmpString);
-
+		  static_cast<const void *>(aMessage));
   return NS_OK;
 }
 
 NS_IMETHODIMP
 EmbedProgress::OnSecurityChange(nsIWebProgress *aWebProgress,
