# proper profile dir 408681
Index: src/EmbedPrivate.cpp
===================================================================
--- mozilla/embedding/browser/gtk/src.orig/EmbedPrivate.cpp
+++ mozilla/embedding/browser/gtk/src/EmbedPrivate.cpp
@@ -543,16 +543,22 @@
       return;
 
     nsCOMPtr<nsILocalFile> greDir;
     rv = NS_NewNativeLocalFile(nsDependentCString(grePath), PR_TRUE,
                                getter_AddRefs(greDir));
     if (NS_FAILED(rv))
       return;
 
+    if (sProfileDir && !sProfileLock) {
+      rv = XRE_LockProfileDirectory(sProfileDir,
+                                    &sProfileLock);
+      if (NS_FAILED(rv)) return;
+    }
+
     rv = XRE_InitEmbedding(greDir, binDir,
                            const_cast<GTKEmbedDirectoryProvider*>(&kDirectoryProvider),
                            nsnull, nsnull);
     if (NS_FAILED(rv))
       return;
 
     if (sProfileDir)
       XRE_NotifyProfile();
@@ -575,16 +581,19 @@
     // we no longer need a reference to the DirectoryServiceProvider
     if (sAppFileLocProvider) {
       NS_RELEASE(sAppFileLocProvider);
       sAppFileLocProvider = nsnull;
     }
 
     // shut down XPCOM/Embedding
     XRE_TermEmbedding();
+
+    NS_IF_RELEASE(sProfileLock);
+    NS_IF_RELEASE(sProfileDir);
   }
 }
 
 /* static */
 void EmbedPrivate::SetPath(const char *aPath)
 {
   if (sPath)
     free(sPath);
@@ -630,18 +639,23 @@
   }
 
   nsresult rv =
     NS_NewNativeLocalFile(nsDependentCString(aDir), PR_TRUE, &sProfileDir);
 
   if (NS_SUCCEEDED(rv) && aName)
     rv = sProfileDir->AppendNative(nsDependentCString(aName));
 
-  if (NS_SUCCEEDED(rv))
+  if (NS_SUCCEEDED(rv)) {
+    PRBool exists = PR_FALSE;
+    rv = sProfileDir->Exists(&exists);
+    if (!exists)
+      rv = sProfileDir->Create(nsIFile::DIRECTORY_TYPE, 0700);
     rv = XRE_LockProfileDirectory(sProfileDir, &sProfileLock);
+  }
 
   if (NS_SUCCEEDED(rv)) {
     if (sWidgetCount)
       XRE_NotifyProfile();
 
     return;
   }
 
