Index: mozilla/embedding/browser/gtk/src/EmbedContentListener.h
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/EmbedContentListener.h
+++ mozilla/embedding/browser/gtk/src/EmbedContentListener.h
@@ -33,18 +33,18 @@
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
 #ifndef __EmbedContentListener_h
 #define __EmbedContentListener_h
 
-#include <nsIURIContentListener.h>
-#include <nsWeakReference.h>
+#include "nsIURIContentListener.h"
+#include "nsWeakReference.h"
 
 class EmbedPrivate;
 
 class EmbedContentListener : public nsIURIContentListener,
                              public nsSupportsWeakReference
 {
  public:
 
Index: mozilla/embedding/browser/gtk/src/EmbedEventListener.cpp
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/EmbedEventListener.cpp
+++ mozilla/embedding/browser/gtk/src/EmbedEventListener.cpp
@@ -30,18 +30,18 @@
  * use your version of this file under the terms of the MPL, indicate your
  * decision by deleting the provisions above and replace them with the notice
  * and other provisions required by the GPL or the LGPL. If you do not delete
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
-#include <nsCOMPtr.h>
-#include <nsIDOMMouseEvent.h>
+#include "nsCOMPtr.h"
+#include "nsIDOMMouseEvent.h"
 
 #include "nsIDOMKeyEvent.h"
 #include "nsIDOMUIEvent.h"
 
 #include "EmbedEventListener.h"
 #include "EmbedPrivate.h"
 
 EmbedEventListener::EmbedEventListener(void)
Index: mozilla/embedding/browser/gtk/src/EmbedEventListener.h
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/EmbedEventListener.h
+++ mozilla/embedding/browser/gtk/src/EmbedEventListener.h
@@ -33,19 +33,19 @@
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
 #ifndef __EmbedEventListener_h
 #define __EmbedEventListener_h
 
-#include <nsIDOMKeyListener.h>
-#include <nsIDOMMouseListener.h>
-#include <nsIDOMUIListener.h>
+#include "nsIDOMKeyListener.h"
+#include "nsIDOMMouseListener.h"
+#include "nsIDOMUIListener.h"
 
 class EmbedPrivate;
 
 class EmbedEventListener : public nsIDOMKeyListener,
                            public nsIDOMMouseListener,
                            public nsIDOMUIListener
 {
  public:
Index: mozilla/embedding/browser/gtk/src/EmbedPrivate.cpp
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/EmbedPrivate.cpp
+++ mozilla/embedding/browser/gtk/src/EmbedPrivate.cpp
@@ -30,53 +30,53 @@
  * use your version of this file under the terms of the MPL, indicate your
  * decision by deleting the provisions above and replace them with the notice
  * and other provisions required by the GPL or the LGPL. If you do not delete
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
-#include <nsIDocShell.h>
-#include <nsIWebProgress.h>
+#include "nsIDocShell.h"
+#include "nsIWebProgress.h"
 #include "nsIWidget.h"
 #include "nsCRT.h"
 #include "nsNetUtil.h"
 #include "nsIWebBrowserStream.h"
 #include "nsIWebBrowserFocus.h"
 #include "nsIDirectoryService.h"
 #include "nsAppDirectoryServiceDefs.h"
 
 // for do_GetInterface
-#include <nsIInterfaceRequestor.h>
+#include "nsIInterfaceRequestor.h"
 // for do_CreateInstance
-#include <nsIComponentManager.h>
+#include "nsIComponentManager.h"
 
 // for initializing our window watcher service
-#include <nsIWindowWatcher.h>
+#include "nsIWindowWatcher.h"
 
-#include <nsILocalFile.h>
+#include "nsILocalFile.h"
 
 #include "nsXULAppAPI.h"
 
 // all of the crap that we need for event listeners
 // and when chrome windows finish loading
-#include <nsIDOMWindow.h>
-#include <nsPIDOMWindow.h>
-#include <nsIDOMWindowInternal.h>
+#include "nsIDOMWindow.h"
+#include "nsPIDOMWindow.h"
+#include "nsIDOMWindowInternal.h"
 
 // For seting scrollbar visibilty
-#include <nsIDOMBarProp.h>
+#include "nsIDOMBarProp.h"
 
 // for the focus hacking we need to do
-#include <nsIFocusController.h>
+#include "nsIFocusController.h"
 
 // app component registration
-#include <nsIGenericFactory.h>
-#include <nsIComponentRegistrar.h>
+#include "nsIGenericFactory.h"
+#include "nsIComponentRegistrar.h"
 
 // all of our local includes
 #include "EmbedPrivate.h"
 #include "EmbedWindow.h"
 #include "EmbedProgress.h"
 #include "EmbedContentListener.h"
 #include "EmbedEventListener.h"
 #include "EmbedWindowCreator.h"
Index: mozilla/embedding/browser/gtk/src/EmbedPrivate.h
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/EmbedPrivate.h
+++ mozilla/embedding/browser/gtk/src/EmbedPrivate.h
@@ -33,29 +33,29 @@
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
 #ifndef __EmbedPrivate_h
 #define __EmbedPrivate_h
 
-#include <nsCOMPtr.h>
-#include <nsString.h>
-#include <nsIWebNavigation.h>
-#include <nsISHistory.h>
+#include "nsCOMPtr.h"
+#include "nsString.h"
+#include "nsIWebNavigation.h"
+#include "nsISHistory.h"
 // for our one function that gets the EmbedPrivate via the chrome
 // object.
-#include <nsIWebBrowserChrome.h>
-#include <nsIAppShell.h>
-#include <nsPIDOMEventTarget.h>
-#include <nsVoidArray.h>
+#include "nsIWebBrowserChrome.h"
+#include "nsIAppShell.h"
+#include "nsPIDOMEventTarget.h"
+#include "nsVoidArray.h"
 // app component registration
-#include <nsIGenericFactory.h>
-#include <nsIComponentRegistrar.h>
+#include "nsIGenericFactory.h"
+#include "nsIComponentRegistrar.h"
 
 #include "gtkmozembedprivate.h"
 
 class EmbedProgress;
 class EmbedWindow;
 class EmbedContentListener;
 class EmbedEventListener;
 
Index: mozilla/embedding/browser/gtk/src/EmbedProgress.h
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/EmbedProgress.h
+++ mozilla/embedding/browser/gtk/src/EmbedProgress.h
@@ -33,18 +33,18 @@
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
 #ifndef __EmbedProgress_h
 #define __EmbedProgress_h
 
-#include <nsIWebProgressListener.h>
-#include <nsWeakReference.h>
+#include "nsIWebProgressListener.h"
+#include "nsWeakReference.h"
 #include "EmbedPrivate.h"
 
 class EmbedProgress : public nsIWebProgressListener,
                       public nsSupportsWeakReference
 {
  public:
   EmbedProgress();
   virtual ~EmbedProgress();
Index: mozilla/embedding/browser/gtk/src/EmbedPrompter.h
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/EmbedPrompter.h
+++ mozilla/embedding/browser/gtk/src/EmbedPrompter.h
@@ -33,17 +33,17 @@
  * use your version of this file under the terms of the MPL, indicate your
  * decision by deleting the provisions above and replace them with the notice
  * and other provisions required by the GPL or the LGPL. If you do not delete
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
-#include <nsString.h>
+#include "nsString.h"
 #include <gtk/gtk.h>
 
 #define EMBED_MAX_BUTTONS 3
 
 class EmbedPrompter {
 
 public:
 
Index: mozilla/embedding/browser/gtk/src/EmbedWindow.h
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/EmbedWindow.h
+++ mozilla/embedding/browser/gtk/src/EmbedWindow.h
@@ -33,26 +33,25 @@
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
 #ifndef __EmbedWindow_h
 #define __EmbedWindow_h
 
-#include <nsString.h>
-#include <nsIWebBrowserChrome.h>
-#include <nsIWebBrowserChromeFocus.h>
-#include <nsIEmbeddingSiteWindow.h>
-#include <nsITooltipListener.h>
-#include <nsISupports.h>
-#include <nsIWebBrowser.h>
-#include <nsIBaseWindow.h>
-#include <nsIInterfaceRequestor.h>
-#include <nsCOMPtr.h>
+#include "nsIWebBrowserChrome.h"
+#include "nsIWebBrowserChromeFocus.h"
+#include "nsIEmbeddingSiteWindow.h"
+#include "nsITooltipListener.h"
+#include "nsISupports.h"
+#include "nsIWebBrowser.h"
+#include "nsIBaseWindow.h"
+#include "nsIInterfaceRequestor.h"
+#include "nsCOMPtr.h"
 #include "nsString.h"
 
 #include <gtk/gtk.h>
 
 class EmbedPrivate;
 
 class EmbedWindow : public nsIWebBrowserChrome,
 		    public nsIWebBrowserChromeFocus,
Index: mozilla/embedding/browser/gtk/src/gtkmozembed_internal.h
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/gtkmozembed_internal.h
+++ mozilla/embedding/browser/gtk/src/gtkmozembed_internal.h
@@ -33,18 +33,18 @@
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
 #ifndef gtkmozembed_internal_h
 #define gtkmozembed_internal_h
 
-#include <nsIWebBrowser.h>
-#include <nsXPCOM.h>
+#include "nsIWebBrowser.h"
+#include "nsXPCOM.h"
 
 struct nsModuleComponentInfo;
 
 #ifdef __cplusplus
 extern "C" {
 #endif /* __cplusplus */
 
 GTKMOZEMBED_API(void,
Index: mozilla/embedding/browser/gtk/src/GtkPromptService.cpp
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/GtkPromptService.cpp
+++ mozilla/embedding/browser/gtk/src/GtkPromptService.cpp
@@ -34,22 +34,22 @@
  * and other provisions required by the GPL or the LGPL. If you do not delete
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
 #include "GtkPromptService.h"
 #include "EmbedPrompter.h"
-#include <nsString.h>
-#include <nsIWindowWatcher.h>
-#include <nsIWebBrowserChrome.h>
-#include <nsIEmbeddingSiteWindow.h>
-#include <nsCOMPtr.h>
-#include <nsIServiceManager.h>
+#include "nsString.h"
+#include "nsIWindowWatcher.h"
+#include "nsIWebBrowserChrome.h"
+#include "nsIEmbeddingSiteWindow.h"
+#include "nsCOMPtr.h"
+#include "nsIServiceManager.h"
 
 GtkPromptService::GtkPromptService()
 {
 }
 
 GtkPromptService::~GtkPromptService()
 {
 }
Index: mozilla/embedding/browser/gtk/src/GtkPromptService.h
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/GtkPromptService.h
+++ mozilla/embedding/browser/gtk/src/GtkPromptService.h
@@ -32,18 +32,18 @@
  * use your version of this file under the terms of the MPL, indicate your
  * decision by deleting the provisions above and replace them with the notice
  * and other provisions required by the GPL or the LGPL. If you do not delete
  * the provisions above, a recipient may use your version of this file under
  * the terms of any one of the MPL, the GPL or the LGPL.
  *
  * ***** END LICENSE BLOCK ***** */
 
-#include <nsIPromptService.h>
-#include <nsString.h>
+#include "nsIPromptService.h"
+#include "nsString.h"
 #include <gtk/gtk.h>
 
 class nsIDOMWindow;
 
 class GtkPromptService : public nsIPromptService
 {
 public:
     GtkPromptService();
