Index: mozilla/embedding/browser/gtk/src/EmbedPrivate.cpp
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/EmbedPrivate.cpp
+++ mozilla/embedding/browser/gtk/src/EmbedPrivate.cpp
@@ -84,16 +84,20 @@
 #include "GtkPromptService.h"
 
 #ifdef MOZ_ACCESSIBILITY_ATK
 #include "nsIAccessibilityService.h"
 #include "nsIAccessible.h"
 #include "nsIDOMDocument.h"
 #endif
 
+#ifdef MOZ_RDF
+#include "nsIExtensionManager.h"
+#endif
+
 PRUint32                 EmbedPrivate::sWidgetCount = 0;
 
 char                    *EmbedPrivate::sPath        = nsnull;
 char                    *EmbedPrivate::sCompPath    = nsnull;
 nsTArray<EmbedPrivate*> *EmbedPrivate::sWindowList  = nsnull;
 nsILocalFile            *EmbedPrivate::sProfileDir  = nsnull;
 nsISupports             *EmbedPrivate::sProfileLock = nsnull;
 GtkWidget               *EmbedPrivate::sOffscreenWindow = 0;
@@ -537,16 +541,24 @@ EmbedPrivate::PushStartup(void)
                            const_cast<GTKEmbedDirectoryProvider*>(&kDirectoryProvider),
                            nsnull, nsnull);
     if (NS_FAILED(rv))
       return;
 
     if (sProfileDir)
       XRE_NotifyProfile();
 
+#ifdef MOZ_RDF
+    PRBool needsRestart;
+    nsCOMPtr<nsIExtensionManager>
+      extensionManager(do_GetService("@mozilla.org/extensions/manager;1", &rv));
+    if (extensionManager)
+      extensionManager->Start(&needsRestart);
+#endif
+
     rv = RegisterAppComponents();
     NS_ASSERTION(NS_SUCCEEDED(rv), "Warning: Failed to register app components.\n");
   }
 }
 
 /* static */
 void
 EmbedPrivate::PopStartup(void)
@@ -969,16 +981,17 @@ EmbedPrivate::RegisterAppComponents(void
   nsCOMPtr<nsIComponentRegistrar> cr;
   nsresult rv = NS_GetComponentRegistrar(getter_AddRefs(cr));
   NS_ENSURE_SUCCESS(rv, rv);
 
   nsCOMPtr<nsIComponentManager> cm;
   rv = NS_GetComponentManager (getter_AddRefs (cm));
   NS_ENSURE_SUCCESS (rv, rv);
 
+  cr->AutoRegister(nsnull);
   for (int i = 0; i < sNumAppComps; ++i) {
     nsCOMPtr<nsIGenericFactory> componentFactory;
     rv = NS_NewGenericFactory(getter_AddRefs(componentFactory),
                               &(sAppComps[i]));
     if (NS_FAILED(rv)) {
       NS_WARNING("Unable to create factory for component");
       continue;  // don't abort registering other components
     }
Index: mozilla/embedding/browser/gtk/src/Makefile.in
===================================================================
--- mozilla.orig/embedding/browser/gtk/src/Makefile.in
+++ mozilla/embedding/browser/gtk/src/Makefile.in
@@ -60,16 +60,17 @@ REQUIRES	= xpcom \
 		  layout \
 		  uriloader \
 		  webbrwsr \
 		  shistory \
 		  embed_base \
 		  windowwatcher \
 		  profdirserviceprovider \
 		  xulapp \
+		  extensions \
 		  $(NULL)
 
 ifdef ACCESSIBILITY
 REQUIRES        += accessibility
 endif
 
 CPPSRCS		= \
 		gtkmozembed2.cpp \
