--- a/gfx/src/thebes/nsThebesImage.cpp	2009-06-03 14:19:04.000000000 +0300
+++ b/gfx/src/thebes/nsThebesImage.cpp	2009-06-03 17:31:05.000000000 +0300
@@ -44,10 +44,14 @@
 
 #include "gfxPlatform.h"
 
 #include "prenv.h"
 
+#ifdef NS_OSSO
+#include <osso-mem.h>
+#endif
+
 static PRBool gDisableOptimize = PR_FALSE;
 
 #ifdef XP_WIN
 static PRUint32 gTotalDDBs = 0;
 static PRUint32 gTotalDDBSize = 0;
@@ -102,10 +106,19 @@ nsThebesImage::Init(PRInt32 aWidth, PRIn
 
     PRBool lowMemory;
     mem->IsLowMemory(&lowMemory);
     if (lowMemory)
         return NS_ERROR_OUT_OF_MEMORY;
+
+#ifdef NS_OSSO
+    // Check if we have enough available memory for the image. At this point
+    // we do not know exact pixel size, so we assume worst case (32 bits).
+    PRUint64 rawSize = PRUint64(aWidth) * PRUint64(aHeight) * 4;
+    osso_mem_usage_t memInfo;
+    if ((osso_mem_get_usage(&memInfo) != 0) || (rawSize > memInfo.usable))
+        return NS_ERROR_OUT_OF_MEMORY;
+#endif
 
     gfxImageSurface::gfxImageFormat format;
     switch(aMaskRequirements)
     {
         case nsMaskRequirements_kNeeds1Bit:
diff -r c510e80d79e7 gfx/src/thebes/Makefile.in
--- a/gfx/src/thebes/Makefile.in	Sat Jun 13 23:04:31 2009 +0300
+++ b/gfx/src/thebes/Makefile.in	Sat Jun 13 23:05:12 2009 +0300
@@ -135,6 +135,10 @@ ifdef MOZ_X11
 EXTRA_DSO_LDOPTS += $(XLDFLAGS) $(XLIBS)
 endif
 
+ifdef NS_OSSO
+EXTRA_DSO_LDOPTS += $(LIBOSSO_LIBS)
+endif
+
 include $(topsrcdir)/config/rules.mk
 
 CXXFLAGS  += $(TK_CFLAGS)
