diff -r bc556f9eaf8b content/base/src/nsScriptLoader.cpp
--- a/content/base/src/nsScriptLoader.cpp	Mon Jun 22 07:45:37 2009 +0300
+++ b/content/base/src/nsScriptLoader.cpp	Mon Jun 22 07:49:10 2009 +0300
@@ -260,7 +260,8 @@ nsScriptLoader::StartLoad(nsScriptLoadRe
   nsCOMPtr<nsILoadGroup> loadGroup = mDocument->GetDocumentLoadGroup();
   nsCOMPtr<nsIStreamLoader> loader;
 
-  nsCOMPtr<nsPIDOMWindow> window(do_QueryInterface(mDocument->GetScriptGlobalObject()));
+  nsCOMPtr<nsPIDOMWindow> window(do_QueryInterface(mDocument->GetScriptGlobalObject(), &rv));
+  NS_ENSURE_TRUE(window, rv);
   nsIDocShell *docshell = window->GetDocShell();
 
   nsCOMPtr<nsIInterfaceRequestor> prompter(do_QueryInterface(docshell));
