diff --git a/toolkit/xre/nsXREDirProvider.cpp b/toolkit/xre/nsXREDirProvider.cpp
--- a/toolkit/xre/nsXREDirProvider.cpp
+++ b/toolkit/xre/nsXREDirProvider.cpp
@@ -1278,16 +1278,19 @@ nsXREDirProvider::AppendSysUserExtension
 
 nsresult
 nsXREDirProvider::AppendProfilePath(nsIFile* aFile)
 {
   NS_ASSERTION(aFile, "Null pointer!");
 
   nsresult rv;
 
+  if (!gAppData)
+    return NS_ERROR_FAILURE;
+
 #if defined (XP_MACOSX)
   if (gAppData->profile) {
     rv = AppendProfileString(aFile, gAppData->profile);
   }
   else {
     // Note that MacOS ignores the vendor when creating the profile hierarchy -
     // all application preferences directories live alongside one another in
     // ~/Library/Application Support/
