diff -r f4c88d2be5bd js/src/jslock.cpp
--- a/js/src/jslock.cpp	Mon Jul 21 12:15:18 2008 -0700
+++ b/js/src/jslock.cpp	Tue Jul 22 00:25:38 2008 +0300
@@ -187,8 +187,14 @@ static JS_INLINE int
 static JS_INLINE int
 NativeCompareAndSwap(jsword *w, jsword ov, jsword nv)
 {
-    volatile int *vp = (volatile int*)w;
-    return !__kernel_cmpxchg(ov, nv, vp);
+    volatile PRInt32 *vp = (volatile PRInt32*)w;
+    PRInt32 failed = 1;
+
+    JS_ASSERT(ov != nv);
+    do {
+        failed = __kernel_cmpxchg(ov, nv, vp);
+    } while (failed && *vp == ov);
+    return !failed;
 }
 
 #else
