diff -r c7d11c7900f8 modules/plugin/base/src/nsPluginInstancePeer.cpp
--- a/modules/plugin/base/src/nsPluginInstancePeer.cpp	Mon Jul 13 08:57:23 2009 +0300
+++ b/modules/plugin/base/src/nsPluginInstancePeer.cpp	Mon Jul 13 08:59:19 2009 +0300
@@ -689,8 +689,10 @@ nsPluginInstancePeerImpl::GetJSContext(J
 {
   *outContext = NULL;
   nsresult rv = NS_ERROR_FAILURE;
+  if (!mOwner)
+    return rv;
+
   nsCOMPtr<nsIDocument> document;
-
   rv = mOwner->GetDocument(getter_AddRefs(document));
 
   if (NS_SUCCEEDED(rv) && document) {
diff -r c7d11c7900f8 modules/plugin/base/src/nsPluginNativeWindowGtk2.cpp
--- a/modules/plugin/base/src/nsPluginNativeWindowGtk2.cpp	Mon Jul 13 08:57:23 2009 +0300
+++ b/modules/plugin/base/src/nsPluginNativeWindowGtk2.cpp	Mon Jul 13 08:59:19 2009 +0300
@@ -162,14 +162,12 @@ nsPluginNativeWindowGtk2::plugin_composi
 {
   nsPluginNativeWindowGtk2 *native_window = (nsPluginNativeWindowGtk2*)data;
   XDamageNotifyEvent *ev;
-  XserverRegion parts;
   ev = (XDamageNotifyEvent *) xevent;
   if (ev->type != xdamage_event_base + XDamageNotify)
     return GDK_FILTER_CONTINUE;
 
   //printf("Damage event %d %d %d %d\n",ev->area.x, ev->area.y, ev->area.width, ev->area.height);
-  parts = XFixesCreateRegion (GDK_DISPLAY(), 0, 0);
-  XDamageSubtract (GDK_DISPLAY(), native_window->mDamage, None, parts);
+  XDamageSubtract (GDK_DISPLAY(), native_window->mDamage, None, None);
 
   /* We try to do our area invalidation here */
   nsPluginRect rect;
