--- mozilla/js/src/jslock.cpp.orig	2008-07-18 15:06:09.000000000 -0400
+++ mozilla/js/src/jslock.cpp	2008-07-18 15:09:27.000000000 -0400
@@ -910,14 +910,14 @@
 
 #ifdef NSPR_LOCK
 
-static JS_INLINE void
+void
 ThinLock(JSThinLock *tl, jsword me)
 {
     JS_ACQUIRE_LOCK((JSLock *) tl->fat);
     tl->owner = me;
 }
 
-static JS_INLINE void
+void
 ThinUnlock(JSThinLock *tl, jsword /*me*/)
 {
     tl->owner = 0;
@@ -1041,7 +1041,7 @@
     js_ResumeThread(tl);
 }
 
-static JS_INLINE void
+void
 ThinLock(JSThinLock *tl, jsword me)
 {
     JS_ASSERT(CURRENT_THREAD_IS_ME(me));
@@ -1055,7 +1055,7 @@
 #endif
 }
 
-static JS_INLINE void
+void
 ThinUnlock(JSThinLock *tl, jsword me)
 {
     JS_ASSERT(CURRENT_THREAD_IS_ME(me));
