diff -r b2de786e8c3d embedding/browser/gtk/tests/Makefile.in
--- a/embedding/browser/gtk/tests/Makefile.in	Fri Jul 24 07:54:09 2009 +0300
+++ b/embedding/browser/gtk/tests/Makefile.in	Fri Jul 24 08:52:30 2009 +0300
@@ -94,6 +94,8 @@ EXTRA_LIBS	+= \
 		$(TK_LIBS) \
 		$(NULL)
 
+EXTRA_LIBS += $(JEMALLOC_LIBS)
+
 ifeq ($(OS_ARCH), SunOS)
 ifndef GNU_CC
 # When using Sun's WorkShop compiler, including
diff -r b2de786e8c3d memory/jemalloc/Makefile.in
--- a/memory/jemalloc/Makefile.in	Fri Jul 24 07:54:09 2009 +0300
+++ b/memory/jemalloc/Makefile.in	Fri Jul 24 08:52:30 2009 +0300
@@ -118,6 +118,9 @@ else
 DIST_INSTALL = 1
 FORCE_STATIC_LIB= 1
 endif
+ifdef MOZ_PLATFORM_HILDON
+FORCE_SHARED_LIB= 1
+endif
 
 EXPORTS = jemalloc.h
 CSRCS   = jemalloc.c
diff -r b2de786e8c3d toolkit/components/startup/src/nsAppStartup.cpp
--- a/toolkit/components/startup/src/nsAppStartup.cpp	Fri Jul 24 07:54:09 2009 +0300
+++ b/toolkit/components/startup/src/nsAppStartup.cpp	Fri Jul 24 08:52:30 2009 +0300
@@ -522,6 +522,8 @@ nsAppStartup::Observe(nsISupports *aSubj
       EnterLastWindowClosingSurvivalArea();
       CloseAllWindows();
       ExitLastWindowClosingSurvivalArea();
+      if (!mShuttingDown)
+        Quit(eConsiderQuit);
     }
   } else if (!strcmp(aTopic, "xul-window-registered")) {
     EnterLastWindowClosingSurvivalArea();
