# https://bugzilla.mozilla.org/show_bug.cgi?id=260106
Index: mozilla/js/src/jsinterp.cpp
===================================================================
--- mozilla.orig/js/src/jsinterp.cpp
+++ mozilla/js/src/jsinterp.cpp
@@ -6049,16 +6049,30 @@
             JS_ASSERT(len <= regs.sp - fp->spbase);
             obj = js_NewArrayObject(cx, len, regs.sp - len, JS_TRUE);
             if (!obj)
                 goto error;
             regs.sp -= len - 1;
             STORE_OPND(-1, OBJECT_TO_JSVAL(obj));
           END_CASE(JSOP_NEWARRAY)
 
+          BEGIN_CASE(JSOP_HOLE)
+            PUSH_OPND(JSVAL_HOLE);
+          END_CASE(JSOP_HOLE)
+
+          BEGIN_CASE(JSOP_NEWARRAY)
+            len = GET_UINT24(regs.pc);
+            JS_ASSERT(len <= regs.sp - fp->spbase);
+            obj = js_NewArrayObject(cx, len, regs.sp - len);
+            if (!obj)
+                goto error;
+            regs.sp -= len - 1;
+            STORE_OPND(-1, OBJECT_TO_JSVAL(obj));
+          END_CASE(JSOP_NEWARRAY)
+
           BEGIN_CASE(JSOP_NEWINIT)
             i = GET_INT8(regs.pc);
             JS_ASSERT(i == JSProto_Array || i == JSProto_Object);
             obj = (i == JSProto_Array)
                   ? js_NewArrayObject(cx, 0, NULL)
                   : js_NewObject(cx, &js_ObjectClass, NULL, NULL, 0);
             if (!obj)
                 goto error;
