Index: content/base/src/nsDOMAttribute.cpp
===================================================================
RCS file: /cvsroot/mozilla/content/base/src/nsDOMAttribute.cpp,v
retrieving revision 1.111
diff -u -8 -p -r1.111 nsDOMAttribute.cpp
--- mozilla/content/base/src/nsDOMAttribute.cpp	28 Mar 2008 14:09:01 -0000	1.111
+++ mozilla/content/base/src/nsDOMAttribute.cpp	13 May 2008 17:38:33 -0000
@@ -191,21 +197,17 @@ nsDOMAttribute::SetValue(const nsAString
 
   return rv;
 }
 
 NS_IMETHODIMP
 nsDOMAttribute::GetSpecified(PRBool* aSpecified)
 {
   NS_ENSURE_ARG_POINTER(aSpecified);
-
-  nsIContent* content = GetContentInternal();
-  *aSpecified = content && content->HasAttr(mNodeInfo->NamespaceID(),
-                                            mNodeInfo->NameAtom());
-
+  *aSpecified = PR_TRUE;
   return NS_OK;
 }
 
 NS_IMETHODIMP
 nsDOMAttribute::GetOwnerElement(nsIDOMElement** aOwnerElement)
 {
   NS_ENSURE_ARG_POINTER(aOwnerElement);
 
