# increasing size of UA string
Index: mozilla/modules/plugin/base/src/nsPluginHostImpl.cpp
===================================================================
--- mozilla.orig/modules/plugin/base/src/nsPluginHostImpl.cpp
+++ mozilla/modules/plugin/base/src/nsPluginHostImpl.cpp
@@ -2898,17 +2898,17 @@
 
   PLUGIN_LOG(PLUGIN_LOG_NORMAL,
   ("nsPluginHostImpl::ReloadPlugins End active_instance_count=%d\n",
   mActivePluginList.mCount));
 
   return rv;
 }
 
-#define NS_RETURN_UASTRING_SIZE 128
+#define NS_RETURN_UASTRING_SIZE 256
 
 
 ////////////////////////////////////////////////////////////////////////
 nsresult nsPluginHostImpl::UserAgent(const char **retstring)
 {
   static char resultString[NS_RETURN_UASTRING_SIZE];
   nsresult res;
 
