Index: mozilla/toolkit/content/about.xhtml
===================================================================
--- mozilla.orig/toolkit/content/about.xhtml
+++ mozilla/toolkit/content/about.xhtml
@@ -69,16 +69,17 @@
   <ul id="aboutPageList">
     <li>&about.copy.beforeLink; <a href="about:credits">&about.copy.linkTitle;</a> &about.copy.afterLink;</li>
     <li>&about.license.beforeLink; <a href="about:license">&about.license.linkTitle;</a> &about.license.afterLink;</li>
     <!-- Release notes link removed pending fix for bug 349985
     <li>&about.relnotes.beforeLink; <a id="releaseNotesURL" href="">&about.relnotes.linkTitle;</a> &about.relnotes.afterLink;</li>
     -->
     <li>&about.buildconfig.beforeLink; <a href="about:buildconfig">&about.buildconfig.linkTitle;</a> &about.buildconfig.afterLink;</li>
      <li>&about.config.beforeLink; <a href="about:config">&about.config.linkTitle;</a>&about.config.afterLink;</li>
+    <li>&about.system.beforeLink; <a href="about:system">&about.system.linkTitle;</a>&about.system.afterLink;</li>
     <script type="application/javascript">
       // get release notes URL from prefs
       try {
         var formatter = Components.classes["@mozilla.org/toolkit/URLFormatterService;1"]
                                   .getService(Components.interfaces.nsIURLFormatter);
         var releaseNotesURL = formatter.formatURLPref("app.releaseNotesURL");
         var relnotes = document.getElementById("releaseNotesURL");
         relnotes.setAttribute("href", releaseNotesURL);
Index: mozilla/toolkit/locales/en-US/chrome/global/about.dtd
===================================================================
--- mozilla.orig/toolkit/locales/en-US/chrome/global/about.dtd
+++ mozilla/toolkit/locales/en-US/chrome/global/about.dtd
@@ -11,13 +11,17 @@
 <!ENTITY about.relnotes.beforeLink    "Read the">
 <!ENTITY about.relnotes.linkTitle     "release notes">
 <!ENTITY about.relnotes.afterLink     "for this version.">
 
 <!ENTITY about.config.beforeLink "Change this application's">
 <!ENTITY about.config.linkTitle "internal settings">
 <!ENTITY about.config.afterLink " (unsupported).">
 
+<!ENTITY about.system.beforeLink "Baseline">
+<!ENTITY about.system.linkTitle "version information">
+<!ENTITY about.system.afterLink "for this product.">
+
 <!ENTITY about.buildconfig.beforeLink "See the">
 <!ENTITY about.buildconfig.linkTitle  "build configuration">
 <!ENTITY about.buildconfig.afterLink  "used for this version.">
 
 <!ENTITY about.buildIdentifier        "Build identifier: ">
