diff -r bbab36cfacc1 toolkit/content/about.xhtml
--- a/toolkit/content/about.xhtml	Thu Dec 03 11:22:34 2009 +0200
+++ b/toolkit/content/about.xhtml	Thu Dec 03 11:25:21 2009 +0200
@@ -59,11 +59,12 @@
 
 <body dir="&locale.dir;">
   <div id="aboutLogoContainer">
-    <a id="vendorURL" href="http://www.mozilla.org/">
-      <img src="about:logo" alt="&brandShortName;"/>
-#expand <p id="version">&about.version; __MOZ_APP_VERSION__</p>
+    <a id="vendorURL" href="http://browser.garage.maemo.org/">
+      <img height="105" src="about:logo" alt="&brandShortName;"/>
     </a>
   </div>
+<p style="color:blue;font-size: 250%;" id="packname">&brandShortName;</p>
+#expand <p id="version">&about.version; __MOZ_APP_VERSION__</p>
 
   <ul id="aboutPageList">
     <li>&about.copy.beforeLink; <a href="about:credits">&about.copy.linkTitle;</a> &about.copy.afterLink;</li>
@@ -72,6 +73,8 @@
     <li>&about.relnotes.beforeLink; <a id="releaseNotesURL" href="">&about.relnotes.linkTitle;</a> &about.relnotes.afterLink;</li>
     -->
     <li>&about.buildconfig.beforeLink; <a href="about:buildconfig">&about.buildconfig.linkTitle;</a> &about.buildconfig.afterLink;</li>
+     <li>&about.config.beforeLink; <a href="about:config">&about.config.linkTitle;</a>&about.config.afterLink;</li>
+    <li>&about.system.beforeLink; <a href="about:system">&about.system.linkTitle;</a>&about.system.afterLink;</li>
     <script type="application/javascript">
       // get release notes URL from prefs
       try {
diff -r bbab36cfacc1 toolkit/locales/en-US/chrome/global/about.dtd
--- a/toolkit/locales/en-US/chrome/global/about.dtd	Thu Dec 03 11:22:34 2009 +0200
+++ b/toolkit/locales/en-US/chrome/global/about.dtd	Thu Dec 03 11:25:21 2009 +0200
@@ -12,6 +12,14 @@
 <!ENTITY about.relnotes.linkTitle     "release notes">
 <!ENTITY about.relnotes.afterLink     "for this version.">
 
+<!ENTITY about.config.beforeLink "Change this application's">
+<!ENTITY about.config.linkTitle "internal settings">
+<!ENTITY about.config.afterLink " (unsupported).">
+
+<!ENTITY about.system.beforeLink "Baseline">
+<!ENTITY about.system.linkTitle "version information">
+<!ENTITY about.system.afterLink "for this product.">
+
 <!ENTITY about.buildconfig.beforeLink "See the">
 <!ENTITY about.buildconfig.linkTitle  "build configuration">
 <!ENTITY about.buildconfig.afterLink  "used for this version.">
diff -r bbab36cfacc1 toolkit/themes/pinstripe/global/netError.css
--- a/toolkit/themes/pinstripe/global/netError.css	Thu Dec 03 11:22:34 2009 +0200
+++ b/toolkit/themes/pinstripe/global/netError.css	Thu Dec 03 11:25:21 2009 +0200
@@ -13,7 +13,7 @@ html {
 
 body {
   margin: 0;
-  padding: 0 1em;
+  padding: 0 0em;
   color: -moz-FieldText;
   font: message-box;
 }
diff -r bbab36cfacc1 toolkit/themes/winstripe/global/about.css
--- a/toolkit/themes/winstripe/global/about.css	Thu Dec 03 11:22:34 2009 +0200
+++ b/toolkit/themes/winstripe/global/about.css	Thu Dec 03 11:25:21 2009 +0200
@@ -1,6 +1,5 @@
 html {
   background: -moz-Dialog;
-  padding: 0 1em;
   font: message-box;
 }
 
@@ -12,7 +11,7 @@ body {
   margin: 4em auto;
   border: 1px solid ThreeDShadow;
   -moz-border-radius: 10px;
-  padding: 3em;
+  padding: 0 0em;
   -moz-padding-start: 30px;
   background: -moz-Field;
 }
@@ -51,3 +50,7 @@ ul > li {
 th, td {
   padding: 0 5px;
 }
+
+*:not(body):not(html) {
+  margin: 1% ! important;
+}
diff -r bbab36cfacc1 toolkit/themes/winstripe/global/netError.css
--- a/toolkit/themes/winstripe/global/netError.css	Thu Dec 03 11:22:34 2009 +0200
+++ b/toolkit/themes/winstripe/global/netError.css	Thu Dec 03 11:25:21 2009 +0200
@@ -13,7 +13,7 @@ html {
 
 body {
   margin: 0;
-  padding: 0 1em;
+  padding: 0 0em;
   color: -moz-FieldText;
   font: message-box;
 }
@@ -42,11 +42,11 @@ ul {
   position: relative;
   min-width: 13em;
   max-width: 52em;
-  margin: 4em auto;
+  margin: 1em auto;
   border: 1px solid ThreeDShadow;
   -moz-border-radius: 10px;
-  padding: 3em;
-  -moz-padding-start: 30px;
+  padding: 1em;
+  -moz-padding-start: 5px;
   background: url("chrome://global/skin/icons/warning-large.png") left 0 no-repeat -moz-Field;
   -moz-background-origin: content;
 }
@@ -60,11 +60,11 @@ body[dir="rtl"] #errorPageContainer {
 }
 
 #errorTitle {
-  -moz-margin-start: 80px;
+  -moz-margin-start: 40px;
 }
 
 #errorLongContent {
-  -moz-margin-start: 80px;
+  -moz-margin-start: 40px;
 }
 
 #errorShortDesc > p {
